---
title: "VNO Paper Figures"
author: "Max Hills"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: html_notebook
---

>This file contains statistics and figures from the paper, "Molecular, Cellular, and Developmental Organization of the Mouse Vomeronasal organ at Single Cell Resolution" by Max Hills Jr., Limei Ma, Ai Fang, Thelma Chiremba, Seth Malloy, Allison Scott, Anoja Perera, and C. Ron Yu.    

***  

```{r setup, include=FALSE}

# R-Markdown Global Settings  

# Set-up a default image size of 10 X 10 inches. Suppress warning
# and error messages. Hide code.
knitr::opts_chunk$set(echo=FALSE, warning=FALSE, message=FALSE, comment="",
                      error=FALSE, fig.width=10, fig.height=10, cache=FALSE)

```

```{r libraries, include=FALSE}

# Library Loading  

suppressPackageStartupMessages({
  library(kit); library(tidyverse); library(stringr); library(plot3D) 
  library(ggplot2); library(data.table); library(org.Mm.eg.db)
  library(dplyr); library(Seurat); library(DBI); library(cowplot)
  library(Matrix); library(glmGamPoi); library(circlize)
  library(tibble); library(ggpubr); library(svglite); library(ggsignif)
  library(stats); library(vegan); library(scales); library(caret)
  library(patchwork); library(plotly); library(geomtextpath)
  library(openxlsx); library(gridExtra); library(ggrepel)
  library(GeneOverlap); library(biomaRt); library(ggpointdensity)
  library(slingshot); library(tradeSeq); library(SingleCellExperiment)
  library(viridis); library(UpSetR); library(pheatmap); library(msigdbr)
  library(fgsea); library(grid); library(IDPmisc); library(MASS)
  library(Rmisc); library(ggplotify); library(ggpmisc); library(scales)
  library(heatmaply); library(dendextend)
})

```

```{r}

# Global Variables

# The home directory
home_dir <- "/l/Yu/YuLab/Bioinformatics/projects/mhh/sc_VNO/"
# Directory for storing data objects at crucial stages
object_dir <- paste0(home_dir,"final.seu.objs/")
# Directory for images from the final analysis
image_dir <- paste0(home_dir,"final_images/")
adobe_images <- paste0(image_dir, "adobe_illustrator_svg_files/")
# Directory for images for the final paper
clustree_dir <- paste0(image_dir,"clustree/")
# A directory for TradeSeq fitGAM output
fitGAM_dir <- paste0(object_dir, "fitGAM/")
moe_dir <- "/l/Yu/YuLab/Bioinformatics/projects/mhh/molng_2407_2453/R_obj/"

# OPEN ```vno.seu.integrated```
load(file = paste0(object_dir,"vno.integrated.postCellID"))
#Load Neuronal Lineage Seurat Object  
load(file = paste0(object_dir,"neuron.integrated.postCellID"))

# Rename unknown_mVSN to sVSN
levels(vno.seu.integrated@active.ident)[levels(vno.seu.integrated@active.ident)=="unknown_mVSN"] <- "sVSN"
levels(neuron.integrated@active.ident)[levels(neuron.integrated@active.ident)=="unknown_mVSN"] <- "sVSN"
levels(vno.seu.integrated$cell_type)[levels(vno.seu.integrated$cell_type)=="unknown_mVSN"] <- "sVSN"
levels(neuron.integrated$cell_type)[levels(neuron.integrated$cell_type)=="unknown_mVSN"] <- "sVSN"

# Get color scheme for the whole vno dataset
vno.cell_type.colors <- hue_pal()(length(levels(vno.seu.integrated@active.ident)))
names(vno.cell_type.colors) <- levels(vno.seu.integrated@active.ident)

# Keep neuron dataset color scheme consistent with the whole vno color scheme, where possible  
neuron.cell_type.colors <- vno.cell_type.colors[c("V1R_mVSN","V2R_mVSN","mOSN","sVSN","V1R_iVSN","V2R_iVSN","GBC")]
neuron.cell_type.colors <- setNames(c(neuron.cell_type.colors,"#00B0F6","#9590FF","#E76BF3"),
                                    c(names(neuron.cell_type.colors),"iOSN","late_INP","early_INP"))

mature.neurons <- subset(neuron.integrated,idents=c("V1R_mVSN","V2R_mVSN","mOSN","sVSN"))
mature.cell_type.colors <- neuron.cell_type.colors[c("V1R_mVSN","V2R_mVSN","mOSN","sVSN")]

global.vars <- c(ls(),"global.vars")

```

***  

# Single cell transcriptomic profile of the whole vomeronasal organ {.tabset .tabset-pills}  

## What cell types, in what quantities, do we find in the whole VNO?  

```{r}

print(paste("There are",dim(vno.seu.integrated)[2],"cells in the whole VNO dataset."))
knitr::kable(sort(table(vno.seu.integrated$cell_type), decreasing = TRUE),col.names=c("cell type","nCells"))

```

## 2D PCA of Clusters  

```{r}

vno.seu.integrated$seurat_clusters <- factor(vno.seu.integrated$seurat_clusters,
                                             levels=as.character(0:30))

```


```{r}

Idents(vno.seu.integrated) <- vno.seu.integrated$seurat_clusters
vno.2d.cluster.pca <- DimPlot(vno.seu.integrated, reduction="pca", label=TRUE, shuffle=TRUE, repel=TRUE) +
  ggtitle("VNO PCA Clusters")

vno.2d.cluster.pca

```

## 3D PCA of Clusters 

```{r include=FALSE}

# Prepare a dataframe for cell plotting. The last column will determine the color of the datapoints.
vno.3d.pca.df <- FetchData(object = vno.seu.integrated, 
                           vars = c("PC_1", "PC_2", "PC_3", "seurat_clusters"))

```

```{r}

nColors.clusters <- length(levels(vno.3d.pca.df$seurat_clusters))
vno.cluster.colors <- hue_pal()(nColors.clusters)

```

```{r include=FALSE}

# Plot your data
vno.3d.pca.cluster <- plot_ly(data = vno.3d.pca.df, 
                              x = ~PC_1, y = ~PC_2, z = ~PC_3, 
                              color = ~seurat_clusters, 
                              type = "scatter3d", 
                              mode = "markers", 
                              colors = vno.cluster.colors,
                              marker = list(size = 1, width=1), # controls size of points
                              text = ~seurat_clusters,
                              hoverinfo="text") %>% layout(title="VNO PCA Clusters")

```

```{r, fig.width=14, fig.height=14}

vno.3d.pca.cluster

```

## 2D UMAP Clusters  

```{r}

fig_1A <- DimPlot(vno.seu.integrated, reduction = "umap", label = TRUE, shuffle = TRUE) +
  ggtitle("VNO Clusters") + theme(aspect.ratio=1)

fig_1A

```

## 3D UMAP Clusters  

```{r include=FALSE}

DefaultAssay(vno.seu.integrated) <- "integrated"
# Re-run UMAP in 50 dimensions and retain calculations for the first 3 dimensions
vno.3d.umap.seu <- RunUMAP(vno.seu.integrated,
                            dims = 1:26,
                            n.components = 3L)

# Prepare a dataframe for cell plotting. The last column will determine the color of the datapoints.
vno.3d.umap.df <- FetchData(object=vno.3d.umap.seu,vars=c("UMAP_1","UMAP_2","UMAP_3","seurat_clusters"))

rm(vno.3d.umap.seu); invisible(gc())

```

```{r}

nColors.clusters <- length(levels(vno.3d.umap.df$seurat_clusters))
vno.cluster.colors <- hue_pal()(nColors.clusters)

```

```{r include=FALSE}

# Plot your data
vno.3d.umap.cluster <- plot_ly(data = vno.3d.umap.df, 
                               x = ~UMAP_1, y = ~UMAP_2, z = ~UMAP_3, 
                               color = ~seurat_clusters, 
                               type = "scatter3d", 
                               mode = "markers", 
                               colors = vno.cluster.colors,
                               marker = list(size = 1, width=1), # controls size of points
                               text = ~seurat_clusters,
                               hoverinfo="text") %>% layout(title="VNO UMAP Clusters")

```

```{r, fig.width=14, fig.height=14}

vno.3d.umap.cluster

```

## UMAP visualization of integrated cell-type clusters for whole-VNO single-cell RNA-seq  

```{r}

Idents(vno.seu.integrated) <- vno.seu.integrated$cell_type

```

>In two-dimensional UMAP space, 18 cell clusters can be clearly identified.

```{r, fig.width=10, fig.height=8}

fig1_a <- DimPlot(vno.seu.integrated, reduction = "umap", label = TRUE, shuffle = TRUE) +
  ggtitle("Vomeronasal Organ Cell Clusters") +
  theme(aspect.ratio=1)

svglite(filename = paste0(adobe_images,"fig1_a.svg"), width=10, height=8)
fig1_a
invisible(dev.off())
fig1_a

```

## 3D UMAP Cell-types  

```{r include=FALSE}

DefaultAssay(vno.seu.integrated) <- "integrated"
# Re-run UMAP in 50 dimensions and retain calculations for the first 3 dimensions
vno.3d.umap.seu <- RunUMAP(vno.seu.integrated,
                           dims = 1:26,
                           n.components = 3L)

# Prepare a dataframe for cell plotting. The last column will determine the color of the datapoints.
vno.3d.umap.df <- FetchData(object=vno.3d.umap.seu,vars=c("UMAP_1","UMAP_2","UMAP_3","cell_type"))

rm(vno.3d.umap.seu); invisible(gc())

```

```{r}

nColors.clusters <- length(levels(vno.3d.umap.df$cell_type))
vno.cluster.colors <- hue_pal()(nColors.clusters)

```

```{r include=FALSE}

# Plot your data
vno.3d.umap.cluster <- plot_ly(data = vno.3d.umap.df, 
                               x = ~UMAP_1, y = ~UMAP_2, z = ~UMAP_3, 
                               color = ~cell_type, 
                               type = "scatter3d", 
                               mode = "markers", 
                               colors = vno.cluster.colors,
                               marker = list(size = 1, width=1), # controls size of points
                               text = ~cell_type,
                               hoverinfo="text") %>% layout(title="VNO UMAP Cell Types")

```

```{r, fig.width=14, fig.height=14}

vno.3d.umap.cluster

```

## Cell-type marker-gene normalized expression across the cell clusters  

>These clusters were curated using known cell markers (Fig. 1b).

```{r, include=FALSE}

# Our marker genes to plot  
plot.genes <- c("Omp","Gnai2","Gng13","Meis2","Gnao1","Robo2","Tfap2e","Gnal","Cnga2","P2ry14","Xist",
                "Gap43","Stmn2","Neurog1","Neurod1","Ccnd1","Ascl1","Krt5","Krt15","Sox9","Hepacam2",
                "Fezf2","Sox2","Plp1","S100b","Cdh5","Cd34","Acta2","Col1a2","Ctss","Dock2","Cd209a",
                "S100a9","Trbc2")

```


```{r, incude=FALSE}

# Get cell expression data for the goi, the cluster, and the cell-type
vno.marker_genes.df <- FetchData(object = vno.seu.integrated, 
                                 vars = c(plot.genes, "seurat_clusters", "cell_type"))
vno.marker_genes.df$Cell <- rownames(vno.marker_genes.df)

# Reshape the data for plotting
vno.marker_genes.df <- reshape2::melt(vno.marker_genes.df, 
                                      id.vars = c("Cell","seurat_clusters","cell_type"), 
                                      measure.vars = plot.genes,
                                      variable.name = "Gene", value.name = "Expr")

```


```{r, include=FALSE}

# Order the gene factor levels
vno.marker_genes.df$Gene <-
  factor(vno.marker_genes.df$Gene,
         levels=c("Omp","Gnai2","Gng13","Meis2","Gnao1","Robo2","Tfap2e","Gnal","Cnga2","P2ry14","Xist",
                "Gap43","Stmn2","Neurog1","Neurod1","Ccnd1","Ascl1","Krt5","Krt15","Sox9","Hepacam2",
                "Fezf2","Sox2","Plp1","S100b","Cdh5","Cd34","Acta2","Col1a2","Ctss","Dock2","Cd209a",
                "S100a9","Trbc2"))

# Order the seurat_clusters factor levels
vno.marker_genes.df$seurat_clusters <- factor(vno.marker_genes.df$seurat_clusters,
                                              levels=c("0","1","2","3","13","23","5","10","15","21","29",
                                                      "9","12","6","7","19","26","17","28","27","4","11",
                                                      "18","30","25","24","16","8","20","14","22"))

```


```{r, include=FALSE}

# Get ordered legend data
legend.df <- data.frame(cluster=levels(vno.marker_genes.df$seurat_clusters),
                        cell_type=c(rep("V1R_mVSN",6),rep("V2R_mVSN",3),rep("mOSN",2),rep("sVSN",2),
                                    rep("V1R_iVSN",1),"V2R_iVSN","INP","GBC",rep("HBC",2),"MV",rep("SUS",4),
                                    "OEC","Endothelial",rep("LP",1),rep("Microglia",1),"Microglia_rp",
                                    "Immune_Fpr","T_cells"))

legend.df$cluster <- factor(legend.df$cluster, levels=levels(vno.marker_genes.df$seurat_clusters))
legend.df$cell_type <- factor(legend.df$cell_type,
                              levels=c("V1R_mVSN","V2R_mVSN","mOSN","sVSN","V1R_iVSN","V2R_iVSN",
                                 "INP","GBC","HBC","MV","SUS","OEC","Endothelial","LP","Microglia",
                                 "Microglia_rp","Immune_Fpr","T_cells"))

```


```{r fig.width=25, fig.height=20}

fig1_b.part1 <- ggplot(vno.marker_genes.df, aes(x=seurat_clusters, Expr, fill=cell_type)) +
  geom_violin(scale = "width", adjust = 1, trim = TRUE) +
  scale_y_continuous(expand = c(0, 0), position="right", labels = function(x)
                     c(rep(x = "", times = length(x)-2), x[length(x) - 1], "")) +
  facet_grid(rows = vars(Gene), scales = "free", switch = "y") +
  theme(legend.position = "none", panel.spacing = unit(0, "lines"),
        plot.title = element_text(size=56, face="bold", hjust=0.7),
        panel.background = element_rect(fill = NA, color = "black"),
        strip.background = element_blank(),
        strip.text.y.left = element_text(angle = 0, size=40),
        axis.title.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.x = element_blank(),
        axis.text.y = element_text(size=40),
        axis.title.y = element_text(size=50, vjust=0.5)) +
  ggtitle("VNO Cell-Type Marker-Gene Normalized Expression by Cluster") + ylab("Normalized Gene Expression")

fig1_b.part2 <- ggplot(legend.df, aes(x = cluster, y = 1, fill = cell_type)) + 
  geom_tile(width=0.97) + 
  scale_fill_manual(values = vno.cell_type.colors) + scale_y_continuous(expand = c(0, 0)) +
  guides(fill = guide_legend(direction = "vertical", label.position = "right",
                       title.theme = element_blank(), keyheight = 0.5, nrow = 3, byrow = TRUE)) +
  theme(legend.key.size = unit(2, 'lines'),
        legend.text = element_text(margin = margin(r=2, unit='pt'), size=30),
        legend.position = "bottom",
        legend.justification = "center",
        legend.margin = margin(0,0,0,0),
        legend.box.margin = margin(-10,5,0,0),
        panel.spacing = unit(3, "lines"),
        panel.background = element_blank(),
        panel.border = element_blank(),
        plot.background = element_blank(),
        plot.margin = margin(0, 7, 7, 7, "pt"),
        axis.text.x = element_text(angle=60, hjust = 0.5, vjust = 0.5, color = "black", size=40),
        axis.title.x = element_text(size=50, vjust=0.5),
        axis.title.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.text.y = element_blank()) + xlab("Cluster")

fig1_b <- plot_grid(fig1_b.part1, fig1_b.part2, ncol = 1, rel_heights = c(0.83, 0.17), 
                    align = "hv", axis = "lr")
svglite(filename = paste0(adobe_images,"fig1_b.svg"),width=25, height=20)
fig1_b
invisible(dev.off())

fig1_b

```

## UMAP of cell-type clusters split by age  

```{r, fig.width=16,fig.height=8}

fig1_c <- DimPlot(vno.seu.integrated, reduction = "umap", shuffle = TRUE, split.by = "age", ncol = 2) +
  theme_void() + theme(aspect.ratio=1) + NoLegend()

svglite(filename = paste0(adobe_images,"fig1_c.svg"), width=10, height=5)
fig1_c
invisible(dev.off())
fig1_c

```

## UMAP of cell-type clusters split by sex  

```{r, fig.width=16,fig.height=8}

fig1_d <- DimPlot(vno.seu.integrated, reduction = "umap", shuffle = TRUE, split.by = "sex", ncol = 2) +
  theme_void() + theme(aspect.ratio=1) + NoLegend()

svglite(filename = paste0(adobe_images,"fig1_d.svg"), width=10, height=5)
fig1_d
invisible(dev.off())
fig1_d

```

## Spatial location of individual VNO cells color-coded according to cell type prediction based on the spatial transcriptomic analysis  

```{r, eval=FALSE}

load(file=paste0(object_dir,"resolve.seu.POSTpredict"))
load(file=paste0(object_dir,"resolve.datasets.POSTpredict"))
load(file=paste0(object_dir,"resolve.predictions.assay"))

```

```{r}

resolve.cell.types <- 
  as.factor(unlist(lapply(colnames(resolve.predictions.assay), function(cell){
  rownames(resolve.predictions.assay)[which.max(x=resolve.predictions.assay[ ,cell])]
})))

```

```{r}

Idents(resolve.seu.integrated) <- resolve.cell.types
levels(resolve.seu.integrated@active.ident)[levels(resolve.seu.integrated@active.ident)=="unknown-mVSN"] <- "sVSN"

levels(resolve.seu.integrated@active.ident) <- gsub("-","_", levels(resolve.seu.integrated@active.ident))

resolve.seu.integrated@active.ident <-
  factor(as.factor(resolve.seu.integrated@active.ident), 
         levels = c("V1R_mVSN","V2R_mVSN","mOSN","sVSN",
                    "V1R_iVSN","V2R_iVSN","INP","GBC","HBC","MV",
                    "SUS","OEC","Endothelial","LP","Microglia",
                    "Microglia_rp","Immune_Fpr","T_cells"))

```

```{r}

resolve.cell_type.colors <- hue_pal()(length(levels(resolve.seu.integrated@active.ident)))

names(resolve.cell_type.colors) <- levels(resolve.seu.integrated@active.ident)

resolve.cell_type.colors["V1R_mVSN"] <- "#00FF00"   #green
resolve.cell_type.colors["V2R_mVSN"] <- "#FF7F50"   #coral
resolve.cell_type.colors["sVSN"] <- "#8F00FF"       #violet 
resolve.cell_type.colors["LP"] <- "#0000FF"         #blue   
resolve.cell_type.colors["SUS"] <- "#FF0000"        #red
resolve.cell_type.colors["HBC"] <- "#FFFF00"        #yellow
resolve.cell_type.colors["V1R_iVSN"] <- "#DFFF00"   #chartreuse 
resolve.cell_type.colors["V2R_iVSN"] <- "#FFA500"   #orange

```

```{r}

colors <- c("V1R_mVSN"="#45B500","V2R_mVSN"="#E7851E","mOSN"="#D39200","sVSN"="#9C8DFF",
            "V1R_iVSN"="#00C087","V2R_iVSN"="#D09400","INP"="#FF61C7","GBC"="#EFFD5F", 
            "HBC"="#FFFF00","MV"="#00C0B2","SUS"="#F8766D","OEC"="#00BCD6",
            "Endothelial"="#00B3F2","LP"="#29A3FF","Microglia"="#00BC51","Microglia_rp"="#D277FF",
            "Immune_Fpr"="#B2A100","T_cells"="#F166E8")

```

```{r}

fig1_f.samples <- resolve.datasets[["slide7.B2_2"]]
cell_names <- colnames(fig1_f.samples)
cell_types <- resolve.seu.integrated@active.ident[cell_names]
fig1_f.samples <- AddMetaData(fig1_f.samples, cell_types, col.name="cell_type")
fig1_f <- SpatialDimPlot(object=fig1_f.samples, group.by="cell_type", 
                         pt.size.factor=2, cols = colors) +
  theme(aspect.ratio=1, legend.title=element_text(size=24),
        legend.text=element_text(size=18),
        legend.box.background=element_rect(fill="blue",colour="transparent"),
        legend.key = element_rect(fill="transparent",colour="transparent")) 

```

```{r, fig.width=12, fig.height=10}

svglite(filename = paste0(adobe_images,"fig1_f.svg"), width=10, height=10)
fig1_f 
invisible(dev.off())
fig1_f

```

## Location of cell belonging to HBC, GBC, INP, and LP cell types, respectively. Heat indicates confidence of predicted values  

```{r}

blank_theme <- theme(panel.background = element_rect(fill = "transparent", colour = NA),
                     plot.background = element_rect(fill = "transparent", colour = NA),
                     panel.grid = element_blank(),
                     panel.border = element_blank(),
                     plot.margin = unit(c(-1, 0, 0, 0), "null"),
                     panel.margin = unit(c(-1, 0, 0, 0), "null"),
                     axis.title=element_blank(),
                     axis.text=element_text(size=20),
                     plot.title=element_text(size=32),
                     legend.text=element_text(size=12,angle=90),
                     legend.title=element_text(size=20),
                     aspect.ratio=1)

```

```{r include=FALSE}

fig1_g.part_1 <- SpatialFeaturePlot(fig1_f.samples, features=rownames(resolve.predictions.assay)[10],
                                    pt.size.factor=4) + blank_theme +
  scale_fill_gradient(low="grey85",high="magenta")

fig1_g.part_2 <- SpatialFeaturePlot(fig1_f.samples, features=rownames(resolve.predictions.assay)[17],
                                    pt.size.factor=4) + blank_theme +
  scale_fill_gradient(low="grey85",high="magenta")

fig1_g.part_3 <- SpatialFeaturePlot(fig1_f.samples, features=rownames(resolve.predictions.assay)[8],
                                    pt.size.factor=4) + blank_theme +
  scale_fill_gradient(low="grey85",high="magenta")

fig1_g.part_4 <- SpatialFeaturePlot(fig1_f.samples, features=rownames(resolve.predictions.assay)[11],
                                    pt.size.factor=4) + blank_theme +
  scale_fill_gradient(low="grey85",high="magenta")


fig1_g <- ggarrange(fig1_g.part_1, fig1_g.part_2, fig1_g.part_3, fig1_g.part_4, ncol=4, nrow=1)

```

```{r, fig.width=28, fig.height=7}

svglite(filename = paste0(adobe_images,"fig1_g.svg"), width=28, height=7)
fig1_g 
invisible(dev.off())
fig1_g

```

# {-}  

```{r, include=FALSE}

current.variables <- c(ls(),"current.variables")
rm(list=setdiff(current.variables,c(global.vars)))
invisible(gc())

```

# Novel neuronal lineage {.tabset .tabset-pills}  

```{r, include=FALSE}

# Differential Expression of Genes Across Cell-Types
V1R_iVSN_V2R_iVSN.DEGs <- 
  read.xlsx(xlsxFile = paste0(object_dir,"V1R_iVSN_V2R_iVSN.DEGs.xlsx"),rowNames=TRUE)
V1R_iVSN_iOSN.DEGs <- 
  read.xlsx(xlsxFile = paste0(object_dir,"V1R_iVSN_iOSN.DEGs.xlsx"),rowNames=TRUE)
V2R_iVSN_iOSN.DEGs <- 
  read.xlsx(xlsxFile = paste0(object_dir,"V2R_iVSN_iOSN.DEGs.xlsx"),rowNames=TRUE)
V1R_mVSN_V2R_mVSN.DEGs <- 
  read.xlsx(xlsxFile = paste0(object_dir,"V1R_mVSN_V2R_mVSN.DEGs.xlsx"),rowNames=TRUE)
V1R_iVSN_late_INP.DEGs <- 
  read.xlsx(xlsxFile = paste0(object_dir,"V1R_iVSN_late_INP.DEGs.xlsx"),rowNames=TRUE)
V2R_iVSN_late_INP.DEGs <- 
  read.xlsx(xlsxFile = paste0(object_dir,"V2R_iVSN_late_INP.DEGs.xlsx"),rowNames=TRUE)
unknown_mVSN_V1R_mVSN.DEGs <- 
  read.xlsx(xlsxFile = paste0(object_dir,"unknown_mVSN_V1R_mVSN.DEGs.xlsx"),rowNames=TRUE)
unknown_mVSN_V2R_mVSN.DEGs <- 
  read.xlsx(xlsxFile = paste0(object_dir,"unknown_mVSN_V2R_mVSN.DEGs.xlsx"),rowNames=TRUE)
unknown_mVSN_mOSN.DEGs <- 
  read.xlsx(xlsxFile = paste0(object_dir,"unknown_mVSN_mOSN.DEGs.xlsx"),rowNames=TRUE)

# Gene-set Enrichment Analysis of GO Terms for Unknown mVSNs vs Other Neuron Types
unknown_mVSN_V1R_mVSN.BP.res <- 
  openxlsx::read.xlsx(xlsxFile=paste0(object_dir,"unknown_mVSN_V1R_mVSN.BP.res.xlsx"),rowNames=TRUE)
unknown_mVSN_V2R_mVSN.BP.res <- 
  openxlsx::read.xlsx(xlsxFile=paste0(object_dir,"unknown_mVSN_V2R_mVSN.BP.res.xlsx"),rowNames=TRUE)
unknown_mVSN_V1R_mVSN.MF.res <- 
  openxlsx::read.xlsx(xlsxFile=paste0(object_dir,"unknown_mVSN_V1R_mVSN.MF.res.xlsx"),rowNames=TRUE)
unknown_mVSN_V2R_mVSN.MF.res <- 
  openxlsx::read.xlsx(xlsxFile=paste0(object_dir,"unknown_mVSN_V2R_mVSN.MF.res.xlsx"),rowNames=TRUE)
unknown_mVSN_V1R_mVSN.CC.res <- 
  openxlsx::read.xlsx(xlsxFile=paste0(object_dir,"unknown_mVSN_V1R_mVSN.CC.res.xlsx"),rowNames=TRUE)
unknown_mVSN_V2R_mVSN.CC.res <- 
  openxlsx::read.xlsx(xlsxFile=paste0(object_dir,"unknown_mVSN_V2R_mVSN.CC.res.xlsx"),rowNames=TRUE)

load(file=paste0(object_dir,"resolve.datasets.POSTpredict"))
load(file=paste0(object_dir,"resolve.predictions.assay"))

```

## UMAP visualization of cell-type clusters for the neuronal lineage  

```{r}

fig2_a <- DimPlot(neuron.integrated, reduction="umap", label=TRUE, shuffle=TRUE, cols=neuron.cell_type.colors) +
  ggtitle("Neuronal Lineage") + NoLegend() +
  theme(aspect.ratio=1)

svglite(filename = paste0(adobe_images,"fig2_a.svg"), width=10, height=10)
fig2_a
invisible(dev.off())
fig2_a

```

>How many mature neurons are OSNs?  

```{r, echo=TRUE}

all_mature_neurons <- c("V1R_mVSN","V2R_mVSN","mOSN","sVSN")
table(neuron.integrated@active.ident)[c("mOSN")]/sum(table(neuron.integrated@active.ident)[all_mature_neurons])

```

```{r}

S1_theme <- theme(plot.title=element_text(size=32),
                  axis.title=element_text(size=24),
                  axis.text =element_text(size=20),
                  legend.text=element_text(size=20),
                  legend.position="right",
                  aspect.ratio=1,
                  plot.margin=unit(c(-1,0.25,-1,0.25),"lines"),
                  panel.spacing=unit(0, "lines"))

```

## Normalized Gap43 and Stmn2 expression in the neuronal lineage  

>In all lineages the cells can be distinguished to have both immature and mature populations, as determined by the expression of Gap43 and Stmn2 genes (Fig. S1).  

```{r, fig.width=15,fig.height=7.5}

figS_1AB <- FeaturePlot(object=neuron.integrated, features=c("Gap43","Stmn2"), order=TRUE,
                      cols=c("cyan","magenta"),keep.scale='all') & S1_theme

figS_1AB

```

## Normalized Xist expression in the neuronal lineage, split by sex  

```{r,fig.width=15,fig.height=7.5}

figS_1C <- FeaturePlot(object=neuron.integrated,features="Xist",order=TRUE,
            cols=c("cyan","magenta"),split.by="sex",keep.scale='all') & S1_theme &
  theme(axis.title.y=element_blank())

figS_1C

```

## Expression of Gnai2 and Gnao1 in the neuronal lineage  

```{r, fig.width=16, fig.height=4}

fig2_b <- FeaturePlot(neuron.integrated, features=c("Gnai2","Gnao1"), order=TRUE, blend=TRUE) +
  theme(aspect.ratio=1)

svglite(filename = paste0(adobe_images,"fig2_b.svg"),width=16, height=4)
fig2_b
invisible(dev.off())

fig2_b

```

## Violin plot of normalized Gnai2 expression in mature V2R neurons, split by sample  

```{r}

# Get dataframe for plotting features of sVSNs
sVSN.compare.df <- FetchData(mature.neurons,vars=c("cell_type","nFeature_RNA","nCount_RNA",
                                                   "percent.ribo","P2ry14","Ap1s3",
                                                   "Cspp1","Muc2","Obp2a","Obp2b","Lcn3",## Up-regulated genes
                                                   "Omp","Tuba1a","Ptma"))               ## Down-regulated genes

my_comparisons = list(c("V1R_mVSN","V2R_mVSN"),c("V2R_mVSN","mOSN"),c("mOSN","sVSN"),
                      c("V1R_mVSN","mOSN"),c("V2R_mVSN","sVSN"),c("V1R_mVSN","sVSN"))

my_theme <- theme(plot.title=element_text(hjust=0.5,size=32,face='bold'), # Center the title
                  axis.title.y=element_text(size=24),
                  axis.text.y=element_text(size=20),
                  axis.title.x=element_blank(),
                  axis.ticks.x=element_blank(), 
                  axis.text.x=element_blank(),aspect.ratio=1,
                  legend.position="right",
                  legend.key.size=unit(8,'mm'),
                  legend.title.align=0,
                  legend.title=element_text(size=24),
                  legend.text=element_text(size=20),
                  plot.margin=unit(c(0,0.25,0,0.25),units='line'),
                  panel.spacing=unit(c(0,0,0,0),units='line'))

my_theme2 <- theme(plot.title=element_text(hjust=0.5,size=32,face='bold'),
                   axis.title=element_text(size=24),
                   axis.text=element_text(size=20),
                   legend.position="none",aspect.ratio=1)

```

```{r}

figS_1D <- VlnPlot(object=subset(neuron.integrated,idents=c("V2R_mVSN")), 
        features="Gnai2", group.by="orig.ident") +
  my_theme2 + ggtitle("Gnai2 in Mature V2R Neurons")
figS_1D

```

## Expression of Cnga2 and Gnal in the OSN lineage  

```{r, fig.width=16, fig.height=8}

fig2_c <- FeaturePlot(neuron.integrated, features=c("Cnga2","Gnal"), 
            cols=c("cyan3","magenta"), order=TRUE, ncol=2) +
  theme(aspect.ratio = 1)

svglite(filename = paste0(adobe_images,"fig2_c.svg"), width=10, height=5)
fig2_c
invisible(dev.off())

fig2_c

```

## Location of mOSNs  

```{r}

fig2d.samples <- resolve.datasets[["slide7.B2_2"]]

```

```{r}

fig2_OSN <- SpatialFeaturePlot(fig2d.samples, features=rownames(resolve.predictions.assay)[12],
                                pt.size.factor=4,) + theme(aspect.ratio=1,legend.position="bottom",
                                                          legend.title=element_blank(), 
                                                          plot.title = element_text(hjust=0.5,size=36)) +
  scale_fill_gradient(low="grey85",high="magenta") + ggtitle("mOSNs")

svglite(filename = paste0(adobe_images,"fig2_OSN.svg"), width=9, height=9)
fig2_OSN
invisible(dev.off())
fig2_OSN

```

## Total number of genes, counts, and percent ribosomal gene expression detected in mature neurons, split by cell type  

```{r,fig.width=18,fig.height=6}

figS_1E <- ggplot(sVSN.compare.df,aes(x=cell_type,y=nFeature_RNA,fill=cell_type)) + geom_violin() +
  scale_fill_manual("Cell Type",values=mature.cell_type.colors) + theme_classic() +
  stat_summary(fun.y=mean,geom='point',size=10,colour="black",shape=95,show_guide=FALSE) +
  stat_compare_means(comparisons=my_comparisons,tip.length=0.01,
                     method="wilcox.test",label="p.signif",vjust=0.3) +
  ggtitle("Total Genes") + ylab("count") + my_theme + NoLegend()

figS_1F <- ggplot(sVSN.compare.df,aes(x=cell_type,y=nCount_RNA,fill=cell_type)) + geom_violin() +
  scale_fill_manual("Cell Type",values=mature.cell_type.colors) + theme_classic() +
  stat_summary(fun.y=mean,geom='point',size=10,colour="black",shape=95,show_guide=FALSE) +
  stat_compare_means(comparisons=my_comparisons,tip.length=0.01,
                     method="wilcox.test",label="p.signif",vjust=0.3) +
  ggtitle("Total Counts") + ylab("count") + my_theme + NoLegend()

figS_1G <- ggplot(sVSN.compare.df,aes(x=cell_type,y=percent.ribo,fill=cell_type)) + geom_violin() +
  scale_fill_manual("Cell Type",values=mature.cell_type.colors) + theme_classic() +
  stat_summary(fun.y=mean,geom='point',size=10,colour="black",shape=95,show_guide=FALSE) +
  stat_compare_means(comparisons=my_comparisons,tip.length=0.01,
                     method="wilcox.test",label="p.signif",vjust=0.3) +
  ggtitle("Ribosomal (Percent)") + ylab("percent") + my_theme + NoLegend()

ggarrange(figS_1E, figS_1F, figS_1G, ncol=3, common.legend=TRUE, legend="bottom", align="hv")

```

## Location of sVSNs in a VNO slice  

```{r}

fig2_sVSN <- SpatialFeaturePlot(fig2d.samples, features=rownames(resolve.predictions.assay)[5],
                                pt.size.factor=4,) + theme(aspect.ratio=1,legend.position="bottom",
                                                          legend.title=element_blank(), 
                                                          plot.title = element_text(hjust=0.5,size=36)) +
  scale_fill_gradient(low="grey85",high="magenta") + ggtitle("sVSNs")

svglite(filename = paste0(adobe_images,"fig2_sVSN.svg"), width=9, height=9)
fig2_sVSN
invisible(dev.off())
fig2_sVSN

```

## Heatmap of normalized expression for a select set of mutually differentially expressed genes between sVSNs and mature V1Rs, V2Rs, and OSNs  

```{r}

svsn.goi <- c("Cspp1","D10Wsu102e","Gm16579","Neat1","P2ry14","Rbm4b","Wnk1","Zfp738",
              "Cox6a1","Cox7a2","Jund","Map1lc3b","Ndufa2","Ndufa7","Omp","Prxl2b")

mSN.matrix <- GetAssayData(subset(neuron.integrated,idents=c("V1R_mVSN","V2R_mVSN","mOSN","sVSN")),slot='data')
sVSN.top.df <- as.data.frame(mSN.matrix[svsn.goi, ])

# The cell IDs for ALL mature sensory neurons
mature.neuron.cellIDs <- WhichCells(object=neuron.integrated, idents=c("V1R_mVSN","V2R_mVSN","mOSN","sVSN"))
# Get corresponding cell-type metadata for column annotations
mature.neuron.annotation <- data.frame(cell_type=neuron.integrated@meta.data[mature.neuron.cellIDs,c("cell_type"),drop=FALSE])
svsn.cell.order <- order(mature.neuron.annotation$cell_type)
mature.neuron.annotation <- as.data.frame(mature.neuron.annotation[svsn.cell.order, ,drop=FALSE])
sVSN.top.df <- sVSN.top.df[ ,svsn.cell.order]

rm(mSN.matrix,mature.neuron.cellIDs)

```

```{r, fig.width=12, fig.height=8}

# Custom Palette for gene expression
colors <- c("#007FFF", "#2E8B57", "#FFFF66", "#FFA500", "#FF0000")
palette <- colorRampPalette(colors)

# Custom Palette for Cell Types
fig2_g.palette <-wesanderson::wes_palette("Darjeeling1")[2:5]
names(fig2_g.palette) <- c("V1R_mVSN","V2R_mVSN","mOSN","sVSN")

# Annotation Color Schematic 
vsn.ann_colors <- list(cell_type=fig2_g.palette)

fig2_g <- pheatmap(mat=sVSN.top.df, 
                   color=palette(n=50), fontsize=10, treeheight_col = 0,
                   cluster_rows=TRUE, cluster_cols=FALSE, #scale="row", 
                   annotation_col=mature.neuron.annotation, show_colnames=FALSE,
                   show_rownames=TRUE, annotation_colors=vsn.ann_colors)

```

## Heatmap of 503 significant (padj ≤ 0.001, fold-change ≥ 1.5) differentially expressed genes in sVSNs  

```{r}

# The gene names for sVSN significantly up-regulated genes against V1R/V2R VSNs and OSNs
sVSN_OSN.up_genes <- rownames(unknown_mVSN_mOSN.DEGs[unknown_mVSN_mOSN.DEGs$p_val_adj < 0.001 &
                                                       unknown_mVSN_mOSN.DEGs$avg_log2FC > log2(1.5), ])
sVSN_V1R.up_genes <- rownames(unknown_mVSN_V1R_mVSN.DEGs[unknown_mVSN_V1R_mVSN.DEGs$p_val_adj < 0.001 & 
                             unknown_mVSN_V1R_mVSN.DEGs$avg_log2FC > log2(1.5), ])
sVSN_V2R.up_genes <- rownames(unknown_mVSN_V2R_mVSN.DEGs[unknown_mVSN_V2R_mVSN.DEGs$p_val_adj < 0.001 &
                             unknown_mVSN_V2R_mVSN.DEGs$avg_log2FC > log2(1.5), ])

sVSN.up_genes <- intersect(sVSN_OSN.up_genes, intersect(sVSN_V1R.up_genes, sVSN_V2R.up_genes))

sVSN_OSN.down_genes <- rownames(unknown_mVSN_mOSN.DEGs[unknown_mVSN_mOSN.DEGs$p_val_adj < 0.001 &
                                                       unknown_mVSN_mOSN.DEGs$avg_log2FC < -log2(1.5), ])
sVSN_V1R.down_genes <- rownames(unknown_mVSN_V1R_mVSN.DEGs[unknown_mVSN_V1R_mVSN.DEGs$p_val_adj < 0.001 & 
                             unknown_mVSN_V1R_mVSN.DEGs$avg_log2FC < -log2(1.5), ])
sVSN_V2R.down_genes <- rownames(unknown_mVSN_V2R_mVSN.DEGs[unknown_mVSN_V2R_mVSN.DEGs$p_val_adj < 0.001 &
                             unknown_mVSN_V2R_mVSN.DEGs$avg_log2FC < -log2(1.5), ])

sVSN.down_genes <- intersect(sVSN_OSN.down_genes, intersect(sVSN_V1R.down_genes, sVSN_V2R.down_genes))

# How many universally significant regulated genes in sVSNs when compared against other mature neuron types
print("sVSN Significant Upregulated Genes")
length(sVSN.up_genes)
print("sVSN Significant Downregulated Genes")
length(sVSN.down_genes)

all.svsn.degs <- c(sVSN.up_genes,sVSN.down_genes)

```

```{r}

# Get count data for mature sensory neurons
mSN.matrix <- GetAssayData(subset(neuron.integrated,idents=c("V1R_mVSN","V2R_mVSN","mOSN","sVSN"),downsample=400),slot='data')
# Subset the matrix for our sVSN DEGs
sVSN.top.df <- as.data.frame(mSN.matrix[all.svsn.degs, ])

# The cell IDs for ALL mature sensory neurons
mature.neuron.cellIDs <- colnames(sVSN.top.df)
# Get corresponding cell-type metadata for column annotations
mature.neuron.annotation <- data.frame(cell_type=neuron.integrated@meta.data[mature.neuron.cellIDs,c("cell_type")])
rownames(mature.neuron.annotation) <- colnames(sVSN.top.df)

rm(mSN.matrix,mature.neuron.cellIDs)

```

```{r}

# Custom Palette for gene expression
colors <- c("#007FFF", "#2E8B57", "#FFFF66", "#FFA500", "#FF0000")
palette <- colorRampPalette(colors)

# Custom Palette for Cell Types
figS_4.palette <-wesanderson::wes_palette("Darjeeling1")[2:5]
names(figS_4.palette) <- c("V1R_mVSN","V2R_mVSN","mOSN","sVSN")

# Annotation Color Schematic 
vsn.ann_colors <- list(cell_type=figS_4.palette)

figS_4 <- pheatmap(mat=sVSN.top.df, 
                   color=palette(n=50), fontsize=8, treeheight_col=0, cellwidth = 0.37,
                   cluster_rows=TRUE, cluster_cols=TRUE, treeheight_row=25, cellheight = 1.3,
                   annotation_col=mature.neuron.annotation, show_colnames=FALSE,
                   show_rownames=FALSE, annotation_colors=vsn.ann_colors,
                   main="Significant Differentially Expressed Genes in sVSNs")

```

## Enriched gene ontology (GO) terms in sVSNs when compared with V1R and V2R VSNs, respectively  

```{r, include=FALSE}

# Combined GO Results for Unknown mVSNs VS V1Rs
unknown_mVSN_V_V1R.GO <- rbind(unknown_mVSN_V1R_mVSN.BP.res,unknown_mVSN_V1R_mVSN.MF.res,unknown_mVSN_V1R_mVSN.CC.res)
# Remove NAs
unknown_mVSN_V_V1R.GO <- na.omit(unknown_mVSN_V_V1R.GO)
# Just significant results
unknown_mVSN_V_V1R.GO <- unknown_mVSN_V_V1R.GO[unknown_mVSN_V_V1R.GO$padj <= 0.05, ]
# Create a column for GO category
unknown_mVSN_V_V1R.GO$Category <- substring(unknown_mVSN_V_V1R.GO$pathway,first=3,last=4)
unknown_mVSN_V_V1R.GO$pathway <- substring(unknown_mVSN_V_V1R.GO$pathway,first=6)
# Order by normalized enrichment score
unknown_mVSN_V_V1R.GO <- unknown_mVSN_V_V1R.GO[order(unknown_mVSN_V_V1R.GO$NES), ]
unknown_mVSN_V_V1R.GO$pathway <- factor(unknown_mVSN_V_V1R.GO$pathway, levels=unknown_mVSN_V_V1R.GO$pathway)

# Combind GO Results for Unknown mVSNs VS V2Rs
unknown_mVSN_V_V2R.GO <- rbind(unknown_mVSN_V2R_mVSN.BP.res,unknown_mVSN_V2R_mVSN.MF.res,unknown_mVSN_V2R_mVSN.CC.res)
# Remove NAs
unknown_mVSN_V_V2R.GO <- na.omit(unknown_mVSN_V_V2R.GO)
# Just significant results
unknown_mVSN_V_V2R.GO <- unknown_mVSN_V_V2R.GO[unknown_mVSN_V_V2R.GO$padj <= 0.05, ]
# Create a column for GO category
unknown_mVSN_V_V2R.GO$Category <- substring(unknown_mVSN_V_V2R.GO$pathway,first=3,last=4) 
unknown_mVSN_V_V2R.GO$pathway <- substring(unknown_mVSN_V_V2R.GO$pathway,first=6)
# Sort by normalized enrichment score
unknown_mVSN_V_V2R.GO <- unknown_mVSN_V_V2R.GO[order(unknown_mVSN_V_V2R.GO$NES), ]
unknown_mVSN_V_V2R.GO$pathway <- factor(unknown_mVSN_V_V2R.GO$pathway, levels=unknown_mVSN_V_V2R.GO$pathway)

```

```{r}

# Remove GO-Terms with 1000 or more associated genes
unknown_mVSN_V_V1R.GO <- unknown_mVSN_V_V1R.GO[unknown_mVSN_V_V1R.GO$size < 1000, ]
unknown_mVSN_V_V2R.GO <- unknown_mVSN_V_V2R.GO[unknown_mVSN_V_V2R.GO$size < 1000, ]

# Terms to remove 
v1r.go.remove <- c("ESTABLISHMENT_OF_PROTEIN_LOCALIZATION_TO_ORGANELLE")
v2r.go.remove <- c("RIBOSOMAL_SUBUNIT","RIBOSOME")

unknown_mVSN_V_V1R.GO <- unknown_mVSN_V_V1R.GO[unknown_mVSN_V_V1R.GO$pathway!=v1r.go.remove, ]
unknown_mVSN_V_V2R.GO <- unknown_mVSN_V_V2R.GO[unknown_mVSN_V_V2R.GO$pathway!=v2r.go.remove, ]

```

```{r}

fig2_e.part1 <- ggplot(unknown_mVSN_V_V1R.GO, aes(x=pathway,y=NES,fill=Category)) +
  geom_bar(stat="identity") +
  theme(aspect.ratio = 1,plot.title=element_text(face="bold")) +
  ggtitle("Enriched GO Terms: sVSNs vs V1R mVSNs") +
  xlab("GO Term") +
  coord_flip()

```

```{r}

fig2_e.part2 <- ggplot(unknown_mVSN_V_V2R.GO, aes(x=pathway,y=NES,fill=Category)) +
  geom_bar(stat="identity") +
  theme(aspect.ratio = 1,plot.title=element_text(face="bold")) +
  ggtitle("Enriched GO Terms in sVSNs vs V2R mVSNs") +
  xlab("GO Term") +
  coord_flip()

```

```{r include=FALSE}

fig2_e <- ggarrange(plotlist=list(V1R=fig2_e.part1,V2R=fig2_e.part2), nrow=2, ncol=1, 
                    common.legend=TRUE, align = "hv", legend = "right") +
  theme(aspect.ratio=1)

svglite(filename = paste0(adobe_images,"fig2_e.svg"),width=8,height=8)
fig2_e
invisible(dev.off())

```

```{r}

fig2_e

```

## Box plots of normalized expression for Muc2, Obp2a, Obp2b, and Lcn3 across mature sensory neurons  

```{r}

my_theme_2 <- theme(axis.title.y=element_text(size=20),             #change font size of y axis label
                    axis.title.x=element_blank(),                   #don't show x axis label
                    plot.title=element_text(size=32,hjust=0.5),     #change font size of plot title, center
                    legend.title=element_text(size=24),             #change font size of legend title
                    legend.text=element_text(size=20),              #change font size of legend text
                    axis.text.y=element_text(size=20),              #change font size of axis tick labels
                    axis.text.x=element_blank(),                    #don't show x-axis tick labels
                    axis.ticks.x=element_blank(),                   #don't show x-axis ticks
                    legend.key.size=unit(8,'mm'),                   #the size of legend 'boxes'
                    legend.title.align=0,                           #left align legend title
                    aspect.ratio=1)                                 #keep plots square

```

```{r}

p1 <- ggplot(sVSN.compare.df,aes(x=cell_type,y=Muc2,fill=cell_type)) + geom_boxplot() +
  scale_fill_manual("Cell Type",values=mature.cell_type.colors) + theme_classic() +
  stat_summary(fun.y=mean,geom='point',size=20,colour="black",shape=95,show_guide=FALSE) +
  stat_compare_means(comparisons=my_comparisons,tip.length=0.01,
                     method="wilcox.test",label="p.signif",vjust=0.3) +
  ggtitle("Muc2") + ylab("normalized count") + my_theme_2
p2 <- ggplot(sVSN.compare.df,aes(x=cell_type,y=Obp2a,fill=cell_type)) + geom_boxplot() +
  scale_fill_manual("Cell Type",values=mature.cell_type.colors) + theme_classic() +
  stat_summary(fun.y=mean,geom='point',size=20,colour="black",shape=95,show_guide=FALSE) +
  stat_compare_means(comparisons=my_comparisons,tip.length=0.01,
                     method="wilcox.test",label="p.signif",vjust=0.3) +
  ggtitle("Obp2a") + ylab("normalized count") + my_theme_2
p3 <- ggplot(sVSN.compare.df,aes(x=cell_type,y=Obp2b,fill=cell_type)) + geom_boxplot() +
  scale_fill_manual("Cell Type",values=mature.cell_type.colors) + theme_classic() +
  stat_summary(fun.y=mean,geom='point',size=20,colour="black",shape=95,show_guide=FALSE) +
  stat_compare_means(comparisons=my_comparisons,tip.length=0.01,
                     method="wilcox.test",label="p.signif",vjust=0.3) +
  ggtitle("Obp2b") + ylab("normalized count") + my_theme_2
p4 <- ggplot(sVSN.compare.df,aes(x=cell_type,y=Lcn3,fill=cell_type)) + geom_boxplot() +
  scale_fill_manual("Cell Type",values=mature.cell_type.colors) + theme_classic() +
  stat_summary(fun.y=mean,geom='point',size=20,colour="black",shape=95,show_guide=FALSE) +
  stat_compare_means(comparisons=my_comparisons,tip.length=0.01,
                     method="wilcox.test",label="p.signif",vjust=0.3) +
  ggtitle("Lcn3") + ylab("normalized count") + my_theme_2

fig2_f <- ggarrange(p1,p2,p3,p4,ncol=2,nrow=2,common.legend=TRUE,align="hv",legend="bottom")
  
```

```{r}

svglite(filename = paste0(adobe_images,"fig2_f.svg"), width=10, height=10)
fig2_f
invisible(dev.off())

```

```{r}

fig2_f

```

## Significantly regulated genes in sVSNs  

>The sVSNs are distinctive from the mature VSNs by a higher expression of P2ry14, Ap1s3, Cspp1, and lower expression of Omp, Tuba1a, and Ptma (Fig. 2f and S6).

```{r, fig.width=15, fig.height=18}

svsn.goi <- c("Muc2","Obp2a","Obp2b","Lcn3",
              "Wnk1","Neat1","P2ry14","Zfp738",
              "Cspp1","Rbm4b","Gm16579","D10Wsu102e",
              "Omp","Jund","Ndufa2","Ndufa7",
              "Cox6a1","Cox7a2","Map1lc3b","Prxl2b")

figS_3 <- FeaturePlot(neuron.integrated, features=svsn.goi, 
            cols=c("cyan3","magenta"), ncol=4, order=TRUE) &
  theme(aspect.ratio = 1,
        text=element_text(size=10),
        plot.title=element_text(size=16),
        axis.title=element_text(size=12))

svglite(filename = paste0(adobe_images,"figS_3.svg"), width=15, height=18)
figS_3
invisible(dev.off())

figS_3

```

# {-}

```{r include=FALSE}

current.variables <- c(ls(),"current.variables")
rm(list=setdiff(current.variables,c(global.vars)))
invisible(gc())

```

# Molecular cascades separating the neuronal lineages {.tabset .tabset-pills}  

## PCA plot of V1R and V2R pseudotime principal curves across cell types  

```{r}

load(file=paste0(object_dir,"v1r.v.v2r.seu"))                       # Seurat Object
load(file=paste0(object_dir,"inp.integrated.preCellID"))            # INP Seurat Object
load(file=paste0(object_dir,"v1r.v.v2r.sds"))                       # Slingshot Object
load(file=paste0(fitGAM_dir,"v1r.v.v2r.sce"))                       # Load the tradeSeq object
load(file=paste0(object_dir,"v1r.v.v2r.patternDEres"))              # patternDE test results object

```

```{r}

# Extract slingshot principal curves for plotting with Seurat/ggplot2
v1r.curve <- slingCurves(v1r.v.v2r.sds)[[2]] 
v1r.curve <- data.frame(v1r.curve$s[v1r.curve$ord, ])
v2r.curve <- slingCurves(v1r.v.v2r.sds)[[1]] 
v2r.curve <- data.frame(v2r.curve$s[v2r.curve$ord, ])

```

```{r}

v1r.v.v2r.seu@active.ident <- v1r.v.v2r.seu$cell_type
# Set the order of the levels for the factor
v1r.v.v2r.seu@active.ident <- factor(v1r.v.v2r.seu@active.ident, 
                                      levels = c("GBC", "INP", "V1R.INP", "V2R.INP",
                                                 "V1R_iVSN", "V2R_iVSN", "V1R_mVSN", "V2R_mVSN"))

```

```{r}

temp.pal.1 <- wesanderson::wes_palette("GrandBudapest1")
temp.pal.2 <- wesanderson::wes_palette("GrandBudapest2")
pal <- c(temp.pal.1,temp.pal.2)
names(pal) <- levels(v1r.v.v2r.seu@active.ident)
# Save the 2D PCA of Cell Types with V1R and V2R Principal Curves
svglite(filename = paste0(adobe_images,"fig3_a.svg"),width=6,height=6)
fig3_a <- DimPlot(v1r.v.v2r.seu, reduction="pca", pt.size=2, shuffle=TRUE, cols=pal) +
  geom_path(data=v1r.curve, aes(x=PC_1, y=PC_2), size=0.05, col="black") +
  geom_path(data=v2r.curve, aes(x=PC_1, y=PC_2), size=0.05, col="blue") +
  geom_labelpath(data=v1r.curve, aes(x=PC_1,y=PC_2), size=4.5, label="V1R", hjust=0.8,
                 fontface="bold") +
  geom_labelpath(data=v2r.curve, aes(x=PC_1,y=PC_2), size=4.5, label="V2R", hjust=0.8, 
                 colour="blue", fontface="bold") +
  theme(aspect.ratio = 1.0, text=element_text(size=22), axis.text = element_text(size=22), legend.position="bottom") +
  guides(color=guide_legend(nrow = 2))
  
rm(temp.pal.1,temp.pal.2)

fig3_a
invisible(dev.off())

```

```{r,fig.width=12,fig.height=12}

# V1R/V2R Principal Curves  
# Plot the 2D PCA of Cell Types with V1R and V2R Principal Curves
fig3_a

```

## Cell density plots across pseudotime for P14 and P56 mice  

```{r}

# Calculate pseudotimes
v1r.v.v2r.pt <- slingPseudotime(v1r.v.v2r.sds, na=FALSE)
v1r.v.v2r.seu <- AddMetaData(v1r.v.v2r.seu, v1r.v.v2r.pt[ ,2], col.name = "v1r.pseudotime")
v1r.v.v2r.seu <- AddMetaData(v1r.v.v2r.seu, v1r.v.v2r.pt[ ,1], col.name = "v2r.pseudotime")

# Add cell-weights
v1r.cw <- as.integer(v1r.v.v2r.seu$cell_type == "GBC" | 
                            v1r.v.v2r.seu$cell_type == "INP" | 
                              v1r.v.v2r.seu$cell_type == "V1R.INP" | 
                                v1r.v.v2r.seu$cell_type == "V1R_iVSN" |
                                  v1r.v.v2r.seu$cell_type == "V1R_mVSN")
v2r.cw <- as.integer(v1r.v.v2r.seu$cell_type == "GBC" | 
                            v1r.v.v2r.seu$cell_type == "INP" | 
                              v1r.v.v2r.seu$cell_type == "V2R.INP" | 
                                v1r.v.v2r.seu$cell_type == "V2R_iVSN" |
                                  v1r.v.v2r.seu$cell_type == "V2R_mVSN")
v1r.v.v2r.seu <- AddMetaData(v1r.v.v2r.seu, v1r.cw, col.name = "v1r.cw")
v1r.v.v2r.seu <- AddMetaData(v1r.v.v2r.seu, v2r.cw, col.name = "v2r.cw")

```

```{r}

v1r.v.v2r.pseudotime <- FetchData(v1r.v.v2r.seu,vars=c("cell_type","age","v1r.pseudotime",
                                                       "v2r.pseudotime","v1r.cw","v2r.cw"))

v1r.v.v2r.pseudotime$cell_type <- 
  factor(v1r.v.v2r.pseudotime$cell_type,
         levels=c("GBC","INP","V1R.INP","V2R.INP","V1R_iVSN",
                  "V2R_iVSN","V1R_mVSN","V2R_mVSN"))

```

[Summary of Kolmogorov-Smirnov Test](https://en.wikipedia.org/wiki/Kolmogorov–Smirnov_test)  

```{r}

# Subset the cells for each lineage
v1r.pseudotime <- v1r.v.v2r.pseudotime[v1r.v.v2r.pseudotime$v1r.cw==1, ]
v1r.pseudotime <- subset(v1r.pseudotime,select=c(cell_type,age,v1r.pseudotime,v1r.cw))
v1r.pseudotime$cell_type <- droplevels(v1r.pseudotime$cell_type)
v2r.pseudotime <- v1r.v.v2r.pseudotime[v1r.v.v2r.pseudotime$v2r.cw==1, ]
v2r.pseudotime <- subset(v2r.pseudotime,select=c(cell_type,age,v2r.pseudotime,v2r.cw))
v2r.pseudotime$cell_type <- droplevels(v2r.pseudotime$cell_type)

# Get age specific cells for statistical test
p14.v1r.pseudotime <- v1r.pseudotime[v1r.pseudotime$age=="P14", ]
p56.v1r.pseudotime <- v1r.pseudotime[v1r.pseudotime$age=="P56", ]
p14.v2r.pseudotime <- v2r.pseudotime[v2r.pseudotime$age=="P14", ]
p56.v2r.pseudotime <- v2r.pseudotime[v2r.pseudotime$age=="P56", ]

###############################
### Kolmogorov-Smirnov Test ###  
###############################
ks.test(p14.v1r.pseudotime$v1r.pseudotime,p56.v1r.pseudotime$v1r.pseudotime)

```

```{r, include=FALSE}

pal <- wesanderson::wes_palette("Darjeeling1",4,"discrete")

fig3_b.part_1 <- ggdensity(data=v1r.pseudotime, x="v1r.pseudotime",
                    color="age",fill="age",title="V1R") +
  theme(aspect.ratio=1, axis.title=element_text(size=20), legend.text=element_text(size=20),
        axis.text=element_text(size=20), plot.title=element_text(hjust=0.5,size=28),
        legend.title=element_text(size=20)) +
  xlab("pseudotime") + 
  annotate("text",x=100,y=0.05,label="p-value < 2.2e-16",size=20/.pt) +
  annotate(
    ymin = -0.001, ymax = -0.0003,
    xmin = c(0, 63, 150, 200),
    xmax = c(63, 150, 200, 237),
    geom = "rect",
    fill = c(pal[2],pal[3],pal[4],pal[1])
  ) +
  annotate("text",x=31,y=-0.0025,label="GBC",size=16/.pt) +
  annotate("text",x=106,y=-0.0025,label="INP",size=16/.pt) +
  annotate("text",x=175,y=-0.0025,label="iVSN",size=16/.pt) +
  annotate("text",x=218,y=-0.0025,label="mVSN",size=16/.pt)
  
```

```{r}

###############################
### Kolmogorov-Smirnov Test ###  
###############################
ks.test(p14.v2r.pseudotime$v2r.pseudotime,p56.v2r.pseudotime$v2r.pseudotime)

```

```{r, include=FALSE}

fig3_b.part_2 <- ggdensity(data=v2r.pseudotime, x="v2r.pseudotime",
                          color="age",fill="age",title="V2R") +
  theme(aspect.ratio=1, axis.title=element_text(size=20), legend.text=element_text(size=20),
        axis.text=element_text(size=20), plot.title=element_text(hjust=0.5,size=28),
        legend.title=element_text(size=20)) +
  xlab("pseudotime") + 
  annotate("text",x=140,y=0.03,label="p-value: 1.892e-13",size=20/.pt) +
  annotate(
    ymin = -0.001, ymax = -0.0003,
    xmin = c(0, 62, 156, 225),
    xmax = c(62, 156, 225, 272),
    geom = "rect",
    fill = c(pal[2],pal[3],pal[4],pal[1])
  ) +
  annotate("text",x=31,y=-0.0025,label="GBC",size=16/.pt) +
  annotate("text",x=109,y=-0.0025,label="INP",size=16/.pt) +
  annotate("text",x=190,y=-0.0025,label="iVSN",size=16/.pt) +
  annotate("text",x=249,y=-0.0025,label="mVSN",size=16/.pt)

```

```{r, include=FALSE}

fig3_b <- ggarrange(plotlist=list(V1R=fig3_b.part_1,V2R=fig3_b.part_2),
                    nrow=1,ncol=2,common.legend=TRUE,legend="right")

svglite(filename = paste0(adobe_images,"fig3_b.svg"), width=12, height=6)
fig3_b
invisible(dev.off())

```

```{r, fig.width=12,fig.height=6}

fig3_b

```

## Heatmap showing expression in pseudotime for genes that differentially expressed between the V1R and V2R lineages. Heat indicates Z-score values  

```{r}

gene.counts <- GetAssayData(object=v1r.v.v2r.seu, slot="data")
gene.tot.counts <- rowSums(gene.counts)
# Only keep genes with more than 100 total normalized counts
gene.tot.counts <- gene.tot.counts[gene.tot.counts >= 300]

sig.v1r.v.v2r.patternDEres <- v1r.v.v2r.patternDEres[names(gene.tot.counts), ]
sig.v1r.v.v2r.patternDEres <- na.omit(sig.v1r.v.v2r.patternDEres)
# Keep genes with an adjusted p-value ≤ 0.001
patternDE.genes <- rownames(sig.v1r.v.v2r.patternDEres)[sig.v1r.v.v2r.patternDEres$padj <= 1e-3]

rm(gene.counts,gene.tot.counts)

```

```{r}

colors <- c("#007FFF", "#2E8B57", "#FFFF66", "#FFA500", "#FF0000")
# Custom color palette  
palette <- colorRampPalette(colors)

### based on mean smoother
patternDE.yhatSmooth <- predictSmooth(models=v1r.v.v2r.sce,gene=patternDE.genes,nPoints=50,tidy=FALSE)
patternDE.yhatSmoothScaled <- t(scale(t(patternDE.yhatSmooth)))
patternDE.heatSmooth_V1R <- pheatmap(patternDE.yhatSmoothScaled[ ,c(51:100)], fontsize = 12,
                                     color=palette(n=50), cluster_cols=FALSE,cluster_rows=TRUE,
                                     show_rownames=FALSE,show_colnames=FALSE,treeheight_row=30,
                                     main="V1R", legend=FALSE, silent=TRUE)

patternDE.matchingHeatmap_V2R <- 
  pheatmap(patternDE.yhatSmoothScaled[patternDE.heatSmooth_V1R$tree_row$order,c(1:50)],
           color=palette(n=50), cluster_cols=FALSE, cluster_rows=FALSE,
           show_rownames=FALSE, show_colnames=FALSE, fontsize = 12,
           main="V2R", legend=TRUE, silent=TRUE)

grid.arrange(patternDE.heatSmooth_V1R[[4]], patternDE.matchingHeatmap_V2R[[4]], ncol = 2, widths=c(1,1.01))

```

## Zoomed in UMAP of cell types early in the neuronal lineage  

```{r}

Idents(inp.integrated) <- as.factor(inp.integrated$cell_type)
# Set the order of the levels for the factor
inp.integrated@active.ident <- factor(inp.integrated@active.ident, 
                                      levels = c("early_INP", "late_INP", "V1R_iVSN",
                                                 "V2R_iVSN", "iOSN", "mOSN"))


```

```{r}

fig3_d <- DimPlot(neuron.integrated,label=TRUE,label.size=8,pt.size=1.5) + 
  NoLegend() + coord_cartesian(ylim=c(-3,4),xlim=c(8.5,15))
fig3_d

```

## Feature plots for select genes expressed early in the neuronal lineage  

```{r}

FeaturePlot(neuron.integrated,features=c("Ascl1"),order=TRUE,pt.size=1,cols=c("cyan","magenta")) +
  coord_cartesian(ylim=c(-3,4),xlim=c(8.5,15))

FeaturePlot(neuron.integrated,features=c("Sox2"),order=TRUE,pt.size=1,cols=c("cyan","magenta")) +
  coord_cartesian(ylim=c(-3,4),xlim=c(8.5,15))

FeaturePlot(neuron.integrated,features=c("Neurod1"),order=TRUE,pt.size=1,cols=c("cyan","magenta")) +
  coord_cartesian(ylim=c(-3,4),xlim=c(8.5,15))

FeaturePlot(neuron.integrated,features=c("Neurog1"),order=TRUE,pt.size=1,cols=c("cyan","magenta")) +
  coord_cartesian(ylim=c(-3,4),xlim=c(8.5,15))

FeaturePlot(neuron.integrated,features=c("Sp8"),order=TRUE,pt.size=1,cols=c("cyan","magenta")) +
  coord_cartesian(ylim=c(-3,4),xlim=c(8.5,15))

```

## Violin plot of normalized expression for Neurog1 and Neurod1, split by cell type  

```{r,fig.width=15,fig.height=7.5}

# Order the cell-types 
# Order the gene factor levels
neuron.integrated$cell_type <-
  factor(neuron.integrated$cell_type, levels=c("GBC","early_INP","late_INP",
                                               "V1R_iVSN","V2R_iVSN","iOSN",
                                               "V1R_mVSN","V2R_mVSN","mOSN","sVSN"))
neuron.cell_type.colors <- 
  neuron.cell_type.colors[c("GBC","early_INP","late_INP",
                            "V1R_iVSN","V2R_iVSN","iOSN",
                            "V1R_mVSN","V2R_mVSN","mOSN","sVSN")]

figS_4A <- VlnPlot(neuron.integrated, features=c("Neurog1"), split.by="cell_type", 
                  cols = neuron.cell_type.colors) & xlab("Cell Type") &
  theme(aspect.ratio=1,axis.text=element_text(size=20),axis.title=element_text(size=24),
        legend.text=element_text(size=20),legend.title=element_text(size=24),plot.title=element_text(size=32))

figS_4B <- VlnPlot(neuron.integrated, features=c("Neurod1"), split.by="cell_type", 
                  cols = neuron.cell_type.colors) & xlab("Cell Type") &
  theme(aspect.ratio=1,axis.text=element_text(size=20),axis.title=element_text(size=24),
        legend.text=element_text(size=20),legend.title=element_text(size=24),plot.title=element_text(size=32))

figS_4AB <- ggarrange(figS_4A,figS_4B,common.legend=TRUE,legend="right",align='hv',ncol=2,nrow=1)

figS_4AB

```

## UMAP of INP, iVSN, iOSN, and mOSN cell types  

```{r}

DimPlot(inp.integrated, pt.size=1, label=TRUE, cols=neuron.cell_type.colors, label.size=8, repel=TRUE) + theme(aspect.ratio = 1)

```

## Normalized expression of candidate genes for V1R/V2R/OSN lineage determination - V1R expressing genes  

```{r}

FeaturePlot(inp.integrated,features=c("Meis2"),order=TRUE,pt.size=1,cols=c("cyan","magenta")) + theme(aspect.ratio=1)

FeaturePlot(inp.integrated,features=c("Sct"),order=TRUE,pt.size=1,cols=c("cyan","magenta")) + theme(aspect.ratio=1)

FeaturePlot(inp.integrated,features=c("Fam183b"),order=TRUE,pt.size=1,cols=c("cyan","magenta")) + theme(aspect.ratio=1)

FeaturePlot(inp.integrated,features=c("Mir100hg"),order=TRUE,pt.size=1,cols=c("cyan","magenta")) + theme(aspect.ratio=1)

FeaturePlot(inp.integrated,features=c("Acad10"),order=TRUE,pt.size=1,cols=c("cyan","magenta")) + theme(aspect.ratio=1)

```

## Normalized expression of candidate genes for V1R/V2R/OSN lineage determination  

```{r}

FeaturePlot(inp.integrated,features=c("Tfap2e"),order=TRUE,pt.size=1,cols=c("cyan","magenta")) +
  theme(aspect.ratio=1)

FeaturePlot(inp.integrated,features=c("Fezf1"),order=TRUE,pt.size=1,cols=c("cyan","magenta")) +
  theme(aspect.ratio=1)

FeaturePlot(inp.integrated,features=c("Olig2"),order=TRUE,pt.size=1,cols=c("cyan","magenta")) +
  theme(aspect.ratio=1)

FeaturePlot(inp.integrated,features=c("Tshz2"),order=TRUE,pt.size=1,cols=c("cyan","magenta")) +
  theme(aspect.ratio=1)

FeaturePlot(inp.integrated,features=c("Nfib"),order=TRUE,pt.size=1,cols=c("cyan","magenta")) +
  theme(aspect.ratio=1)

FeaturePlot(inp.integrated,features=c("Bcl11b"),order=TRUE,pt.size=1,cols=c("cyan","magenta")) +
  theme(aspect.ratio=1)

```

## Feature plots of normalized gene expression for early differentially expressed genes between V1R, V2R, and OSN lineages  

```{r}

corner_theme <- theme(panel.background = element_rect(fill = "transparent", colour = NA),
                      plot.background = element_rect(fill = "transparent", colour = NA),
                      panel.grid = element_blank(),
                      panel.border = element_blank(),
                      plot.margin = unit(c(-1, -1, 0, -1), "null"),
                      panel.margin = unit(c(-1, -1, 0, -1), "null"),
                      axis.text=element_text(size=20),
                      axis.title=element_text(size=24),
                      plot.title=element_text(size=32),
                      legend.text=element_text(size=20),
                      aspect.ratio=1)

y_theme <- theme(panel.background = element_rect(fill = "transparent", colour = NA),
                 plot.background = element_rect(fill = "transparent", colour = NA),
                 panel.grid = element_blank(),
                 panel.border = element_blank(),
                 plot.margin = unit(c(-1, -1, 0, -1), "null"),
                 panel.margin = unit(c(-1, -1, 0, -1), "null"),
                 axis.title.x=element_blank(),
                 axis.text=element_text(size=20),
                 axis.title=element_text(size=24),
                 plot.title=element_text(size=32),
                 legend.text=element_text(size=20),
                 aspect.ratio=1)

x_theme <- theme(panel.background = element_rect(fill = "transparent", colour = NA),
                 plot.background = element_rect(fill = "transparent", colour = NA),
                 panel.grid = element_blank(),
                 panel.border = element_blank(),
                 plot.margin = unit(c(-1, -1, 0, -1), "null"),
                 panel.margin = unit(c(-1, -1, 0, -1), "null"),
                 axis.title.y=element_blank(),
                 axis.text=element_text(size=20),
                 axis.title=element_text(size=24),
                 plot.title=element_text(size=32),
                 legend.text=element_text(size=20),
                 aspect.ratio=1)
  
blank_theme <- theme(panel.background = element_rect(fill = "transparent", colour = NA),
                     plot.background = element_rect(fill = "transparent", colour = NA),
                     panel.grid = element_blank(),
                     panel.border = element_blank(),
                     plot.margin = unit(c(-1, -1, 0, -1), "null"),
                     panel.margin = unit(c(-1, -1, 0, -1), "null"),
                     axis.title=element_blank(),
                     axis.text=element_text(size=20),
                     plot.title=element_text(size=32),
                     legend.text=element_text(size=20),
                     aspect.ratio=1)

```

```{r,fig.width=16,fig.height=16}

figS_4C <- FeaturePlot(inp.integrated,features=c("Krt7"),order=TRUE,pt.size=1,cols=c("cyan","magenta")) +
  y_theme

figS_4D <- FeaturePlot(inp.integrated,features=c("Krt8"),order=TRUE,pt.size=1,cols=c("cyan","magenta")) +
  blank_theme

figS_4E <- FeaturePlot(inp.integrated,features=c("Emx2"),order=TRUE,pt.size=1,cols=c("cyan","magenta")) +
  blank_theme

figS_4F <- FeaturePlot(inp.integrated,features=c("Gmnc"),order=TRUE,pt.size=1,cols=c("cyan","magenta")) +
  blank_theme

figS_4G <- FeaturePlot(inp.integrated,features=c("Cdkn1a"),order=TRUE,pt.size=1,cols=c("cyan","magenta")) +
  y_theme

figS_4H <- FeaturePlot(inp.integrated,features=c("Mcidas"),order=TRUE,pt.size=1,cols=c("cyan","magenta")) +
  blank_theme

figS_4I <- FeaturePlot(inp.integrated,features=c("Cdk1"),order=TRUE,pt.size=1,cols=c("cyan","magenta")) +
  blank_theme

figS_4J <- FeaturePlot(inp.integrated,features=c("Phlda1"),order=TRUE,pt.size=1,cols=c("cyan","magenta")) +
  blank_theme

figS_4K <- FeaturePlot(inp.integrated,features=c("Misp"),order=TRUE,pt.size=1,cols=c("cyan","magenta")) +
  y_theme

figS_4L <- FeaturePlot(inp.integrated,features=c("Clmp"),order=TRUE,pt.size=1,cols=c("cyan","magenta")) +
  blank_theme

figS_4M <- FeaturePlot(inp.integrated,features=c("Tppp3"),order=TRUE,pt.size=1,cols=c("cyan","magenta")) +
  blank_theme

figS_4N <- FeaturePlot(inp.integrated,features=c("Ina"),order=TRUE,pt.size=1,cols=c("cyan","magenta")) +
  blank_theme

figS_4O <- FeaturePlot(inp.integrated,features=c("Mss51"),order=TRUE,pt.size=1,cols=c("cyan","magenta")) +
  corner_theme

figS_4P <- FeaturePlot(inp.integrated,features=c("Dnase2a"),order=TRUE,pt.size=1,cols=c("cyan","magenta")) +
  x_theme

figS_4Q <- FeaturePlot(inp.integrated,features=c("Tex15"),order=TRUE,pt.size=1,cols=c("cyan","magenta")) +
  x_theme

figS_4R <- FeaturePlot(inp.integrated,features=c("Crabp1"),order=TRUE,pt.size=1,cols=c("cyan","magenta")) +
  x_theme

ggarrange(figS_4C,figS_4D,figS_4E,figS_4F,figS_4G,figS_4H,figS_4I,figS_4J,figS_4K,figS_4L,figS_4M,figS_4N,
          figS_4O,figS_4P,figS_4Q,figS_4R,nrow=4,ncol=4,align='hv')

```

## Feature plots of normalized expression for Notch1 and Dll4  

```{r,fig.width=16,fig.height=8}

figS_3S <- FeaturePlot(inp.integrated,features=c("Notch1"),order=TRUE,pt.size=1.5,cols=c("cyan","magenta")) +
  theme(aspect.ratio=1)
figS_3T <- FeaturePlot(inp.integrated,features=c("Dll4"),order=TRUE,pt.size=1.5,cols=c("cyan","magenta")) +
  theme(aspect.ratio=1)

ggarrange(figS_3S,figS_3T,ncol=2,align='hv')

```

# {-}

```{r include=FALSE}

current.variables <- c(ls(),"current.variables")
rm(list=setdiff(current.variables,c(global.vars)))
invisible(gc())

```

# Receptor expression in the VNO {.tabset .tabset-pills}  

## Expression level of individual receptors (total raw counts) plotted against the number of cells expressing the receptor for V1R, V2R, VNO-Olfr, and MOE-Olfr  

```{r}

load(file=paste0(moe_dir,"moe.integrated.preCellID"))
# Load data for panels E and F
vr.cor.transcription_factor <- readRDS(file=paste0(object_dir,"vr.cor.transcription_factor.rds"))
v1r.cor.transcription_factor <- readRDS(file=paste0(object_dir,"v1r.cor.transcription_factor.rds"))
v2r.cor.transcription_factor <- readRDS(file=paste0(object_dir,"v2r.cor.transcription_factor.rds"))
v1r.transcription_factor.b_in_a.mat <- readRDS(file=paste0(object_dir,"v1r.transcription_factor.b_in_a.mat.rds"))
v2r.transcription_factor.b_in_a.mat <- readRDS(file=paste0(object_dir,"v2r.transcription_factor.b_in_a.mat.rds"))
vno.receptor.identity <- readRDS(file=paste0(object_dir,"vno.receptor.identity.matrix.rds"))

```

```{r}

moe.mature.neurons <- subset(moe.seu.integrated,idents=c("0","2","3","7","8",
                                                         "16","17","18","26"))

```

```{r, include=FALSE}

receptor.names <- grep("^(Vmn1r[0-9]{1,3}$|Vmn2r[0-9]{1,3}$|Olfr[0-9]{1,4}$)",
                       rownames(mature.neurons), value=TRUE)

# Get all receptor names in the mature.neurons dataset (including pseudogenes)
psg.receptor.names <- grep("^(Vmn1r|Vmn2r|Olfr)",
                       rownames(mature.neurons), value=TRUE)

moe.olfr.names <- grep("^(Olfr)", rownames(moe.mature.neurons), value=TRUE)

# Get V1R, V2R, and Olfr names.
V1R.names <- grep("^Vmn1r",receptor.names,value=TRUE)
V2R.names <- grep("^Vmn2r",receptor.names,value=TRUE)
Olfr.names <- grep("^Olfr",receptor.names,value=TRUE)

# Retrieve the cell barcode X receptor count matrix for all chemo-sensory receptor genes
receptor.df <- FetchData(mature.neurons, vars=receptor.names, slot="counts")
psg.receptor.df <- FetchData(mature.neurons, vars=psg.receptor.names, slot="counts")
moe.receptor.df <- FetchData(moe.mature.neurons, vars=moe.olfr.names, slot="counts")

```

```{r, include=FALSE}

v1r.df <- receptor.df[ ,V1R.names]
v2r.df <- receptor.df[ ,V2R.names]
olfr.df <- receptor.df[ ,Olfr.names]

# Total_Counts is based on raw counts
v1r.total.counts <- 
  setNames(nm=c('gene','Total_Counts'),stack(sort(colSums(v1r.df),decreasing=TRUE))[2:1])
v2r.total.counts <- 
  setNames(nm=c('gene','Total_Counts'),stack(sort(colSums(v2r.df),decreasing=TRUE))[2:1])
olfr.total.counts <- 
  setNames(nm=c('gene','Total_Counts'),stack(sort(colSums(olfr.df),decreasing=TRUE))[2:1])
moe.olfr.total.counts <- 
  setNames(nm=c('gene','Total_Counts'),stack(sort(colSums(moe.receptor.df),decreasing=TRUE))[2:1])

# nCells is based on simple presence
v1r.nCells <- 
  setNames(nm=c("gene","nCells"),stack(sort(colSums(v1r.df > 0),decreasing=TRUE))[2:1])
v2r.nCells <- 
  setNames(nm=c("gene","nCells"),stack(sort(colSums(v2r.df > 0),decreasing=TRUE))[2:1])
olfr.nCells <- 
  setNames(nm=c("gene","nCells"),stack(sort(colSums(olfr.df > 0),decreasing=TRUE))[2:1])
moe.olfr.nCells <- 
  setNames(nm=c("gene","nCells"),stack(sort(colSums(moe.receptor.df > 0),decreasing=TRUE))[2:1])

v1r.metadata <- merge(v1r.nCells,v1r.total.counts)
# Remove unexpressed receptors
v1r.metadata <- v1r.metadata[!(v1r.metadata$nCells==0), ]
# Calculate average expression
v1r.metadata$avg_exp <- v1r.metadata$Total_Counts/v1r.metadata$nCells
# Set the gene factor to ordered by avg_exp decreasing
v1r.metadata$gene <- factor(v1r.metadata$gene,levels=v1r.metadata$gene[order(v1r.metadata$avg_exp,decreasing=TRUE)])
# Set the dataframe to the same order
v1r.metadata <- v1r.metadata[order(v1r.metadata$avg_exp,decreasing = TRUE), ]
# Reset rownames
rownames(v1r.metadata) <- NULL
# Get the rank of receptors by total counts
v1r.metadata$rank <- as.numeric(rownames(v1r.metadata))

v2r.metadata <- merge(v2r.nCells,v2r.total.counts)
v2r.metadata <- v2r.metadata[!(v2r.metadata$nCells==0), ]
v2r.metadata$avg_exp <- v2r.metadata$Total_Counts/v2r.metadata$nCells
# Set the gene factor to ordered by avg_exp decreasing
v2r.metadata$gene <- factor(v2r.metadata$gene,levels=v2r.metadata$gene[order(v2r.metadata$avg_exp,decreasing=TRUE)])
# Set the dataframe to the same order
v2r.metadata <- v2r.metadata[order(v2r.metadata$avg_exp,decreasing = TRUE), ]
# Reset rownames
rownames(v2r.metadata) <- NULL
# Get the rank of receptors by total counts
v2r.metadata$rank <- as.numeric(rownames(v2r.metadata))

olfr.metadata <- merge(olfr.nCells,olfr.total.counts)
olfr.metadata <- olfr.metadata[!(olfr.metadata$nCells==0), ]
olfr.metadata$avg_exp <- olfr.metadata$Total_Counts/olfr.metadata$nCells
# Set the gene factor to ordered by avg_exp decreasing
olfr.metadata$gene <- factor(olfr.metadata$gene,levels=olfr.metadata$gene[order(olfr.metadata$avg_exp,decreasing=TRUE)])
# Set the dataframe to the same order
olfr.metadata <- olfr.metadata[order(olfr.metadata$avg_exp,decreasing = TRUE), ]
# Reset rownames
rownames(olfr.metadata) <- NULL
# Get the rank of receptors by total counts
olfr.metadata$rank <- as.numeric(rownames(olfr.metadata))

moe.olfr.metadata <- merge(moe.olfr.nCells,moe.olfr.total.counts)
# Remove Genes Unexpressed in Cells
moe.olfr.metadata <- moe.olfr.metadata[!(moe.olfr.metadata$nCells==0), ]
moe.olfr.metadata$avg_exp <- moe.olfr.metadata$Total_Counts/moe.olfr.metadata$nCells
# Set the gene factor to ordered by avg_exp decreasing
moe.olfr.metadata$gene <- factor(moe.olfr.metadata$gene,levels=moe.olfr.metadata$gene[order(moe.olfr.metadata$avg_exp,decreasing=TRUE)])
# Set the dataframe to the same order
moe.olfr.metadata <- moe.olfr.metadata[order(moe.olfr.metadata$avg_exp,decreasing = TRUE), ]
# Reset rownames
rownames(moe.olfr.metadata) <- NULL
# Get the rank of receptors by total counts
moe.olfr.metadata$rank <- as.numeric(rownames(moe.olfr.metadata))

v1r.metadata$proportion <- v1r.metadata$avg_exp/sum(v1r.metadata$avg_exp) 
v2r.metadata$proportion <- v2r.metadata$avg_exp/sum(v2r.metadata$avg_exp) 
olfr.metadata$proportion <- olfr.metadata$avg_exp/sum(olfr.metadata$avg_exp) 
moe.olfr.metadata$proportion <- moe.olfr.metadata$avg_exp/sum(moe.olfr.metadata$avg_exp) 

```

```{r, include=FALSE}

vno.receptor.df.list <- list(V1R=v1r.total.counts,V2R=v2r.total.counts,
                             Olfr=olfr.total.counts,MOE.Olfr=moe.olfr.metadata)

vno.receptor.pie.list <- lapply(names(vno.receptor.df.list), function(name){
  df <- vno.receptor.df.list[[name]]
  zeroes <- sum(df[ ,"Total_Counts"] == 0)
  low <- sum(df[ ,"Total_Counts"] > 0 & df[ ,"Total_Counts"] < 100)
  medium <- sum(df[ ,"Total_Counts"] >= 100 & df[ ,"Total_Counts"] < 1000)
  high <- sum(df[ ,"Total_Counts"] >= 1000 & df[ ,"Total_Counts"] < 10000)
  very_high <- sum(df[ ,"Total_Counts"] >= 10000)
  c(zeroes,low,medium,high,very_high)
})

names(vno.receptor.pie.list) <- names(vno.receptor.df.list)

lbls <- c("x = 0","0 < x < 100","100 ≤ x < 1000",
          "1000 ≤ x < 10000","x ≥ 10000")

# Order the Receptors factor levels for plotting
lbls <- factor(as.factor(lbls), levels=c("x = 0","0 < x < 100","100 ≤ x < 1000",
                                         "1000 ≤ x < 10000","x ≥ 10000"))

v1r.pie.df <- data.frame(Expression_Range=lbls,Counts=vno.receptor.pie.list$V1R)
v2r.pie.df <- data.frame(Expression_Range=lbls,Counts=vno.receptor.pie.list$V2R)
olfr.pie.df <- data.frame(Expression_Range=lbls,Counts=vno.receptor.pie.list$Olfr)
moe.olfr.pie.df <- data.frame(Expression_Range=lbls,Counts=vno.receptor.pie.list$MOE.Olfr)

```

```{r}

# Create a color vector to distinguish genes from pseudogenes.
v1r.dist.plot <- v1r.metadata %>% mutate(color = ifelse(str_detect(v1r.metadata$gene, "-ps"), "gray", "gold2"))

fig4_a.part_2 <- ggplot(v1r.dist.plot, aes(x=nCells,y=Total_Counts,label=gene)) +
  geom_point(shape=21,size=4,fill=v1r.dist.plot$color,color="gold2") + theme_classic() + scale_color_identity() +
  geom_text_repel(data=v1r.dist.plot[v1r.dist.plot$nCells>5000|v1r.dist.plot$Total_Counts>80000, ],
                  nudge_x=125,nudge_y=-25,point.padding=0,size=7,min.segment.length=0.25) +
  theme(axis.title = element_text(size=24),
        axis.text = element_text(size=20),aspect.ratio=1) +
  ylab("Total Raw Count") +
  stat_poly_line(method="lm", formula = y ~ x + 0) +
  stat_poly_eq(formula = y ~ x + 0, aes(label = paste(after_stat(eq.label),
                                                      after_stat(rr.label), 
                                                      sep = "*\", \"*")),size=7,label.y=0.85,label.x=0.5)

svglite(filename = paste0(adobe_images,"fig4_a.part_2.svg"),width=8,height=8)
fig4_a.part_2
invisible(dev.off())

fig4_a.part_2

```

```{r}

# Create a color vector to distinguish genes from pseudogenes.
v2r.dist.plot <- v2r.metadata %>% mutate(color = ifelse(str_detect(v2r.metadata$gene, "-ps"), "gray", "purple"))

fig4_b.part_2 <- ggplot(v2r.dist.plot, aes(x=nCells,y=Total_Counts,label=gene)) +
  geom_point(shape=21,size=4,fill=v2r.dist.plot$color,color="purple") + theme_classic() + scale_color_identity() +
  geom_text_repel(data=v2r.dist.plot[v2r.dist.plot$nCells>6000|v2r.dist.plot$Total_Counts>30000, ],
                  nudge_x=-50,nudge_y=50,point.padding=0,size=7,min.segment.length=0.25) +
  theme(axis.title = element_text(size=24),
        axis.text = element_text(size=20),aspect.ratio=1) +
  ylab("Total Raw Count") +
  stat_poly_line(method="lm", formula = y ~ x + 0) +
  stat_poly_eq(formula = y ~ x + 0, aes(label = paste(after_stat(eq.label),
                                                      after_stat(rr.label), 
                                                      sep = "*\", \"*")),size=7,label.y=0.9)

fig4_b.part_2

```

```{r}

# Create a color vector to distinguish genes from pseudogenes.
olfr.dist.plot <- olfr.metadata %>% mutate(color = ifelse(str_detect(olfr.metadata$gene, "-ps"), "gray", "green3"))

fig4_c.part_2 <- ggplot(olfr.dist.plot, aes(nCells,Total_Counts,label=gene)) +
  geom_point(shape=21,size=4,fill=olfr.dist.plot$color,color="green3") + theme_classic() + scale_color_identity() +
  geom_text_repel(data=olfr.dist.plot[olfr.dist.plot$nCells>6000|olfr.dist.plot$Total_Counts>10000, ],
                  nudge_x=-50,nudge_y=-75,point.padding=0,size=7,min.segment.length=0.25) +
  theme(axis.title = element_text(size=24),
        axis.text = element_text(size=20),aspect.ratio=1) +
  ylab("Total Raw Count") +
  stat_poly_line(method="lm", formula = y ~ x + 0) +
  stat_poly_eq(formula = y ~ x + 0, aes(label = paste(after_stat(eq.label),
                                                      after_stat(rr.label), 
                                                      sep = "*\", \"*")),size=7,label.y=0.9,label.x=0.5)

fig4_c.part_2

```

Main Olfactory Epithelium - Olfrs  

```{r}

fig4_d.part_2 <- ggplot(moe.olfr.metadata, aes(nCells,Total_Counts,label=gene)) +
  geom_point(shape=21,size=3,fill="magenta",color="magenta") + theme_classic() +
  geom_text_repel(data=moe.olfr.metadata[moe.olfr.metadata$nCells>1000|moe.olfr.metadata$Total_Counts>3000, ],
                  nudge_x=-25,nudge_y=-75,point.padding=0,size=7) +
  theme(text=element_text(size=20),
        axis.title=element_text(size=24),
        axis.text=element_text(size=20),
        aspect.ratio=1) +
  ylab("Total Raw Count") +
  stat_poly_line(method="lm", formula = y ~ x + 0) +
  stat_poly_eq(formula = y ~ x + 0, aes(label = paste(after_stat(eq.label),
                                                      after_stat(rr.label), 
                                                      sep = "*\", \"*")),size=7,label.y=0.9,label.x=0.5)

fig4_d.part_2

```

## Ranked distribution of average expression per cell for the four receptor classes. Pie charts show the number of cells expressing a receptor at the specified range  

```{r, include=FALSE}

v1r.pie.df2 <- v1r.pie.df %>% 
  mutate(csum = rev(cumsum(rev(Counts))), 
         pos = Counts/2 + lead(csum, 1),
         pos = if_else(is.na(pos), Counts/2, pos)) # isn't working
v1r.pie.df2[5,4] <- v1r.pie.df$Counts[5]/2

fig4_a.part_1 <- ggplot(v1r.pie.df2, aes(x="", y=Counts, fill=Expression_Range)) +
    geom_col(color="black") + theme_void() +
    coord_polar("y", start=0, direction=-1) + ggtitle("V1R") +
    theme(text=element_text(size=24),plot.title=element_text(hjust=0.5)) +
    geom_label_repel(aes(label=Counts),position=position_stack(vjust = 0.5),
                     size=6,segment.color = "black",segment.size = 0.5,segment.curvature = 0.2) +
  guides(fill = guide_legend(override.aes = list(label = "")))

```

```{r, include=FALSE}

# Create the gamma model with a 15 degree polynomial.
v1r.gamma.model <- glm(avg_exp ~ poly(rank, 15, raw=TRUE), data=v1r.dist.plot,
                       control=glm.control(maxit=1000), family=Gamma(link="log"))
# Create the vector of predicted values, based on our model.
v1r.dist.plot$predicted <- predict(v1r.gamma.model, type="response")

# Plot the data and the fitted model
fig4_a.part_3 <- ggplot(v1r.dist.plot, aes(x = rank, y = avg_exp, color=color)) +
  geom_point(size=4) + theme_classic() + scale_color_identity() +
  geom_line(aes(y = predicted), color = "black") +
  theme(axis.title=element_text(size=24),
        axis.text=element_text(size=20),
        aspect.ratio=1) +
  geom_label_repel(aes(label = gene, color="black"), size=8, label.size = 0, point.padding=0.5,
                   data = ~ subset(., gene == v1r.metadata$gene[1:6]), label.padding=0.1, 
                   nudge_x=70,nudge_y=-10,min.segment.length=0.25, box.padding=0.1) +
  labs(x="rank", y="avg_exp")

```

```{r,fig.width=16,fig.height=8}

ggarrange(fig4_a.part_3,fig4_a.part_1,ncol=2,nrow=1)

```

```{r, include=FALSE}

v2r.pie.df2 <- v2r.pie.df %>% 
  mutate(csum = rev(cumsum(rev(Counts))), 
         pos = Counts/2 + lead(csum, 1),
         pos = if_else(is.na(pos), Counts/2, pos)) # isn't working
v2r.pie.df2[5,4] <- v2r.pie.df$Counts[5]/2

fig4_b.part_1 <- ggplot(v2r.pie.df2, aes(x="", y=Counts, fill=Expression_Range)) +
                    geom_col(color="black") + theme_void() + 
                    coord_polar("y", start=0, direction=-1) + ggtitle("V2R") +
                    theme(text=element_text(size=24),plot.title=element_text(hjust=0.5)) +
                    geom_label_repel(aes(label=Counts),position=position_stack(vjust = 0.5),
                                     size=6,segment.color = "black",segment.size = 0.5,segment.curvature = 0.2) +
  guides(fill = guide_legend(override.aes = list(label = "")))


```

```{r}

v2r.gamma.model <- glm(avg_exp ~ poly(rank, 23, raw=TRUE), data=v2r.metadata,
                       control=glm.control(maxit=1000), family=Gamma(link="log"))
# Create the vector of predicted values, based on our model.
v2r.dist.plot$predicted <- predict(v2r.gamma.model, type="response")

# Plot the data and the fitted model
fig4_b.part_3 <- ggplot(v2r.dist.plot, aes(x = rank, y = avg_exp, color=color)) +
  geom_point(size=4) + theme_classic() + scale_color_identity() +
  geom_line(aes(y = predicted), color = "black") +
  theme(axis.title=element_text(size=24),
        axis.text=element_text(size=20),
        aspect.ratio=1) +
  geom_label_repel(aes(label = gene,color="black"), size=8, label.size = 0, point.padding=0.5,
                   data = ~ subset(., gene == v2r.metadata$gene[1:6]), label.padding=0.1, 
                   nudge_x=30, min.segment.length=0.25, box.padding=0.1) +
  labs(x = "rank", y = "avg_exp")

```

```{r,fig.width=16,fig.height=8}

ggarrange(fig4_b.part_3,fig4_b.part_1,ncol=2,nrow=1)

```

```{r}

olfr.pie.df2 <- olfr.pie.df %>% 
  mutate(csum = rev(cumsum(rev(Counts))), 
         pos = Counts/2 + lead(csum, 1),
         pos = if_else(is.na(pos), Counts/2, pos)) # isn't working
olfr.pie.df2[5,4] <- olfr.pie.df$Counts[5]/2

fig4_c.part_1 <- ggplot(olfr.pie.df2, aes(x="", y=Counts, fill=Expression_Range)) +
                    geom_col(color="black") + theme_void() + 
                    coord_polar("y", start=0, direction=-1) + ggtitle("VNO-Olfr") +
                    theme(text=element_text(size=24),plot.title=element_text(hjust=0.5)) +
                    geom_label_repel(aes(label=Counts),position=position_stack(vjust = 0.5),
                                     size=6,segment.color = "black",segment.size = 0.5,segment.curvature = 0.2) +
  guides(fill = guide_legend(override.aes = list(label = "")))
                    
```

```{r}

olfr.gamma.model <- glm(avg_exp ~ poly(rank, 15, raw=TRUE), data=olfr.metadata,
                       control=glm.control(maxit=1000), family=Gamma(link="log"))
olfr.dist.plot$predicted <- predict(olfr.gamma.model, type="response")

# Plot the data and the fitted model
fig4_c.part_3 <- ggplot(olfr.dist.plot, aes(x = rank, y = avg_exp, color=color)) +
  geom_point(size=4) + theme_classic() + scale_color_identity() +
  geom_line(aes(y = predicted), color = "black") +
  theme(axis.title=element_text(size=24),
        axis.text=element_text(size=20),
        aspect.ratio=1) +
  geom_label_repel(aes(label=gene,color="black"), size=8, label.size=0, point.padding=0.5, max.overlaps=1,
                   data = ~ subset(., gene==olfr.metadata$gene[1:7], label.padding=0.1), 
                   nudge_x=150, nudge_y=-1, min.segment.length=0.25, box.padding=0.1, xlim=c(75,600)) +
  labs(x = "rank", y = "avg_exp")

```

```{r,fig.width=16,fig.height=8}

ggarrange(fig4_c.part_3,fig4_c.part_1,ncol=2,nrow=1)

```

```{r}

moe.olfr.pie.df2 <- moe.olfr.pie.df %>% 
  mutate(csum = rev(cumsum(rev(Counts))), 
         pos = Counts/2 + lead(csum, 1),
         pos = if_else(is.na(pos), Counts/2, pos)) # isn't working
moe.olfr.pie.df2[5,4] <- moe.olfr.pie.df$Counts[5]/2

fig4_d.part_1 <- ggplot(moe.olfr.pie.df2, aes(x="", y=Counts, fill=Expression_Range)) +
                    geom_col(color="black") + theme_void() + 
                    coord_polar("y", start=0, direction=-1) + ggtitle("MOE-Olfr") +
                    theme(text=element_text(size=24),plot.title=element_text(hjust=0.5)) +
                    geom_label_repel(aes(label=Counts),position=position_stack(vjust = 0.5),
                                     size=6,segment.color = "black",segment.size = 0.5,segment.curvature = 0.2) +
  guides(fill = guide_legend(override.aes = list(label = "")))
                    

```

```{r}

moe.olfr.gamma.model <- glm(avg_exp ~ poly(rank, 22, raw=TRUE), data=moe.olfr.metadata,
                       control=glm.control(maxit=1000), family=Gamma(link="log"))
moe.olfr.metadata$predicted <- predict(moe.olfr.gamma.model, type="response")

# Plot the data and the fitted model
fig4_d.part_3 <- ggplot(moe.olfr.metadata, aes(x = rank, y = avg_exp)) +
  geom_point(color="magenta",size=5) + theme_classic() +
  geom_line(aes(y = predicted), color = "black") +
  theme(axis.title=element_text(size=40),
        axis.text=element_text(size=33),
        aspect.ratio=1) +
  geom_label_repel(aes(label = gene), size=12, label.size = NA, point.padding=0.5,
                   data = ~ subset(., gene == moe.olfr.metadata$gene[1:6]), label.padding=0.1, 
                   nudge_x=120, nudge_y=-2, min.segment.length=0, box.padding=0.1, xlim=c(100,900)) +
  labs(x = "rank", y = "avg_exp")

```

```{r,fig.width=16,fig.height=8}

ggarrange(fig4_d.part_3,fig4_d.part_1,ncol=2,nrow=1)

```

## Receptor distributions with **pseudogenes included**  

```{r, include=FALSE}

# Get V1R, V2R, and Olfr names.
V1R.names <- grep("^Vmn1r",psg.receptor.names,value=TRUE)
V2R.names <- grep("^Vmn2r",psg.receptor.names,value=TRUE)
Olfr.names <- grep("^Olfr",psg.receptor.names,value=TRUE)

v1r.df <- psg.receptor.df[ ,V1R.names]
v2r.df <- psg.receptor.df[ ,V2R.names]
olfr.df <- psg.receptor.df[ ,Olfr.names]

# Total_Counts is based on raw counts
v1r.total.counts <- 
  setNames(nm=c('gene','Total_Counts'),stack(sort(colSums(v1r.df),decreasing=TRUE))[2:1])
v2r.total.counts <- 
  setNames(nm=c('gene','Total_Counts'),stack(sort(colSums(v2r.df),decreasing=TRUE))[2:1])
olfr.total.counts <- 
  setNames(nm=c('gene','Total_Counts'),stack(sort(colSums(olfr.df),decreasing=TRUE))[2:1])
moe.olfr.total.counts <- 
  setNames(nm=c('gene','Total_Counts'),stack(sort(colSums(moe.receptor.df),decreasing=TRUE))[2:1])

# nCells is based on simple presence
v1r.nCells <- 
  setNames(nm=c("gene","nCells"),stack(sort(colSums(v1r.df > 0),decreasing=TRUE))[2:1])
v2r.nCells <- 
  setNames(nm=c("gene","nCells"),stack(sort(colSums(v2r.df > 0),decreasing=TRUE))[2:1])
olfr.nCells <- 
  setNames(nm=c("gene","nCells"),stack(sort(colSums(olfr.df > 0),decreasing=TRUE))[2:1])
moe.olfr.nCells <- 
  setNames(nm=c("gene","nCells"),stack(sort(colSums(moe.receptor.df > 0),decreasing=TRUE))[2:1])

v1r.metadata <- merge(v1r.nCells,v1r.total.counts)
# Remove unexpressed receptors
v1r.metadata <- v1r.metadata[!(v1r.metadata$nCells==0), ]
# Calculate average expression
v1r.metadata$avg_exp <- v1r.metadata$Total_Counts/v1r.metadata$nCells
# Set the gene factor to ordered by avg_exp decreasing
v1r.metadata$gene <- factor(v1r.metadata$gene,levels=v1r.metadata$gene[order(v1r.metadata$avg_exp,decreasing=TRUE)])
# Set the dataframe to the same order
v1r.metadata <- v1r.metadata[order(v1r.metadata$avg_exp,decreasing = TRUE), ]
# Reset rownames
rownames(v1r.metadata) <- NULL
# Get the rank of receptors by total counts
v1r.metadata$rank <- as.numeric(rownames(v1r.metadata))

v2r.metadata <- merge(v2r.nCells,v2r.total.counts)
v2r.metadata <- v2r.metadata[!(v2r.metadata$nCells==0), ]
v2r.metadata$avg_exp <- v2r.metadata$Total_Counts/v2r.metadata$nCells
# Set the gene factor to ordered by avg_exp decreasing
v2r.metadata$gene <- factor(v2r.metadata$gene,levels=v2r.metadata$gene[order(v2r.metadata$avg_exp,decreasing=TRUE)])
# Set the dataframe to the same order
v2r.metadata <- v2r.metadata[order(v2r.metadata$avg_exp,decreasing = TRUE), ]
# Reset rownames
rownames(v2r.metadata) <- NULL
# Get the rank of receptors by total counts
v2r.metadata$rank <- as.numeric(rownames(v2r.metadata))

olfr.metadata <- merge(olfr.nCells,olfr.total.counts)
olfr.metadata <- olfr.metadata[!(olfr.metadata$nCells==0), ]
olfr.metadata$avg_exp <- olfr.metadata$Total_Counts/olfr.metadata$nCells
# Set the gene factor to ordered by avg_exp decreasing
olfr.metadata$gene <- factor(olfr.metadata$gene,levels=olfr.metadata$gene[order(olfr.metadata$avg_exp,decreasing=TRUE)])
# Set the dataframe to the same order
olfr.metadata <- olfr.metadata[order(olfr.metadata$avg_exp,decreasing = TRUE), ]
# Reset rownames
rownames(olfr.metadata) <- NULL
# Get the rank of receptors by total counts
olfr.metadata$rank <- as.numeric(rownames(olfr.metadata))

moe.olfr.metadata <- merge(moe.olfr.nCells,moe.olfr.total.counts)
# Remove Genes Unexpressed in Cells
moe.olfr.metadata <- moe.olfr.metadata[!(moe.olfr.metadata$nCells==0), ]
moe.olfr.metadata$avg_exp <- moe.olfr.metadata$Total_Counts/moe.olfr.metadata$nCells
# Set the gene factor to ordered by avg_exp decreasing
moe.olfr.metadata$gene <- factor(moe.olfr.metadata$gene,levels=moe.olfr.metadata$gene[order(moe.olfr.metadata$avg_exp,decreasing=TRUE)])
# Set the dataframe to the same order
moe.olfr.metadata <- moe.olfr.metadata[order(moe.olfr.metadata$avg_exp,decreasing = TRUE), ]
# Reset rownames
rownames(moe.olfr.metadata) <- NULL
# Get the rank of receptors by total counts
moe.olfr.metadata$rank <- as.numeric(rownames(moe.olfr.metadata))

v1r.metadata$proportion <- v1r.metadata$avg_exp/sum(v1r.metadata$avg_exp) 
v2r.metadata$proportion <- v2r.metadata$avg_exp/sum(v2r.metadata$avg_exp) 
olfr.metadata$proportion <- olfr.metadata$avg_exp/sum(olfr.metadata$avg_exp) 
moe.olfr.metadata$proportion <- moe.olfr.metadata$avg_exp/sum(moe.olfr.metadata$avg_exp) 

```

```{r, include=FALSE}

vno.receptor.df.list <- list(V1R=v1r.total.counts,V2R=v2r.total.counts,
                             Olfr=olfr.total.counts,MOE.Olfr=moe.olfr.metadata)

vno.receptor.pie.list <- lapply(names(vno.receptor.df.list), function(name){
  df <- vno.receptor.df.list[[name]]
  zeroes <- sum(df[ ,"Total_Counts"] == 0)
  low <- sum(df[ ,"Total_Counts"] > 0 & df[ ,"Total_Counts"] < 100)
  medium <- sum(df[ ,"Total_Counts"] >= 100 & df[ ,"Total_Counts"] < 1000)
  high <- sum(df[ ,"Total_Counts"] >= 1000 & df[ ,"Total_Counts"] < 10000)
  very_high <- sum(df[ ,"Total_Counts"] >= 10000)
  c(zeroes,low,medium,high,very_high)
})

names(vno.receptor.pie.list) <- names(vno.receptor.df.list)

lbls <- c("x = 0","0 < x < 100","100 ≤ x < 1000",
          "1000 ≤ x < 10000","x ≥ 10000")

# Order the Receptors factor levels for plotting
lbls <- factor(as.factor(lbls), levels=c("x = 0","0 < x < 100","100 ≤ x < 1000",
                                         "1000 ≤ x < 10000","x ≥ 10000"))

v1r.pie.df <- data.frame(Expression_Range=lbls,Counts=vno.receptor.pie.list$V1R)
v2r.pie.df <- data.frame(Expression_Range=lbls,Counts=vno.receptor.pie.list$V2R)
olfr.pie.df <- data.frame(Expression_Range=lbls,Counts=vno.receptor.pie.list$Olfr)
moe.olfr.pie.df <- data.frame(Expression_Range=lbls,Counts=vno.receptor.pie.list$MOE.Olfr)

```

```{r}

# Create a color vector to distinguish genes from pseudogenes.
v1r.dist.plot <- v1r.metadata %>% mutate(color = ifelse(str_detect(v1r.metadata$gene, "-ps"), "gray", "gold2"))

fig4_a.part_2 <- ggplot(v1r.dist.plot, aes(x=nCells,y=Total_Counts,label=gene)) +
  geom_point(shape=21,size=4,fill=v1r.dist.plot$color,color="gold2") + theme_classic() + scale_color_identity() +
  geom_text_repel(data=v1r.dist.plot[v1r.dist.plot$nCells>5000|v1r.dist.plot$Total_Counts>80000, ],
                  nudge_x=125,nudge_y=-25,point.padding=0,size=7,min.segment.length=0.25) +
  theme(axis.title = element_text(size=24),
        axis.text = element_text(size=20),aspect.ratio=1) +
  ylab("Total Raw Count") +
  stat_poly_line(method="lm", formula = y ~ x + 0) +
  stat_poly_eq(formula = y ~ x + 0, aes(label = paste(after_stat(eq.label),
                                                      after_stat(rr.label), 
                                                      sep = "*\", \"*")),size=7,label.y=0.85,label.x=0.5)

fig4_a.part_2

```

```{r}

# Create a color vector to distinguish genes from pseudogenes.
v2r.dist.plot <- v2r.metadata %>% mutate(color = ifelse(str_detect(v2r.metadata$gene, "-ps"), "gray", "purple"))

fig4_b.part_2 <- ggplot(v2r.dist.plot, aes(x=nCells,y=Total_Counts,label=gene)) +
  geom_point(shape=21,size=4,fill=v2r.dist.plot$color,color="purple") + theme_classic() + scale_color_identity() +
  geom_text_repel(data=v2r.dist.plot[v2r.dist.plot$nCells>6000|v2r.dist.plot$Total_Counts>30000, ],
                  nudge_x=-50,nudge_y=50,point.padding=0,size=7,min.segment.length=0.25) +
  theme(axis.title = element_text(size=24),
        axis.text = element_text(size=20),aspect.ratio=1) +
  ylab("Total Raw Count") +
  stat_poly_line(method="lm", formula = y ~ x + 0) +
  stat_poly_eq(formula = y ~ x + 0, aes(label = paste(after_stat(eq.label),
                                                      after_stat(rr.label), 
                                                      sep = "*\", \"*")),size=7,label.y=0.9)

fig4_b.part_2

```

```{r}

# Create a color vector to distinguish genes from pseudogenes.
olfr.dist.plot <- olfr.metadata %>% mutate(color = ifelse(str_detect(olfr.metadata$gene, "-ps"), "gray", "green3"))

fig4_c.part_2 <- ggplot(olfr.dist.plot, aes(nCells,Total_Counts,label=gene)) +
  geom_point(shape=21,size=4,fill=olfr.dist.plot$color,color="green3") + theme_classic() + scale_color_identity() +
  geom_text_repel(data=olfr.dist.plot[olfr.dist.plot$nCells>6000|olfr.dist.plot$Total_Counts>10000, ],
                  nudge_x=-50,nudge_y=-75,point.padding=0,size=7,min.segment.length=0.25) +
  theme(axis.title = element_text(size=24),
        axis.text = element_text(size=20),aspect.ratio=1) +
  ylab("Total Raw Count") +
  stat_poly_line(method="lm", formula = y ~ x + 0) +
  stat_poly_eq(formula = y ~ x + 0, aes(label = paste(after_stat(eq.label),
                                                      after_stat(rr.label), 
                                                      sep = "*\", \"*")),size=7,label.y=0.9,label.x=0.5)

fig4_c.part_2

```

## Rank by average count distributions for V1R, V2R, and VSN-OR, **pseudogenes included**

```{r}

# Create the gamma model with a 15 degree polynomial.
v1r.gamma.model <- glm(avg_exp ~ poly(rank, 15, raw=TRUE), data=v1r.dist.plot,
                       control=glm.control(maxit=1000), family=Gamma(link="log"))
# Create the vector of predicted values, based on our model.
v1r.dist.plot$predicted <- predict(v1r.gamma.model, type="response")

# Plot the data and the fitted model
fig4_a.part_3 <- ggplot(v1r.dist.plot, aes(x = rank, y = avg_exp, color=color)) +
  geom_point(size=4) + theme_classic() + scale_color_identity() +
  geom_line(aes(y = predicted), color = "black") +
  theme(axis.title=element_text(size=24),
        axis.text=element_text(size=20),
        aspect.ratio=1) +
  geom_label_repel(aes(label = gene, color="black"), size=8, label.size = 0, point.padding=0.5,
                   data = ~ subset(., gene == v1r.metadata$gene[1:6]), label.padding=0.1, 
                   nudge_x=70,nudge_y=-10,min.segment.length=0.25, box.padding=0.1) +
  labs(x="rank", y="avg_exp")

fig4_a.part_3

```

```{r}

v2r.gamma.model <- glm(avg_exp ~ poly(rank, 23, raw=TRUE), data=v2r.metadata,
                       control=glm.control(maxit=1000), family=Gamma(link="log"))
# Create the vector of predicted values, based on our model.
v2r.dist.plot$predicted <- predict(v2r.gamma.model, type="response")

# Plot the data and the fitted model
fig4_b.part_3 <- ggplot(v2r.dist.plot, aes(x = rank, y = avg_exp, color=color)) +
  geom_point(size=4) + theme_classic() + scale_color_identity() +
  geom_line(aes(y = predicted), color = "black") +
  theme(axis.title=element_text(size=24),
        axis.text=element_text(size=20),
        aspect.ratio=1) +
  geom_label_repel(aes(label = gene,color="black"), size=8, label.size = 0, point.padding=0.5,
                   data = ~ subset(., gene == v2r.metadata$gene[1:6]), label.padding=0.1, 
                   nudge_x=30, min.segment.length=0.25, box.padding=0.1) +
  labs(x = "rank", y = "avg_exp")

fig4_b.part_3

```

```{r}

olfr.gamma.model <- glm(avg_exp ~ poly(rank, 15, raw=TRUE), data=olfr.metadata,
                       control=glm.control(maxit=1000), family=Gamma(link="log"))
olfr.dist.plot$predicted <- predict(olfr.gamma.model, type="response")

# Plot the data and the fitted model
fig4_c.part_3 <- ggplot(olfr.dist.plot, aes(x = rank, y = avg_exp, color=color)) +
  geom_point(size=4) + theme_classic() + scale_color_identity() +
  geom_line(aes(y = predicted), color = "black") +
  theme(axis.title=element_text(size=24),
        axis.text=element_text(size=20),
        aspect.ratio=1) +
  geom_label_repel(aes(label=gene,color="black"), size=8, label.size=0, point.padding=0.5, max.overlaps=1,
                   data = ~ subset(., gene==olfr.metadata$gene[1:7], label.padding=0.1), 
                   nudge_x=150, nudge_y=-1, min.segment.length=0.25, box.padding=0.1, xlim=c(75,600)) +
  labs(x = "rank", y = "avg_exp")

fig4_c.part_3

```

## Heatmaps showing the correlation of transcription factor expression among the V1Rs or V2Rs

```{r,fig.width=14,fig.height=14}

fig4_i <-
  pheatmap(v1r.cor.transcription_factor,fontsize=10,treeheight_row=25,treeheight_col=25)

# Save image to svg file
grid.draw(rectGrob(gp=gpar(fill="white", lwd=0)))
grid.draw(fig4_i[[4]])

```

```{r,fig.width=14,fig.height=14}

fig4_j <-
  pheatmap(v2r.cor.transcription_factor,fontsize=10,treeheight_row=25,treeheight_col=25)

# Save image to svg file
svglite(filename = paste0(adobe_images,"fig4_j.svg"),height=14,width=14)
grid.draw(rectGrob(gp=gpar(fill="white", lwd=0)))
grid.draw(fig4_j[[4]])
invisible(dev.off())

```

## Distribution density plot showing relationship between similarity of transcription factor (TF) gene expression profiles and receptor sequence similarity  

```{r}

# Find the index of each significant receptor co-expressed with a transcription factor gene within the receptor sequence identity matrix
match.idx <- match(x=rownames(vr.cor.transcription_factor),table=rownames(vno.receptor.identity))
# Remove NAs for genes not present in the sequence identity matrix
match.idx <- na.omit(match.idx)
# Subset our sequence identity matrix to match our vr.cor.transcription_factor matrix
vno.receptor.identity <- vno.receptor.identity[match.idx,match.idx]
# Missing receptors
missing.receptors <- setdiff(rownames(vr.cor.transcription_factor), rownames(vno.receptor.identity))
# Remove missing receptors
vr.cor.transcription_factor.filtered <- 
  vr.cor.transcription_factor[!(rownames(vr.cor.transcription_factor) %in% missing.receptors),!(colnames(vr.cor.transcription_factor) %in% missing.receptors)]

```

```{r}

receptor.seq_sim.cor.tf_sim.df <- expand.grid(rownames(vr.cor.transcription_factor.filtered), 
                                              colnames(vr.cor.transcription_factor.filtered), 
                                              stringsAsFactors=FALSE)

vr.transcription_factor.cor.vec <- sapply(1:nrow(receptor.seq_sim.cor.tf_sim.df), function(i){
  row.gene <- receptor.seq_sim.cor.tf_sim.df[i,1]
  col.gene <- receptor.seq_sim.cor.tf_sim.df[i,2]
  val <- vr.cor.transcription_factor.filtered[row.gene,col.gene]
  return(val)
})
vr.sequence.cor.vec <- sapply(1:nrow(receptor.seq_sim.cor.tf_sim.df), function(i){
  row.gene <- receptor.seq_sim.cor.tf_sim.df[i,1]
  col.gene <- receptor.seq_sim.cor.tf_sim.df[i,2]
  val <- vno.receptor.identity[row.gene,col.gene]
  return(val)
})

receptor.seq_sim.cor.tf_sim.df$tf_exp_cor <- vr.transcription_factor.cor.vec
receptor.seq_sim.cor.tf_sim.df$vr_seq_sim <- vr.sequence.cor.vec

# Create rownames based on geneA_geneB
new_rn <- paste(receptor.seq_sim.cor.tf_sim.df$Var1,receptor.seq_sim.cor.tf_sim.df$Var2,sep="_")
rownames(receptor.seq_sim.cor.tf_sim.df) <- new_rn

# Remove self-identity rows, i.e., Vmn1r1 and Vmn1r1
receptor.seq_sim.cor.tf_sim.df <- receptor.seq_sim.cor.tf_sim.df[!(receptor.seq_sim.cor.tf_sim.df$Var1 == receptor.seq_sim.cor.tf_sim.df$Var2), ]
# Remove gene names from dataframe
receptor.seq_sim.cor.tf_sim.df <- receptor.seq_sim.cor.tf_sim.df[ ,-c(1:2)]

# Remove duplicate rows
receptor.seq_sim.cor.tf_sim.df <- receptor.seq_sim.cor.tf_sim.df %>% distinct()

saveRDS(receptor.seq_sim.cor.tf_sim.df,file=paste0(object_dir,"receptor.seq_sim.cor.tf_sim.df.rds"))

```

```{r}

figS_9A <- ggplot(data=receptor.seq_sim.cor.tf_sim.df, aes(x=vr_seq_sim, y=tf_exp_cor)) + 
  geom_point(size=0.5) + theme_classic() +
  geom_pointdensity() + scale_color_viridis() + 
  labs(y="Transcription Factor Similarity",x="VR Similarity") +
  stat_summary_bin(fun.y='mean', bins=100, color='orange', 
                   geom='line', orientation='x', size=1.5, show.legend=TRUE) + 
  stat_summary_bin(fun.y='median', bins=100, color = 'blue', 
                   geom='line', linetype='dashed', orientation='x', size=1.5, show.legend=TRUE) +
  stat_summary_bin(fun.y='mean', bins=100, 
                   color='magenta', geom='line', orientation='y', size=1.5, show.legend=TRUE) + 
  stat_summary_bin(fun.y='median', bins=100, color = 'cyan', 
                   geom='line', linetype='dashed', orientation='y', size=1.5, show.legend=TRUE) +
  theme(legend.key.size =  unit(0.5, "in"),axis.title=element_text(size=24),
        axis.text=element_text(size=20),legend.title=element_text(size=20),
        legend.text=element_text(size=20), aspect.ratio=1)

figS_9A
  
```

## Heatmaps showing V1R-TF and V2R-TF associations. Heat shows average expression level for TF genes for a given receptor type  

```{r,fig.width=8,fig.height=16}

# The dendrogram for rows
dst <- dist(v1r.transcription_factor.b_in_a.mat)
hc_row <- hclust(dst)
row_dend <- as.dendrogram(hc_row)
row_dend <- seriate_dendrogram(row_dend, dst, method="OLO")

# The dendrogram for columns
dst <- dist(t(v1r.transcription_factor.b_in_a.mat))
hc_row <- hclust(dst)
col_dend <- as.dendrogram(hc_row)
col_dend <- seriate_dendrogram(col_dend, dst, method="OLO")

figS_9B <- pheatmap(v1r.transcription_factor.b_in_a.mat,fontsize=15,treeheight_row=25,treeheight_col=25,
                    cluster_rows=as.hclust(row_dend), cluster_cols=as.hclust(col_dend))

grid.draw(rectGrob(gp=gpar(fill="white", lwd=0)))
grid.draw(figS_9B[[4]])

```

```{r,fig.height=16,fig.width=16}

# The dendrogram for rows
dst <- dist(v2r.transcription_factor.b_in_a.mat)
hc_row <- hclust(dst)
row_dend <- as.dendrogram(hc_row)
row_dend <- seriate_dendrogram(row_dend, dst, method="OLO")

# The dendrogram for columns
dst <- dist(t(v2r.transcription_factor.b_in_a.mat))
hc_row <- hclust(dst)
col_dend <- as.dendrogram(hc_row)
col_dend <- seriate_dendrogram(col_dend, dst, method="OLO")

figS_9C <- pheatmap(v2r.transcription_factor.b_in_a.mat,fontsize=12,treeheight_row=25,treeheight_col=25,
                    cluster_rows=as.hclust(row_dend), cluster_cols=as.hclust(col_dend))

grid.draw(rectGrob(gp=gpar(fill="white", lwd=0)))
grid.draw(figS_9C[[4]])

```
# {-}

```{r include=FALSE}

current.variables <- c(ls(),"current.variables")
rm(list=setdiff(current.variables,c(global.vars)))
invisible(gc())

```

# Co-expression of chemosensory receptors {.tabset .tabset-pills}  

```{r}

my_theme_3 <- theme(axis.title.y=element_text(size=12),                         #change font size of y axis label
                    axis.title.x=element_blank(),                               #don't show x axis label
                    plot.title=element_text(size=16,hjust=0.5,face='bold'),     #change font size of plot title, center
                    legend.title=element_text(size=12),                         #change font size of legend title
                    legend.text=element_text(size=10),                          #change font size of legend text
                    axis.text.y=element_text(size=12),                          #change font size of axis tick labels
                    axis.text.x=element_blank(),                                #don't show x-axis tick labels
                    axis.ticks.x=element_blank(),                               #don't show x-axis ticks
                    legend.key.size=unit(8,'mm'),                               #the size of legend 'boxes'
                    legend.title.align=0,                                       #left align legend title
                    aspect.ratio=1)                                             #keep plots square

```

```{r}

metadata.compare <- FetchData(mature.neurons,vars=c("cell_type","percent.receptors",
                                                    "receptor.species","total.receptor.cnt",
                                                    "first.receptor.percent","second.receptor.percent",
                                                    "third.receptor.percent","first.receptor.count",
                                                    "second.receptor.count","third.receptor.count"))

```

```{r}

receptor.percent.df <- 
  pivot_longer(data=metadata.compare[ ,c(1,5:7)],
               cols=ends_with("percent"),
               names_to="receptor",
               names_pattern="(.*)\\.receptor\\.percent",
               values_to="percent")

```

## Percent of all read counts from a receptor gene, number of receptor species present per cell, and total counts from vomeronasal receptors, separated by cell type  

```{r,fig.width=18,fig.height=6}

my_comparisons <- list(c("V1R_mVSN","V2R_mVSN"),c("V2R_mVSN","mOSN"),c("mOSN","sVSN"),
                      c("V1R_mVSN","mOSN"),c("V2R_mVSN","sVSN"),c("V1R_mVSN","sVSN"))

figS_6A <- ggplot(metadata.compare,aes(x=cell_type,y=percent.receptors,fill=cell_type)) + geom_boxplot() +
  scale_fill_manual("Cell Type",values=mature.cell_type.colors) + theme_classic() +
  stat_summary(fun.y=mean,geom='point',size=10,colour="black",shape=95,show_guide=FALSE) +
  stat_compare_means(comparisons=my_comparisons,tip.length=0.01,
                     method="wilcox.test",label="p.signif",vjust=0.3) +
  ggtitle("percent.receptors") + ylab("receptor/total (%)") + my_theme_3 + scale_y_log10()
figS_6B <- ggplot(metadata.compare,aes(x=cell_type,y=receptor.species,fill=cell_type)) + geom_boxplot() +
  scale_fill_manual("Cell Type",values=mature.cell_type.colors) + theme_classic() +
  stat_summary(fun.y=mean,geom='point',size=10,colour="black",shape=95,show_guide=FALSE) +
  stat_compare_means(comparisons=my_comparisons,tip.length=0.01,
                     method="wilcox.test",label="p.signif",vjust=0.3) +
  ggtitle("receptor.species") + ylab("number of receptor species/cell") + my_theme_3 + scale_y_log10()
figS_6C <- ggplot(metadata.compare,aes(x=cell_type,y=total.receptor.cnt,fill=cell_type)) + geom_boxplot() +
  scale_fill_manual("Cell Type",values=mature.cell_type.colors) + theme_classic() +
  stat_summary(fun.y=mean,geom='point',size=10,colour="black",shape=95,show_guide=FALSE) +
  stat_compare_means(comparisons=my_comparisons,tip.length=0.01,
                     method="wilcox.test",label="p.signif",vjust=0.3) +
  ggtitle("total.receptor.cnt") + ylab("total normalized count") + my_theme_3 + scale_y_log10()

ggarrange(figS_6A,figS_6B,figS_6C,ncol=3,nrow=1,common.legend=TRUE,align="hv",legend="bottom")
  
```

## Shannon Indices showing the specificity of receptor expression. High values indicate more co-expression  

```{r}

alpha.diversity.df <- FetchData(neuron.integrated,vars=c("age","cell_type","Shannon.Index"))

```

```{r}

fig5_a <- ggplot(alpha.diversity.df,aes(x=cell_type,y=Shannon.Index,fill=cell_type)) + geom_boxplot() +
  scale_fill_manual("Cell Type",values=neuron.cell_type.colors) + #theme_classic() + 
  stat_summary(fun.y=mean,geom='point',size=10,colour="white",shape=95,show_guide=FALSE) +
  facet_grid(. ~ age) +
  # stat_compare_means(comparisons=alpha_comparisons.list,tip.length=0.01,
  #                    method="wilcox.test",label="p.signif",vjust=0.3) +
  ylab("Shannon Index") + xlab("Cell Type") + theme_bw() +
  theme(aspect.ratio=2, plot.title=element_text(size=32,face="bold"),
        axis.text.x=element_text(angle=45,vjust=1,hjust=1),
        axis.text=element_text(size=20), axis.title=element_text(size=24), axis.ticks.x=element_blank(),
        strip.text.x = element_text(size = 24)) + 
  NoLegend() 

fig5_a

```

## Proportions of first, second, and third most expressed receptor gene as percent of total receptor counts, separated by cell type  

```{r,fig.width=18,fig.height=6}

figS_6.part_2 <- ggplot(data=receptor.percent.df,aes(x = receptor, y = percent, fill = receptor)) +
    geom_boxplot() + theme_classic() +
    facet_grid(~ cell_type) + my_theme_3 + theme(legend.position="bottom", text=element_text(size=20),
                                                 legend.text=element_text(size=16),
                                                 axis.title.y=element_text(size=16)) 

figS_6.part_2

```

## Prevalence and level of receptor co-expression by age and cell-type for the V1R and V2R lineages, respectively  

```{r, include=FALSE}

receptor.metadata <- FetchData(neuron.integrated,
                               vars=c("cell_type","age","first.receptor.count","receptor.species"))

```

```{r}

V1R.receptor.metadata <- receptor.metadata[receptor.metadata$cell_type %in% c("late_INP","V1R_iVSN","V1R_mVSN"), ]

```

```{r,include=FALSE}

##### V1R #####
# Group data by age/cell_type combinations, and how many N cells 
V1R.co_exp.cell_type <- 
   summarySE(V1R.receptor.metadata, 
             measurevar = "first.receptor.count", 
             groupvars = c("age","cell_type"))

V1R.co_exp.cell_type$None <- unlist(lapply(rownames(V1R.co_exp.cell_type), function(i){
   age <- V1R.co_exp.cell_type[i, "age"]
   cell_type <- V1R.co_exp.cell_type[i, "cell_type"]
   nrow(V1R.receptor.metadata[V1R.receptor.metadata$age == age & 
                            V1R.receptor.metadata$cell_type == cell_type &
                            V1R.receptor.metadata$receptor.species == 0, ])/V1R.co_exp.cell_type[i, "N"]
}))

V1R.co_exp.cell_type$One <- unlist(lapply(rownames(V1R.co_exp.cell_type), function(i){
   age <- V1R.co_exp.cell_type[i, "age"]
   cell_type <- V1R.co_exp.cell_type[i, "cell_type"]
   nrow(V1R.receptor.metadata[V1R.receptor.metadata$age == age & 
                            V1R.receptor.metadata$cell_type == cell_type &
                            V1R.receptor.metadata$receptor.species == 1, ])/V1R.co_exp.cell_type[i, "N"]
}))

V1R.co_exp.cell_type$Two <- unlist(lapply(rownames(V1R.co_exp.cell_type), function(i){
   age <- V1R.co_exp.cell_type[i, "age"]
   cell_type <- V1R.co_exp.cell_type[i, "cell_type"]
   nrow(V1R.receptor.metadata[V1R.receptor.metadata$age == age & 
                            V1R.receptor.metadata$cell_type == cell_type &
                            V1R.receptor.metadata$receptor.species == 2, ])/V1R.co_exp.cell_type[i, "N"]
}))

V1R.co_exp.cell_type$Three_or_more <- unlist(lapply(rownames(V1R.co_exp.cell_type), function(i){
   age <- V1R.co_exp.cell_type[i, "age"]
   cell_type <- V1R.co_exp.cell_type[i, "cell_type"]
   nrow(V1R.receptor.metadata[V1R.receptor.metadata$age == age & 
                            V1R.receptor.metadata$cell_type == cell_type &
                            V1R.receptor.metadata$receptor.species >= 3, ])/V1R.co_exp.cell_type[i, "N"]
}))

```

```{r,include=FALSE}

V1R.plot.df <- V1R.co_exp.cell_type %>% gather(Receptors, Ratio, c("Three_or_more","Two","One","None"))

```

```{r,include=FALSE}

## Order the factors and round the floats in the DF  
##### V1R #####
# Order the Receptors factor levels for plotting
V1R.plot.df$Receptors <-
  factor(as.factor(V1R.plot.df$Receptors), 
         levels = c("Three_or_more","Two","One","None"))

# Order the age factor levels for plotting
V1R.plot.df$age <-
  factor(as.factor(V1R.plot.df$age), 
         levels = c("P14","P56"))

# Order the cell_type factor levels for plotting
V1R.plot.df$cell_type <-
  factor(as.factor(V1R.plot.df$cell_type), 
         levels = c("late_INP","V1R_iVSN","V1R_mVSN"))

# Round the ratio to 4 digits
V1R.plot.df$Ratio <- round(V1R.plot.df$Ratio, digits = 4)

```

```{r}

fig5_b <- ggplot(V1R.plot.df, aes(x = age, y = Ratio, fill = Receptors, label = Ratio)) + 
  geom_col() + theme(aspect.ratio=3,text=element_text(size=24),legend.position="top") +
  facet_grid(. ~ cell_type) +
  geom_text(data = subset(V1R.plot.df, Ratio > 0), size=5, position=position_stack(vjust = 0.5)) +
  geom_text(data = subset(V1R.plot.df, Receptors == "None"), size=5, aes(x=age, y=-0.01, label=paste0("N=",N)))

fig5_b

```

```{r}

V2R.receptor.metadata <- receptor.metadata[receptor.metadata$cell_type %in% c("late_INP","V2R_iVSN","V2R_mVSN"), ]

```

```{r}

##### V2R #####

V2R.co_exp.cell_type <- 
   summarySE(V2R.receptor.metadata, 
             measurevar = "first.receptor.count", 
             groupvars = c("age","cell_type"))

V2R.co_exp.cell_type$None <- unlist(lapply(rownames(V2R.co_exp.cell_type), function(i){
   age <- V2R.co_exp.cell_type[i, "age"]
   cell_type <- V2R.co_exp.cell_type[i, "cell_type"]
   nrow(V2R.receptor.metadata[V2R.receptor.metadata$age == age & 
                            V2R.receptor.metadata$cell_type == cell_type &
                            V2R.receptor.metadata$receptor.species == 0, ])/V2R.co_exp.cell_type[i, "N"]
}))

V2R.co_exp.cell_type$One <- unlist(lapply(rownames(V2R.co_exp.cell_type), function(i){
   age <- V2R.co_exp.cell_type[i, "age"]
   cell_type <- V2R.co_exp.cell_type[i, "cell_type"]
   nrow(V2R.receptor.metadata[V2R.receptor.metadata$age == age & 
                            V2R.receptor.metadata$cell_type == cell_type &
                            V2R.receptor.metadata$receptor.species == 1, ])/V2R.co_exp.cell_type[i, "N"]
}))

V2R.co_exp.cell_type$Two <- unlist(lapply(rownames(V2R.co_exp.cell_type), function(i){
   age <- V2R.co_exp.cell_type[i, "age"]
   cell_type <- V2R.co_exp.cell_type[i, "cell_type"]
   nrow(V2R.receptor.metadata[V2R.receptor.metadata$age == age & 
                            V2R.receptor.metadata$cell_type == cell_type &
                            V2R.receptor.metadata$receptor.species == 2, ])/V2R.co_exp.cell_type[i, "N"]
}))

V2R.co_exp.cell_type$Three_or_more <- unlist(lapply(rownames(V2R.co_exp.cell_type), function(i){
   age <- V2R.co_exp.cell_type[i, "age"]
   cell_type <- V2R.co_exp.cell_type[i, "cell_type"]
   nrow(V2R.receptor.metadata[V2R.receptor.metadata$age == age & 
                            V2R.receptor.metadata$cell_type == cell_type &
                            V2R.receptor.metadata$receptor.species >= 3, ])/V2R.co_exp.cell_type[i, "N"]
}))

```

```{r}

V2R.plot.df <- V2R.co_exp.cell_type %>% gather(Receptors, Ratio, c("Three_or_more","Two","One","None"))

```

```{r}

##### V2R #####

# Order the Receptors factor levels for plotting
V2R.plot.df$Receptors <-
  factor(as.factor(V2R.plot.df$Receptors), 
         levels = c("Three_or_more","Two","One","None"))

# Order the age factor levels for plotting
V2R.plot.df$age <-
  factor(as.factor(V2R.plot.df$age), 
         levels = c("P14","P56"))

# Order the cell_type factor levels for plotting
V2R.plot.df$cell_type <-
  factor(as.factor(V2R.plot.df$cell_type), 
         levels = c("early_INP","late_INP","V2R_iVSN","V2R_mVSN"))

# Round the ratio to 4 digits
V2R.plot.df$Ratio <- round(V2R.plot.df$Ratio, digits = 4)

```

```{r}

fig5_c <- ggplot(V2R.plot.df, aes(x = age, y = Ratio, fill = Receptors, label = Ratio)) +
  geom_col() + theme(aspect.ratio=3,text=element_text(size=24),legend.position="top") +
  facet_grid(. ~ cell_type) +
  geom_text(data = subset(V2R.plot.df, Ratio > 0), size=5, position = position_stack(vjust = 0.5)) +
  geom_text(data = subset(V2R.plot.df, Receptors == "None"), size=5, aes(x=age, y=-0.01, label=paste0("N=",N)))

fig5_c

```

## Circos plot of genomic loci for significantly co-expressed receptor pairs in the V1R, V2R, and across-type populations  

```{r}

load(file=paste0(object_dir, "sig.v1r.coexpress"))
load(file=paste0(object_dir, "v1r.circos.bed"))
load(file=paste0(object_dir, "v1r.bed1"))
load(file=paste0(object_dir, "v1r.bed2"))
load(file=paste0(object_dir, "sig.v2r.coexpress"))
load(file=paste0(object_dir, "v2r.circos.bed"))
load(file=paste0(object_dir, "v2r.bed1"))
load(file=paste0(object_dir, "v2r.bed2"))
load(file=paste0(object_dir, "sig.interclass.coexpress"))
load(file=paste0(object_dir, "interclass.circos.bed"))
load(file=paste0(object_dir, "interclass.bed1"))
load(file=paste0(object_dir, "interclass.bed2"))

```

```{r}

v1r.sectors <- lapply(unique(v1r.circos.bed$chr), function(chr){
  padding <- 100000
  sector.start <- min(v1r.circos.bed[v1r.circos.bed$chr == chr, "start"]) - padding
  sector.end <- max(v1r.circos.bed[v1r.circos.bed$chr == chr, "end"]) + padding
  return(c(chr,sector.start,sector.end))
})

v1r.chr.coord <- as.data.frame(do.call(rbind, v1r.sectors))
colnames(v1r.chr.coord) <- c("chr","start","end")

major.tix <- seq(from=0,to=250000000,by=100000)

```

```{r, fig.width=40,fig.height=40}

font.scale <- 7

circos.par(start.degree=90, gap.degree=2, track.margin=c(0.0005,0.0005))
circos.genomicInitialize(v1r.chr.coord,plotType = NULL)
# Add the geneSymbol labels
circos.genomicLabels(v1r.circos.bed, labels.column=4, side="outside", labels_height=0.29, cex=font.scale,
                     padding=0.04, connection_height=mm_h(15), line_lwd=5)

# Draw the sectors and add chr ID as text
circos.track(ylim=c(0,1), panel.fun=function(x,y){
  chr=CELL_META$sector.index
  xlim=CELL_META$xlim
  ylim=CELL_META$ylim
  circos.rect(xlim[1], 0, xlim[2], 1, col="darkred")
  circos.text(mean(xlim), mean(ylim), chr, cex=font.scale, col="white", facing="clockwise", niceFacing=TRUE)
  circos.xaxis(h=1, labels=FALSE, direction="outside", minor.ticks=1,
               major.tick.length=0.05, major.at=major.tix, lwd=5)
  }, track.height=0.22, bg.border=NA)
circos.genomicLink(v1r.bed1, v1r.bed2, col=sig.v1r.coexpress$color, lwd=5, border=NA)

```

```{r}

v2r.sectors <- lapply(unique(v2r.circos.bed$chr), function(chr){
  padding <- 100000
  sector.start <- min(v2r.circos.bed[v2r.circos.bed$chr == chr, "start"]) - padding
  sector.end <- max(v2r.circos.bed[v2r.circos.bed$chr == chr, "end"]) + padding
  return(c(chr,sector.start,sector.end))
})

v2r.chr.coord <- as.data.frame(do.call(rbind, v2r.sectors))
colnames(v2r.chr.coord) <- c("chr","start","end")

cols.num <- c("start","end")
v2r.chr.coord[cols.num] <- sapply(v2r.chr.coord[cols.num],as.numeric)

```

```{r, fig.width=50,fig.height=50}

font.scale <- 7

circos.par(start.degree=90, gap.degree=2, track.margin=c(0.0005,0.0005))
circos.genomicInitialize(v2r.chr.coord,plotType = NULL)
# Add the geneSymbol labels
circos.genomicLabels(v2r.circos.bed, labels.column=4, side="outside", labels_height=0.29, cex=font.scale,
                     padding=0.04, connection_height=mm_h(15), line_lwd=5)

# Draw the sectors and add chr ID as text
circos.track(ylim=c(0,1), panel.fun=function(x,y){
  chr=CELL_META$sector.index
  xlim=CELL_META$xlim
  ylim=CELL_META$ylim
  circos.rect(xlim[1], 0, xlim[2], 1, col="darkblue")
  circos.text(mean(xlim), mean(ylim), chr, cex=font.scale, col="white", facing="clockwise", niceFacing=TRUE)
  circos.xaxis(h=1, labels=FALSE, direction="outside", minor.ticks=1,
               major.tick.length=0.05, major.at=major.tix, lwd=5)
  }, track.height=0.22, bg.border=NA)
circos.genomicLink(v2r.bed1, v2r.bed2, col=sig.v2r.coexpress$color, lwd=5, border=NA)

```

```{r}

interclass.sectors <- lapply(unique(interclass.circos.bed$chr), function(chr){
  padding <- 100000
  sector.start <- min(interclass.circos.bed[interclass.circos.bed$chr == chr, "start"]) - padding
  sector.end <- max(interclass.circos.bed[interclass.circos.bed$chr == chr, "end"]) + padding
  return(c(chr,sector.start,sector.end))
})

interclass.chr.coord <- as.data.frame(do.call(rbind, interclass.sectors))
colnames(interclass.chr.coord) <- c("chr","start","end")

major.tix <- seq(from=0,to=250000000,by=100000)

```

```{r, fig.width=50,fig.height=50}

font.scale <- 7

circos.par(start.degree=90, gap.degree=2, track.margin=c(0.0005,0.0005))
circos.genomicInitialize(interclass.chr.coord,plotType = NULL)
# Add the geneSymbol labels
circos.genomicLabels(interclass.circos.bed, labels.column=4, side="outside", labels_height=0.29, cex=font.scale,
                     padding=0.04, connection_height=mm_h(15), line_lwd=5)

# Draw the sectors and add chr ID as text
circos.track(ylim=c(0,1), panel.fun=function(x,y){
  chr=CELL_META$sector.index
  xlim=CELL_META$xlim
  ylim=CELL_META$ylim
  circos.rect(xlim[1], 0, xlim[2], 1, col="darkgreen")
  circos.text(mean(xlim), mean(ylim), chr, cex=font.scale, col="white", facing="clockwise", niceFacing=TRUE)
  circos.xaxis(h=1, labels=FALSE, direction="outside", minor.ticks=1,
               major.tick.length=0.05, major.at=major.tix, lwd=5)
  }, track.height=0.22, bg.border=NA)
circos.genomicLink(interclass.bed1, interclass.bed2, col=sig.interclass.coexpress$color, lwd=5, border=NA)

```

# {-}

```{r include=FALSE}

current.variables <- c(ls(),"current.variables")
rm(list=setdiff(current.variables,c(global.vars)))
invisible(gc())

```

# Axon guidance molecules associated with receptor genes {.tabset .tabset-pills}  

```{r}

vr.cor.membrane.ag <- readRDS(file=paste0(object_dir,"vr.cor.membrane.ag.rds")) 
# Data frame with receptor sequence similarity scores and receptor average axon-guidance gene expression  
receptor.seq_sim.cor.ag_sim.df <- readRDS(file=paste0(object_dir,"receptor.seq_sim.cor.ag_sim.df.rds"))
# All significant VRs overlapping with surface expressed axon-guidance genes 
membrane.ag.b_in_a.mat <- readRDS(file=paste0(object_dir,"membrane.ag.b_in_a.mat.rds"))
# The above subset for strong groupings and split into V1R, V2R, and mixed-class matrices
v1r.membrane.ag.b_in_a.mat <- readRDS(file=paste0(object_dir,"v1r.membrane.ag.b_in_a.mat.rds"))
v2r.membrane.ag.b_in_a.mat <- readRDS(file=paste0(object_dir,"v2r.membrane.ag.b_in_a.mat.rds"))

```

## Distribution density plot showing relationship between similarity of axon guidance gene expression profiles and receptor sequence similarity  

```{r,fig.width=12,fig.height=12}

fig6_b <- ggplot(data=receptor.seq_sim.cor.ag_sim.df, aes(x=vr_seq_sim, y=pm_ag_exp_cor)) + 
  geom_point(size=0.5) + theme_classic() +
  geom_pointdensity() + scale_color_viridis() + 
  labs(y="AG profile Similarity",x="VR Similarity") +
  stat_summary_bin(fun.y='mean', bins=100, color='orange', 
                   geom='line', orientation='x', size=1.5, show.legend=TRUE) + 
  stat_summary_bin(fun.y='median', bins=100, color = 'dodgerblue', 
                   geom='line', linetype='dashed', orientation='x', size=1.5, show.legend=TRUE) +
  stat_summary_bin(fun.y='mean', bins=100, 
                   color='magenta', geom='line', orientation='y', size=1.5, show.legend=TRUE) + 
  stat_summary_bin(fun.y='median', bins=100, color = 'cyan', 
                   geom='line', linetype='dashed', orientation='y', size=1.5, show.legend=TRUE) +
  theme(legend.position="left",aspect.ratio=1,axis.text=element_text(size=20),legend.text=element_text(size=24),
        axis.title=element_text(size=32),legend.title=element_text(size=24),panel.border=element_blank(),
        plot.background=element_blank())
fig6_b
  
```

## Heatmap showing the correlation among VRs in their AG gene expression  

```{r,fig.height=50,fig.width=50}

membrane.ag.cluster.n <- 20

fig6_a <-
  pheatmap(vr.cor.membrane.ag,fontsize=18,
           cutree_rows=membrane.ag.cluster.n,
           cutree_cols=membrane.ag.cluster.n)

```

## Heatmap showing the V1R-AG gene associations  

```{r,fig.height=9.5,fig.width=5}

# The dendrogram for rows
dst <- dist(v1r.membrane.ag.b_in_a.mat)
hc_row <- hclust(dst)
row_dend <- as.dendrogram(hc_row)
row_dend <- seriate_dendrogram(row_dend, dst, method="OLO")

# The dendrogram for columns
dst <- dist(t(v1r.membrane.ag.b_in_a.mat))
hc_row <- hclust(dst)
col_dend <- as.dendrogram(hc_row)
col_dend <- seriate_dendrogram(col_dend, dst, method="OLO")

fig6_c <- pheatmap(v1r.membrane.ag.b_in_a.mat,fontsize=10,treeheight_row=25,treeheight_col=25,
                   cluster_rows=as.hclust(row_dend), cluster_cols=as.hclust(col_dend))

```

## Heatmap showing the V2R-AG gene associations  

```{r,fig.height=15,fig.width=7.5}

# The dendrogram for rows
dst <- dist(v2r.membrane.ag.b_in_a.mat)
hc_row <- hclust(dst)
row_dend <- as.dendrogram(hc_row)
row_dend <- seriate_dendrogram(row_dend, dst, method="OLO")

# The dendrogram for columns
dst <- dist(t(v2r.membrane.ag.b_in_a.mat))
hc_row <- hclust(dst)
col_dend <- as.dendrogram(hc_row)
col_dend <- seriate_dendrogram(col_dend, dst, method="OLO")

fig6_d <- pheatmap(v2r.membrane.ag.b_in_a.mat,cluster_rows=as.hclust(row_dend),
                   cluster_cols=as.hclust(col_dend),fontsize=10,treeheight_row=25,
                   treeheight_col=25)

```

# {-}

```{r include=FALSE}

current.variables <- c(ls(),"current.variables")
rm(list=setdiff(current.variables,c(global.vars)))
invisible(gc())

```

# Axon guidance molecules associated with receptor genes {.tabset .tabset-pills}  

```{r}

# Load the significant co-expression data frames
sig.vr.coexpression <- openxlsx::read.xlsx(xlsxFile=paste0(object_dir,"sig.mVSN.co.xlsx"))
rownames(sig.vr.coexpression) <- paste(sig.vr.coexpression$setA,sig.vr.coexpression$setB,sep="_")
sig.vr.ag.coexpression <- openxlsx::read.xlsx(xlsxFile=paste0(object_dir,"sig.vr.ag.coexpression.xlsx"))
rownames(sig.vr.ag.coexpression) <- paste(sig.vr.ag.coexpression$setA,sig.vr.ag.coexpression$setB,sep="_")
sig.vr.tf.coexpression <- openxlsx::read.xlsx(xlsxFile=paste0(object_dir,"sig.vr.tf.coexpression.xlsx"))
rownames(sig.vr.tf.coexpression) <- paste(sig.vr.tf.coexpression$setA,sig.vr.tf.coexpression$setB,sep="_")
vr.ag_tf.results <- openxlsx::read.xlsx(xlsxFile=paste0(object_dir,"vr.ag_tf.results.xlsx"))
rownames(vr.ag_tf.results) <- paste(vr.ag_tf.results$setA,vr.ag_tf.results$setB,sep="_")

# Load the gene names for each category
vr.names <- readRDS(file=paste0(object_dir,"final.receptor.names.rds"))
ag.names <- readRDS(file=paste0(object_dir,"final.membrane_ag.names.rds"))
tf.names <- readRDS(file=paste0(object_dir,"final.dna_binding_tf.names.rds"))

```

```{r}

VRs <- unique(sort(vr.ag_tf.results$VR))
AGs <- unique(sort(vr.ag_tf.results$AG))
TFs <- unique(sort(vr.ag_tf.results$TF))

```

```{r}

# Create a list of matrices with the Jaccard Index for AG-TF co-expression, for each VR
vno.mat.list <- lapply(VRs, function(vr){    # Loop through unique VRs
  # Create an empty matrix for this VR with AGs on columns and TFs on rows
  empty_mat <- matrix(data=0,ncol=length(AGs),nrow=length(TFs))
  rownames(empty_mat) <- TFs
  colnames(empty_mat) <- AGs
  # Subset our results for rows that contain this VR
  sub.df <- vr.ag_tf.results[vr.ag_tf.results$VR==vr, ]
  # For our results data frame, iterate over rows
  for(i in 1:nrow(sub.df)){
    # The AG for this row
    ag.name <- sub.df[i, ]$AG
    # The TF for this row
    tf.name <- sub.df[i, ]$TF
    # The Jaccard Index for this AG and this TF for this VR
    empty_mat[tf.name,ag.name] <- sub.df[i, ]$jaccard.index
  }
  return(empty_mat)    # Return the matrix for this VR
})
# Name matrices for their VR
names(vno.mat.list) <- VRs

```

```{r}

# Convert all matrices to vectors and check their pearson correlation
vno.vec.mat <- sapply(vno.mat.list, as.vector)
vno.cor.mat <- cor(vno.vec.mat)

```

```{r}

fig7_a.colors <- wesanderson::wes_palette("Darjeeling1")
# Create a row annotation data frame for figure 7A
fig7_a.row_annotations <- sapply(rownames(vno.cor.mat), function(rec){
  if(startsWith(x=rec,prefix="Fpr")){
    return("Fpr")
  } else if(startsWith(x=rec,prefix="Olfr")){
    return("Olfr")
  } else if(startsWith(x=rec,prefix="Vmn1r")){
    return("V1R")
  } else if(startsWith(x=rec,prefix="Vmn2r")){
    return("V2R")
  }
})

fig7_a.row_annotations <- as.data.frame(fig7_a.row_annotations)
colnames(fig7_a.row_annotations) <- "type"

fig7_a.ann_colors <- list(type=c(Fpr=fig7_a.colors[5],
                                 Olfr=fig7_a.colors[2],
                                 V1R=fig7_a.colors[3],
                                 V2R=fig7_a.colors[4]))

```

## Correlation heatmap between receptor types calculated from co-expressed AG and TF genes  

```{r,fig.width=12,fig.height=12}

# Plot the cross VR-specific correlation of AGxTF co-expression
fig7_a <-
  pheatmap(vno.cor.mat,fontsize=16,annotation_row=fig7_a.row_annotations, cellwidth=3.6,
           show_rownames=FALSE,show_colnames=FALSE,annotation_colors=fig7_a.ann_colors)

```

## 3D Heatmap showing the Jaccard Indices between AG and TF genes for each of the VRs in the dataset  

```{r}

vr.order.3d <- readRDS(file=paste0(object_dir,"vr.order.3d.rds"))
ag.order.3d <- readRDS(file=paste0(object_dir,"ag.order.3d.rds"))
tf.order.3d <- readRDS(file=paste0(object_dir,"tf.order.3d.rds"))

```

```{r}

# Mimic the color scheme of pheatmap
fig7_palette <- colorRampPalette(rev(RColorBrewer::brewer.pal(n=7, name="RdYlBu")))(100)
# Subset our results data frame
fig7_b.scatter.df <- vr.ag_tf.results[ ,c("VR","AG","TF","jaccard.index")]

```

```{r}

# Convert categorical variables to factors
fig7_b.scatter.df$AG <- factor(fig7_b.scatter.df$AG, levels = rev(ag.order.3d))
fig7_b.scatter.df$TF <- factor(fig7_b.scatter.df$TF, levels = rev(tf.order.3d))
fig7_b.scatter.df$VR <- factor(fig7_b.scatter.df$VR, levels = rev(vr.order.3d))

# Convert factors to underlying numeric representation
fig7_b.scatter.df$AG_numeric <- as.numeric(fig7_b.scatter.df$AG)
fig7_b.scatter.df$TF_numeric <- as.numeric(fig7_b.scatter.df$TF)
fig7_b.scatter.df$VR_numeric <- as.numeric(fig7_b.scatter.df$VR)

# Plotting the 3D scatterplot
scatter3D(
  x = fig7_b.scatter.df$AG_numeric,
  y = fig7_b.scatter.df$TF_numeric,
  z = fig7_b.scatter.df$VR_numeric,
  colvar = fig7_b.scatter.df$jaccard.index,  # Color by jaccard.index
  col=fig7_palette,
  pch = 19,  # Point shape
  theta = 45, phi = 30,  # Viewing angle
  xlab = "AG", ylab = "TF", zlab = "VR",  # Axis labels
  clab = "Jaccard Index",
  cex = 0.25    # Shrink the size of the data points
)

```

## Interactive 3D Heatmap VR-AG-TF Jaccard Indices  

```{r}

palette <- colorRampPalette(wesanderson::wes_palette("Zissou1"))

scatter3d.colors <- palette(n=100)

```


```{r}

plotly.scatter.df <- vr.ag_tf.results[ ,c("VR","AG","TF","jaccard.index")]

vr.ag.tf.scatter3d <- plot_ly(data = plotly.scatter.df,
                                    x = ~TF, y = ~AG, z = ~VR,
                                    color = ~jaccard.index,
                                    type = "scatter3d",
                                    mode = "markers",
                                    colors = scatter3d.colors,
                                    marker = list(size = 1, width=1), # controls size of points
                                    hoverinfo="all") %>% layout(title="VR-AG-TF Co-expression",
                                                                scene=list(xaxis=list(title='TF',
                                                                                      categoryorder="array",
                                                                                      categoryarray=tf.order.3d),
                                                                           yaxis=list(title='AG',
                                                                                      categoryorder="array",
                                                                                      categoryarray=ag.order.3d),
                                                                           zaxis=list(title='VR',
                                                                                      categoryorder="array",
                                                                                      categoryarray=vr.order.3d)))

```

```{r, fig.width=14, fig.height=14}

vr.ag.tf.scatter3d

```


## Heatmaps showing Jaccard Indices of TF-AG associations for V1R and V2R types. The lists of TF and AG genes here are abridged from the full list to enhance visualization  

```{r}

vr.goi <- c("Vmn1r69","Vmn1r185","Vmn1r34","Vmn1r56","Vmn1r60","Vmn1r61","Vmn2r1",
         "Vmn2r3","Vmn2r7","Vmn2r73","Vmn2r88","Vmn2r8","Vmn2r91","Vmn2r118")

ag.goi <- c("Flot1","Lgr4","Pcdh9","Kirrel2","Cdk5r2","Ncam1","Efna5","Nrp2","B3gnt2","Pcdh10",
            "Epha5","Robo2","Pcdh17","Tenm2","Gap43","Kirrel3","Plxnd1","App","Bsg","Efna3")

tf.goi <- c("Trps1","Pbx2","Meis2","Nfix","Tfap2e","Pou2f1","Batf3","Tcf4","Sp8","Cebpb",
            "Nhlh1","Foxg1","Foxp1","Atf5","Jund","Ebf2","Ebf3","Klf7","Emx2","Ikzf4")

filtered.vr.ag_tf.results <- vr.ag_tf.results[vr.ag_tf.results$VR %in% vr.goi &
                                                vr.ag_tf.results$AG %in% ag.goi &
                                                vr.ag_tf.results$TF %in% tf.goi, ]

```

```{r}

vr.order.3d.filtered <- vr.order.3d[vr.order.3d %in% vr.goi]
ag.order.3d.filtered <- ag.order.3d[ag.order.3d %in% ag.goi]
tf.order.3d.filtered <- tf.order.3d[tf.order.3d %in% tf.goi]

```

```{r}

filtered.vno.mat.list <- lapply(vr.order.3d.filtered, function(vr){
  empty_mat <- matrix(data=0,ncol=length(ag.order.3d.filtered),nrow=length(tf.order.3d.filtered))
  rownames(empty_mat) <- tf.order.3d.filtered
  colnames(empty_mat) <- ag.order.3d.filtered
  sub.df <- filtered.vr.ag_tf.results[filtered.vr.ag_tf.results$VR==vr, ]
  for(i in 1:nrow(sub.df)){
    ag.name <- sub.df[i, ]$AG
    tf.name <- sub.df[i, ]$TF
    empty_mat[tf.name,ag.name] <- sub.df[i, ]$jaccard.index
  }
  return(empty_mat)
})
names(filtered.vno.mat.list) <- vr.order.3d.filtered

```

```{r}

# Re-order the matrices by our new ag and tf orders
filtered.vno.mat.list <- lapply(filtered.vno.mat.list, function(mat){
  mat <- mat[tf.order.3d.filtered,ag.order.3d.filtered]
  return(mat)
})

```

```{r,fig.width=18,fig.height=6}

# Correlated V1Rs (Thelma)
Vmn1r185 <- pheatmap(filtered.vno.mat.list$Vmn1r185, fontsize = 16, cluster_cols = FALSE, 
                     cluster_rows = FALSE, main="Vmn1r185", show_rownames = FALSE, silent=TRUE)
Vmn1r34 <- pheatmap(filtered.vno.mat.list$Vmn1r34, fontsize = 16, cluster_cols = FALSE, 
                    cluster_rows = FALSE, main="Vmn1r34", show_rownames = FALSE, silent=TRUE)
Vmn1r69 <- pheatmap(filtered.vno.mat.list$Vmn1r69, fontsize = 16, cluster_cols = FALSE, 
                    cluster_rows = FALSE, main="Vmn1r69", silent=TRUE)

grid.arrange(Vmn1r185[[4]], Vmn1r34[[4]], Vmn1r69[[4]], ncol = 3, widths=c(1,1,1.14))

```

```{r,fig.width=18,fig.height=6}

Vmn1r56 <- pheatmap(filtered.vno.mat.list$Vmn1r56, fontsize = 16, cluster_cols = FALSE, 
                     cluster_rows = FALSE, main="Vmn1r56", show_rownames = FALSE, silent=TRUE)
Vmn1r60 <- pheatmap(filtered.vno.mat.list$Vmn1r60, fontsize = 16, cluster_cols = FALSE, 
                    cluster_rows = FALSE, main="Vmn1r60", show_rownames = FALSE, silent=TRUE)
Vmn1r61 <- pheatmap(filtered.vno.mat.list$Vmn1r61, fontsize = 16, cluster_cols = FALSE, 
                    cluster_rows = FALSE, main="Vmn1r61", silent=TRUE)


#svglite(filename = paste0(image_dir,"fig7_d.svg"),width=24,height=8)
grid.arrange(Vmn1r56[[4]], Vmn1r60[[4]], Vmn1r61[[4]], ncol = 3, widths=c(1,1,1.14))
#invisible(dev.off())

```

```{r,fig.width=18,fig.height=6}

Vmn2r1 <- pheatmap(filtered.vno.mat.list$Vmn2r1, fontsize = 16, cluster_cols = FALSE, 
                     cluster_rows = FALSE, main="Vmn2r1", show_rownames = FALSE, silent=TRUE)
Vmn2r3 <- pheatmap(filtered.vno.mat.list$Vmn2r3, fontsize = 16, cluster_cols = FALSE, 
                    cluster_rows = FALSE, main="Vmn2r3", show_rownames = FALSE, silent=TRUE)
Vmn2r7 <- pheatmap(filtered.vno.mat.list$Vmn2r7, fontsize = 16, cluster_cols = FALSE, 
                    cluster_rows = FALSE, main="Vmn2r7", silent=TRUE)


#svglite(filename = paste0(image_dir,"fig7_e.svg"),width=24,height=8)
grid.arrange(Vmn2r1[[4]], Vmn2r3[[4]], Vmn2r7[[4]], ncol = 3, widths=c(1,1,1.14))
#invisible(dev.off())

```

```{r,fig.width=18,fig.height=6}

Vmn2r8 <- pheatmap(filtered.vno.mat.list$Vmn2r8, fontsize = 16, cluster_cols = FALSE, 
                     cluster_rows = FALSE, main="Vmn2r8", show_rownames = FALSE, silent=TRUE)
Vmn2r73 <- pheatmap(filtered.vno.mat.list$Vmn2r73, fontsize = 16, cluster_cols = FALSE, 
                    cluster_rows = FALSE, main="Vmn2r73", show_rownames = FALSE, silent=TRUE)
Vmn2r88 <- pheatmap(filtered.vno.mat.list$Vmn2r88, fontsize = 16, cluster_cols = FALSE, 
                    cluster_rows = FALSE, main="Vmn2r88", silent=TRUE)


#svglite(filename = paste0(image_dir,"fig7_f.svg"),width=24,height=8)
grid.arrange(Vmn2r8[[4]], Vmn2r73[[4]], Vmn2r88[[4]], ncol = 3, widths=c(1,1,1.14))
#invisible(dev.off())

```

```{r,fig.width=10,fig.height=5}

Vmn2r91 <- pheatmap(filtered.vno.mat.list$Vmn2r91, fontsize = 16, cluster_cols = FALSE, 
                     cluster_rows = FALSE, main="Vmn2r91", show_rownames = FALSE, silent=TRUE)
Vmn2r118 <- pheatmap(filtered.vno.mat.list$Vmn2r118, fontsize = 16, cluster_cols = FALSE, 
                    cluster_rows = FALSE, main="Vmn2r118", silent=TRUE)

#svglite(filename = paste0(image_dir,"fig7_g.svg"),width=16,height=8)
grid.arrange(Vmn2r91[[4]], Vmn2r118[[4]], ncol = 2, widths=c(1,1.14))
#invisible(dev.off())

```

# {-}

```{r include=FALSE}

current.variables <- c(ls(),"current.variables")
rm(list=setdiff(current.variables,c(global.vars)))
invisible(gc())

```

# Data Tables {.tabset .tabset-pills}  

## Significant VR co-expression  

```{r}

# Load the significant co-expression data frames
sig.vr.coexpression <- openxlsx::read.xlsx(xlsxFile=paste0(object_dir,"sig.mVSN.co.xlsx"))
sig.vr.coexpression

```

## Significant VR / AG Co-expression  

```{r}

sig.vr.ag.coexpression <- openxlsx::read.xlsx(xlsxFile=paste0(object_dir,"sig.vr.ag.coexpression.xlsx"))
sig.vr.ag.coexpression

```

## Significant VR / TF Co-expression  

```{r}

sig.vr.tf.coexpression <- openxlsx::read.xlsx(xlsxFile=paste0(object_dir,"sig.vr.tf.coexpression.xlsx"))
sig.vr.tf.coexpression

```

## VR-specific AG / TF co-expression  

```{r}

vr.ag_tf.results <- openxlsx::read.xlsx(xlsxFile=paste0(object_dir,"vr.ag_tf.results.xlsx"))
vr.ag_tf.results

```

```{r}

# Differential Expression of Genes Across Cell-Types
V1R_iVSN_V2R_iVSN.DEGs <- 
  read.xlsx(xlsxFile = paste0(object_dir,"V1R_iVSN_V2R_iVSN.DEGs.xlsx"),rowNames=TRUE)
V1R_mVSN_V2R_mVSN.DEGs <- 
  read.xlsx(xlsxFile = paste0(object_dir,"V1R_mVSN_V2R_mVSN.DEGs.xlsx"),rowNames=TRUE)
unknown_mVSN_V1R_mVSN.DEGs <- 
  read.xlsx(xlsxFile = paste0(object_dir,"unknown_mVSN_V1R_mVSN.DEGs.xlsx"),rowNames=TRUE)
unknown_mVSN_V2R_mVSN.DEGs <- 
  read.xlsx(xlsxFile = paste0(object_dir,"unknown_mVSN_V2R_mVSN.DEGs.xlsx"),rowNames=TRUE)

```

## V1R_iVSN vs V2R_iVSN differentially expressed genes (DEGs)  

```{r}
V1R_iVSN_V2R_iVSN.DEGs
```

## V1R_mVSN vs V2R_mVSN DEGs  

```{r}
V1R_mVSN_V2R_mVSN.DEGs
```

## sVSN vs V1R_mVSN DEGs  

```{r}
unknown_mVSN_V1R_mVSN.DEGs
```

## sVSN vs V2R_mVSN DEGs  

```{r}
unknown_mVSN_V2R_mVSN.DEGs
```

## Early INP - V1R vs V2R  

```{r}

# Load the significant co-expression data frames
early_INP.v1r_v2r.DEGs <- openxlsx::read.xlsx(xlsxFile=paste0(object_dir,"v1r.v.v2r.inp1.markers.xlsx"))
early_INP.v1r_v2r.DEGs[early_INP.v1r_v2r.DEGs$p_val_adj<=0.05, ]

```

## Late INP - V1R vs V2R  

```{r}

# Load the significant co-expression data frames
late_INP.v1r_v2r.DEGs <- openxlsx::read.xlsx(xlsxFile=paste0(object_dir,"v1r.v.v2r.inp2.markers.xlsx"))
late_INP.v1r_v2r.DEGs[late_INP.v1r_v2r.DEGs$p_val_adj <= 0.05, ]

```

























```{r}

### Supplemental 6 - Cell Receptor Profile by Type  

my_theme_3 <- theme(axis.title.y=element_text(size=12),                         #change font size of y axis label
                    axis.title.x=element_blank(),                               #don't show x axis label
                    plot.title=element_text(size=16,hjust=0.5,face='bold'),     #change font size of plot title, center
                    legend.title=element_text(size=12),                         #change font size of legend title
                    legend.text=element_text(size=10),                          #change font size of legend text
                    axis.text.y=element_text(size=12),                          #change font size of axis tick labels
                    axis.text.x=element_blank(),                                #don't show x-axis tick labels
                    axis.ticks.x=element_blank(),                               #don't show x-axis ticks
                    legend.key.size=unit(8,'mm'),                               #the size of legend 'boxes'
                    legend.title.align=0,                                       #left align legend title
                    aspect.ratio=1)                                             #keep plots square

```

```{r}

metadata.compare <- FetchData(mature.neurons,vars=c("cell_type","percent.receptors",
                                                    "receptor.species","total.receptor.cnt",
                                                    "first.receptor.percent","second.receptor.percent",
                                                    "third.receptor.percent","first.receptor.count",
                                                    "second.receptor.count","third.receptor.count"))

metadata.compare[ ,c(1,5:7)]

```

```{r}

receptor.percent.df <- 
  pivot_longer(data=metadata.compare[ ,c(1,5:7)],
               cols=ends_with("percent"),
               names_to="receptor",
               names_pattern="(.*)\\.receptor\\.percent",
               values_to="percent")

receptor.percent.df

```


```{r}

my_comparisons <- list(c("V1R_mVSN","V2R_mVSN"),c("V2R_mVSN","mOSN"),c("mOSN","sVSN"),
                      c("V1R_mVSN","mOSN"),c("V2R_mVSN","sVSN"),c("V1R_mVSN","sVSN"))

figS_6A <- ggplot(metadata.compare,aes(x=cell_type,y=percent.receptors,fill=cell_type)) + geom_boxplot() +
  scale_fill_manual("Cell Type",values=mature.cell_type.colors) + theme_classic() +
  stat_summary(fun.y=mean,geom='point',size=10,colour="black",shape=95,show_guide=FALSE) +
  stat_compare_means(comparisons=my_comparisons,tip.length=0.01,
                     method="wilcox.test",label="p.signif",vjust=0.3) +
  ggtitle("percent.receptors") + ylab("receptor/total (%)") + my_theme_3 + scale_y_log10()
figS_6B <- ggplot(metadata.compare,aes(x=cell_type,y=receptor.species,fill=cell_type)) + geom_boxplot() +
  scale_fill_manual("Cell Type",values=mature.cell_type.colors) + theme_classic() +
  stat_summary(fun.y=mean,geom='point',size=10,colour="black",shape=95,show_guide=FALSE) +
  stat_compare_means(comparisons=my_comparisons,tip.length=0.01,
                     method="wilcox.test",label="p.signif",vjust=0.3) +
  ggtitle("receptor.species") + ylab("number of receptor species/cell") + my_theme_3 + scale_y_log10()
figS_6C <- ggplot(metadata.compare,aes(x=cell_type,y=total.receptor.cnt,fill=cell_type)) + geom_boxplot() +
  scale_fill_manual("Cell Type",values=mature.cell_type.colors) + theme_classic() +
  stat_summary(fun.y=mean,geom='point',size=10,colour="black",shape=95,show_guide=FALSE) +
  stat_compare_means(comparisons=my_comparisons,tip.length=0.01,
                     method="wilcox.test",label="p.signif",vjust=0.3) +
  ggtitle("total.receptor.cnt") + ylab("total normalized count") + my_theme_3 + scale_y_log10()

figS_6.part_1 <- ggarrange(figS_6A,figS_6B,figS_6C,ncol=3,nrow=1,common.legend=TRUE,align="hv",legend="bottom")
  
```

```{r}

svglite(filename = paste0(adobe_images,"figS_6.part_1.svg"), width=15, height=5)
figS_6.part_1
invisible(dev.off())

```


```{r,fig.width=15,fig.height=5}

#### A-C  
figS_6.part_1

```


```{r,fig.width=15,fig.height=5}

#### D-F  
figS_6.part_2 <- ggplot(data=receptor.percent.df,aes(x = receptor, y = percent, fill = receptor)) +
    geom_boxplot() + theme_classic() + theme(legend.position="bottom") +
    facet_grid(~ cell_type) + my_theme_3

figS_6.part_2

```

```{r}

svglite(filename = paste0(adobe_images,"figS_6.part_2.svg"), width=15, height=5)
figS_6.part_2
invisible(dev.off())

```


```{r}

#### Do sVSNs Express Fewer Axon-Guidance and Synaptic Genes?  

# Get all axon-guidance and synaptic genes
ag.genes <- read.delim(paste0(home_dir,"axon_guidance/membrane.ag.txt"), header=FALSE)[ ,1]
synapse.genes <- read.delim(paste0(home_dir,"axon_guidance/synapse.txt"), header=FALSE)[ ,1]
ag.genes <- ag.genes[!(ag.genes=="Symbol")]

```

```{r}

mature.neuron.ctrl <- subset(mature.neurons,idents=c("sVSN"),invert=TRUE)
mature.neuron.svsn <- subset(mature.neurons,idents=c("sVSN"))

ag.features.ctrl <- FetchData(object=mature.neuron.ctrl,vars=ag.genes,slot='data') 
ag.features.svsn <- FetchData(object=mature.neuron.svsn,vars=ag.genes,slot='data')

synapse.features.ctrl <- FetchData(object=mature.neuron.ctrl,vars=synapse.genes,slot='data')
synapse.features.svsn <- FetchData(object=mature.neuron.svsn,vars=synapse.genes,slot='data')

```

```{r}

paste("Non-sVSNs express",sum(colSums(ag.features.ctrl)!=0),"surface axon-guidance genes",
      "while sVSNs express",sum(colSums(ag.features.svsn)!=0),
      "of a total of",length(ag.genes),"surface axon-guidance genes.")
paste("Non-sVSNs express",sum(colSums(synapse.features.ctrl)!=0),"synaptic genes",
      "while sVSNs express",sum(colSums(synapse.features.svsn)!=0),
      "of a total of",length(synapse.genes),"synaptic genes.")

```





```{r}

## D - U Neuronal UMAP <ZOOMED> and Genes-of-Interest  
corner_theme <- theme(panel.background = element_rect(fill = "transparent", colour = NA),
                      plot.background = element_rect(fill = "transparent", colour = NA),
                      panel.grid = element_blank(),
                      panel.border = element_blank(),
                      plot.margin = unit(c(-1, -1, 0, -1), "null"),
                      panel.margin = unit(c(-1, -1, 0, -1), "null"),
                      axis.text=element_text(size=20),
                      axis.title=element_text(size=24),
                      plot.title=element_text(size=32),
                      legend.text=element_text(size=20),
                      aspect.ratio=1)

y_theme <- theme(panel.background = element_rect(fill = "transparent", colour = NA),
                 plot.background = element_rect(fill = "transparent", colour = NA),
                 panel.grid = element_blank(),
                 panel.border = element_blank(),
                 plot.margin = unit(c(-1, -1, 0, -1), "null"),
                 panel.margin = unit(c(-1, -1, 0, -1), "null"),
                 axis.title.x=element_blank(),
                 axis.text=element_text(size=20),
                 axis.title=element_text(size=24),
                 plot.title=element_text(size=32),
                 legend.text=element_text(size=20),
                 aspect.ratio=1)

x_theme <- theme(panel.background = element_rect(fill = "transparent", colour = NA),
                 plot.background = element_rect(fill = "transparent", colour = NA),
                 panel.grid = element_blank(),
                 panel.border = element_blank(),
                 plot.margin = unit(c(-1, -1, 0, -1), "null"),
                 panel.margin = unit(c(-1, -1, 0, -1), "null"),
                 axis.title.y=element_blank(),
                 axis.text=element_text(size=20),
                 axis.title=element_text(size=24),
                 plot.title=element_text(size=32),
                 legend.text=element_text(size=20),
                 aspect.ratio=1)
  
blank_theme <- theme(panel.background = element_rect(fill = "transparent", colour = NA),
                     plot.background = element_rect(fill = "transparent", colour = NA),
                     panel.grid = element_blank(),
                     panel.border = element_blank(),
                     plot.margin = unit(c(-1, -1, 0, -1), "null"),
                     panel.margin = unit(c(-1, -1, 0, -1), "null"),
                     axis.title=element_blank(),
                     axis.text=element_text(size=20),
                     plot.title=element_text(size=32),
                     legend.text=element_text(size=20),
                     aspect.ratio=1)

```

```{r,fig.width=24,fig.height=12}

fig3_d_thru_u <- ggarrange(fig3_d,fig3_e,fig3_f,fig3_g,fig3_h,fig3_i,
                           fig3_j,fig3_k,fig3_l,fig3_m,fig3_n,fig3_o,
                           fig3_p,fig3_q,fig3_r,fig3_s,fig3_t,fig3_u,
                           ncol=6,nrow=3,align="hv")
svglite(filename = paste0(adobe_images,"fig3_d_thru_u.svg"),width=24,height=12)
fig3_d_thru_u
invisible(dev.off())

fig3_d_thru_u

```




```{r,include=FALSE}

### Supplemental 4 - V1R/V2R/OSN DEGs  
# Blank panels for A & B to fit.
figS_4.blank1 <- FeaturePlot(inp.integrated,features=c("Omp"),order=TRUE,pt.size=1,cols=c("cyan","magenta")) +
  y_theme

figS_4.blank2 <- FeaturePlot(inp.integrated,features=c("Gap43"),order=TRUE,pt.size=1,cols=c("cyan","magenta")) +
  y_theme


figS_4S <- FeaturePlot(inp.integrated,features=c("Notch1"),order=TRUE,pt.size=1,cols=c("cyan","magenta")) +
  x_theme

figS_4T <- FeaturePlot(inp.integrated,features=c("Dll4"),order=TRUE,pt.size=1,cols=c("cyan","magenta")) +
  x_theme

```


```{r,fig.width=15,fig.height=19}

#### C-T  
figS_4CT <- ggarrange(figS_4.blank1,figS_4.blank2,figS_4C,figS_4D,figS_4E,figS_4F,figS_4G,figS_4H,
                          figS_4I,figS_4J,figS_4K,figS_4L,figS_4M,figS_4N,figS_4O,figS_4P,
                          figS_4Q,figS_4R,figS_4S,figS_4T,
                          ncol=4,nrow=5,align="hv")
svglite(filename = paste0(adobe_images,"figS_4CT.svg"),width=15,height=19)
figS_4CT
invisible(dev.off())

figS_4CT

```



```{r}

#### O-P  
figS_4.13 <- FeaturePlot(inp.integrated,features=c("Notch1"),order=TRUE,pt.size=2,cols=c("cyan","magenta")) +
  corner_theme

figS_4.14 <- FeaturePlot(inp.integrated,features=c("Dll4"),order=TRUE,pt.size=2,cols=c("cyan","magenta")) +
  x_theme

figS_4.OP <- ggarrange(figS_4.13,figS_4.14,ncol=2,nrow=1,align="hv")

svglite(filename = paste0(adobe_images,"figS_4.OP.svg"),width=15,height=7.5)
figS_4.OP
invisible(dev.off())

```

```{r,fig.width=15,fig.height=7.5}

figS_4.OP

```


```{r,include=FALSE}

## Figure 3 Supplemental 3  
goi <- c("Notch1","Dll4","Bcl11b","Meis2","Sct","Tfap2e")
# create the plot list with geneSymbol as title, padj as subtitle
fig3_supplemental_3.pl <- lapply(goi, function(gene){
  tradeSeq::plotSmoothers(models=v1r.v.v2r.sce, counts=assays(v1r.v.v2r.sce)$counts,
                gene=gene, alpha=1, size=1, border=FALSE) +
    ggtitle(gene,
            subtitle=paste("padj:",format(v1r.v.v2r.patternDEres[gene, "padj"], scientific=TRUE, digits=6))) +
    theme(aspect.ratio=1, text=element_text(size=12), axis.title=element_text(size=16),
          plot.title=element_text(size=18), axis.text=element_text(size=12))
})

# What are the dimensions of the multi-plot page?
fig3_supplemental_3 <- ggarrange(plotlist=fig3_supplemental_3.pl, nrow=2, ncol=3, common.legend=TRUE)

```

```{r,fig.width=12,fig.height=8}

svglite(filename = paste0(adobe_images,"fig3_supplemental_3.svg"), width=12, height=8)
fig3_supplemental_3
invisible(dev.off())
fig3_supplemental_3

```

```{r,fig.width=12,fig.height=8}

FeaturePlot(inp.integrated,features=c("Notch1","Dll4","Bcl11b","Meis2","Sct","Tfap2e"),ncol=3,
            order=TRUE, cols=c("cyan","magenta"))

```


```{r include=FALSE}
# {-}

current.variables <- c(ls(),"current.variables")
rm(list=setdiff(current.variables,c(global.vars)))
invisible(gc())

```



```{r}

## E - V1R nCells  
v1r.gamma.model <- glm(avg_exp ~ poly(rank, 15, raw=TRUE), data=v1r.metadata,
                       control=glm.control(maxit=1000), family=Gamma(link="log"))
v1r.metadata$predicted <- predict(v1r.gamma.model, type="response")

# Plot the data and the fitted model
p <- ggplot(v1r.metadata, aes(x = rank, y = avg_exp)) +
  geom_point(color="gold2",size=5) + theme_classic() +
  geom_line(aes(y = predicted), color = "black") +
  theme(plot.title=element_text(size=32),
        axis.title=element_text(size=24),
        axis.text=element_text(size=20),
        aspect.ratio=1) +
  geom_label_repel(aes(label = gene), size=7, label.size = 0,
                   data = ~ subset(., gene == v1r.metadata$gene[1:6]), 
                   nudge_x=30,nudge_y=10,min.segment.length=0) +
  labs(x="rank", y="avg_exp")
p

```

```{r}

fig4_e <- ggbarplot(v1r.metadata,x="gene",y="nCells",fill="gold2",color="gold2",width=1,
                    title="V1R") +
  theme(plot.title=element_text(size=64),
        axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        axis.title.y=element_text(size=48),
        axis.text.y=element_text(size=40), aspect.ratio=1) +
  geom_label_repel(aes(label = gene), size=12, label.size = 0,
                   data = ~ subset(., gene == v1r.metadata$gene[1:6]), 
                   nudge_x=50,nudge_y=50,min.segment.length=0)# +
  #scale_y_continuous(expand=c(0,0),limits=c(0,175000))

# Save image to svg file
svglite(filename = paste0(adobe_images,"fig4_e.svg"),height=15,width=15)
fig4_e
invisible(dev.off())

```

```{r,fig.width=15,fig.height=15}
fig4_e
```


```{r}

## F - V2R Total Counts  
# Order by most expressed genes
## set the levels in order we want
rownames(v2r.metadata) <- NULL
v2r.metadata$gene <- 
  factor(v2r.metadata$gene,levels=v2r.metadata$gene[order(v2r.metadata$Total_Counts,decreasing=TRUE)])
v2r.metadata <- v2r.metadata[order(v2r.metadata$Total_Counts,decreasing = TRUE), ]

fig4_f <- ggbarplot(v2r.metadata,x="gene",y="Total_Counts",fill="purple",color="purple",width=1,
                    title="V2R") +
  theme(plot.title=element_text(size=64),
        axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        axis.title.y=element_text(size=48),
        axis.text.y=element_text(size=40), aspect.ratio=1) +
  geom_label_repel(aes(label = gene), size=12, label.size = 0,
                   data = ~ subset(., gene == v2r.metadata$gene[1:6]), 
                   nudge_x=30,nudge_y=30,min.segment.length=0) +
  scale_y_continuous(expand=c(0,0),limits=c(0,50000))

# Save image to svg file
svglite(filename = paste0(adobe_images,"fig4_f.svg"),height=15,width=15)
fig4_f
invisible(dev.off())

```

```{r,fig.width=15,fig.height=15}
fig4_f
```


```{r}

## G - Olfr Total Counts  
# Order by most expressed genes
## set the levels in order we want
rownames(olfr.metadata) <- NULL
olfr.metadata$gene <- 
  factor(olfr.metadata$gene,levels=olfr.metadata$gene[order(olfr.metadata$Total_Counts,decreasing=TRUE)])
olfr.metadata <- olfr.metadata[order(olfr.metadata$Total_Counts,decreasing = TRUE), ]

fig4_g <- ggbarplot(olfr.metadata,x="gene",y="Total_Counts",fill="green3",color="green3",width=1,
                    title="Olfr") +
  theme(plot.title=element_text(size=64),
        axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        axis.title.y=element_text(size=48),
        axis.text.y=element_text(size=40), aspect.ratio=1) +
  geom_label_repel(aes(label = gene), size=12, label.size = 0,
                   data = ~ subset(., gene == olfr.metadata$gene[1:6]), 
                   nudge_x=200,nudge_y=50,min.segment.length=0) +
  scale_y_continuous(expand=c(0,0),limits=c(0,20000))

# Save image to svg file
svglite(filename = paste0(adobe_images,"fig4_g.svg"),height=15,width=15)
fig4_g
invisible(dev.off())

```

```{r,fig.width=15,fig.height=15}
fig4_g
```


```{r}
## H - MOE Olfr Total Counts  

# Order by most expressed genes
## set the levels in order we want
rownames(moe.olfr.metadata) <- NULL
moe.olfr.metadata$gene <- 
  factor(moe.olfr.metadata$gene,levels=moe.olfr.metadata$gene[order(moe.olfr.metadata$Total_Counts,decreasing=TRUE)])
moe.olfr.metadata <- moe.olfr.metadata[order(moe.olfr.metadata$Total_Counts,decreasing = TRUE), ]

fig4_h <- ggbarplot(moe.olfr.metadata,x="gene",y="Total_Counts",fill="magenta",color="magenta",width=1,
                    title="MOE Olfr") +
  theme(plot.title=element_text(size=64),
        axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        axis.title.y=element_text(size=48),
        axis.text.y=element_text(size=40), aspect.ratio=1) +
  geom_label_repel(aes(label = gene), size=12, label.size = 0,
                   data = ~ subset(., gene == moe.olfr.metadata$gene[1:6]), 
                   nudge_x=200,nudge_y=50,min.segment.length=0) +
  scale_y_continuous(expand=c(0,0),limits=c(0,6000))

# Save image to svg file
svglite(filename = paste0(adobe_images,"fig4_h.svg"),height=15,width=15)
fig4_h
invisible(dev.off())

```

```{r,fig.width=15,fig.height=15}
fig4_h
```


```{r,fig.width=14,fig.height=14}
## I - V1R Transcription Factor Expression Correlation  

fig4_i <-
  pheatmap(v1r.cor.transcription_factor,fontsize=10,treeheight_row=25,treeheight_col=25)

# Save image to svg file
svglite(filename = paste0(adobe_images,"fig4_i.svg"),height=14,width=14)
grid.draw(rectGrob(gp=gpar(fill="white", lwd=0)))
grid.draw(fig4_i[[4]])
invisible(dev.off())

```


```{r,fig.width=14,fig.height=14}
## J - V2R Transcription Factor Expression Correlation  

fig4_j <-
  pheatmap(v2r.cor.transcription_factor,fontsize=10,treeheight_row=25,treeheight_col=25)

# Save image to svg file
svglite(filename = paste0(adobe_images,"fig4_j.svg"),height=14,width=14)
grid.draw(rectGrob(gp=gpar(fill="white", lwd=0)))
grid.draw(fig4_j[[4]])
invisible(dev.off())

```


```{r}
### Supplemental 9A - Scatter Plot: (VR Sequence Similarity) vs (VR-Specific Average-Transcription-Factor Gene Expression Correlation)  

# Find the index of each significant receptor co-expressed with a transcription factor gene within the receptor sequence identity matrix
match.idx <- match(x=rownames(vr.cor.transcription_factor),table=rownames(vno.receptor.identity))
# Remove NAs for genes not present in the sequence identity matrix
match.idx <- na.omit(match.idx)
# Subset our sequence identity matrix to match our vr.cor.transcription_factor matrix
vno.receptor.identity <- vno.receptor.identity[match.idx,match.idx]
# Missing receptors
missing.receptors <- setdiff(rownames(vr.cor.transcription_factor), rownames(vno.receptor.identity))
# Remove missing receptors
vr.cor.transcription_factor.filtered <- 
  vr.cor.transcription_factor[!(rownames(vr.cor.transcription_factor) %in% missing.receptors),!(colnames(vr.cor.transcription_factor) %in% missing.receptors)]

```

```{r}

receptor.seq_sim.cor.tf_sim.df <- expand.grid(rownames(vr.cor.transcription_factor.filtered), 
                                              colnames(vr.cor.transcription_factor.filtered), 
                                              stringsAsFactors=FALSE)

vr.transcription_factor.cor.vec <- sapply(1:nrow(receptor.seq_sim.cor.tf_sim.df), function(i){
  row.gene <- receptor.seq_sim.cor.tf_sim.df[i,1]
  col.gene <- receptor.seq_sim.cor.tf_sim.df[i,2]
  val <- vr.cor.transcription_factor.filtered[row.gene,col.gene]
  return(val)
})
vr.sequence.cor.vec <- sapply(1:nrow(receptor.seq_sim.cor.tf_sim.df), function(i){
  row.gene <- receptor.seq_sim.cor.tf_sim.df[i,1]
  col.gene <- receptor.seq_sim.cor.tf_sim.df[i,2]
  val <- vno.receptor.identity[row.gene,col.gene]
  return(val)
})

receptor.seq_sim.cor.tf_sim.df$tf_exp_cor <- vr.transcription_factor.cor.vec
receptor.seq_sim.cor.tf_sim.df$vr_seq_sim <- vr.sequence.cor.vec

# Create rownames based on geneA_geneB
new_rn <- paste(receptor.seq_sim.cor.tf_sim.df$Var1,receptor.seq_sim.cor.tf_sim.df$Var2,sep="_")
rownames(receptor.seq_sim.cor.tf_sim.df) <- new_rn

# Remove self-identity rows, i.e., Vmn1r1 and Vmn1r1
receptor.seq_sim.cor.tf_sim.df <- receptor.seq_sim.cor.tf_sim.df[!(receptor.seq_sim.cor.tf_sim.df$Var1 == receptor.seq_sim.cor.tf_sim.df$Var2), ]
# Remove gene names from dataframe
receptor.seq_sim.cor.tf_sim.df <- receptor.seq_sim.cor.tf_sim.df[ ,-c(1:2)]

# Remove duplicate rows
receptor.seq_sim.cor.tf_sim.df <- receptor.seq_sim.cor.tf_sim.df %>% distinct()

saveRDS(receptor.seq_sim.cor.tf_sim.df,file=paste0(object_dir,"receptor.seq_sim.cor.tf_sim.df.rds"))

```

```{r}

figS_9A <- ggplot(data=receptor.seq_sim.cor.tf_sim.df, aes(x=vr_seq_sim, y=tf_exp_cor)) + 
  geom_point(size=0.5) + theme_classic() +
  geom_pointdensity() + scale_color_viridis() + 
  labs(y="Transcription Factor Similarity",x="VR Similarity") +
  stat_summary_bin(fun.y='mean', bins=100, color='orange', 
                   geom='line', orientation='x', size=1.5, show.legend=TRUE) + 
  stat_summary_bin(fun.y='median', bins=100, color = 'blue', 
                   geom='line', linetype='dashed', orientation='x', size=1.5, show.legend=TRUE) +
  stat_summary_bin(fun.y='mean', bins=100, 
                   color='magenta', geom='line', orientation='y', size=1.5, show.legend=TRUE) + 
  stat_summary_bin(fun.y='median', bins=100, color = 'cyan', 
                   geom='line', linetype='dashed', orientation='y', size=1.5, show.legend=TRUE) +
  theme(legend.key.size =  unit(0.5, "in"),axis.title=element_text(size=24),
        axis.text=element_text(size=20),legend.title=element_text(size=20),
        legend.text=element_text(size=20), aspect.ratio=1)
  
```

```{r, fig.width=10, fig.height=10}

svglite(filename = paste0(adobe_images,"figS_9A.svg"), width=10, height=10)
figS_9A
invisible(dev.off())
figS_9A

```


```{r,fig.width=8,fig.height=16}

### Supplemental - V1R-specific Average Transcription Factor Expression  
# The dendrogram for rows
dst <- dist(v1r.transcription_factor.b_in_a.mat)
hc_row <- hclust(dst)
row_dend <- as.dendrogram(hc_row)
row_dend <- seriate_dendrogram(row_dend, dst, method="OLO")

# The dendrogram for columns
dst <- dist(t(v1r.transcription_factor.b_in_a.mat))
hc_row <- hclust(dst)
col_dend <- as.dendrogram(hc_row)
col_dend <- seriate_dendrogram(col_dend, dst, method="OLO")

figS_9B <- pheatmap(v1r.transcription_factor.b_in_a.mat,fontsize=15,treeheight_row=25,treeheight_col=25,
                    cluster_rows=as.hclust(row_dend), cluster_cols=as.hclust(col_dend))

# Save image to svg file
svglite(filename = paste0(adobe_images,"figS_9B.svg"),width=8,height=16)
grid.draw(rectGrob(gp=gpar(fill="white", lwd=0)))
grid.draw(figS_9B[[4]])
invisible(dev.off())

```


```{r,fig.height=15,fig.width=15}

### Supplemental - V2R-specific Average Transcription Factor Expression  
# The dendrogram for rows
dst <- dist(v2r.transcription_factor.b_in_a.mat)
hc_row <- hclust(dst)
row_dend <- as.dendrogram(hc_row)
row_dend <- seriate_dendrogram(row_dend, dst, method="OLO")

# The dendrogram for columns
dst <- dist(t(v2r.transcription_factor.b_in_a.mat))
hc_row <- hclust(dst)
col_dend <- as.dendrogram(hc_row)
col_dend <- seriate_dendrogram(col_dend, dst, method="OLO")

figS_9C <- pheatmap(v2r.transcription_factor.b_in_a.mat,fontsize=12,treeheight_row=25,treeheight_col=25,
                    cluster_rows=as.hclust(row_dend), cluster_cols=as.hclust(col_dend))

# Save image to svg file
svglite(filename = paste0(adobe_images,"figS_9C.svg"),width=15,height=15)
grid.draw(rectGrob(gp=gpar(fill="white", lwd=0)))
grid.draw(figS_9C[[4]])
invisible(dev.off())

```


```{r}
# {-}  

current.variables <- c(ls(),"current.variables")
rm(list=setdiff(current.variables,c(global.vars)))
invisible(gc())

```



```{r}
# Figure 5 - Receptor Co-expression {.tabset .tabset-pills}  

## A - alpha-diversity  

alpha.diversity.df <- FetchData(neuron.integrated,vars=c("age","cell_type","Shannon.Index"))

```

```{r}

fig5_a <- ggplot(alpha.diversity.df,aes(x=cell_type,y=Shannon.Index,fill=cell_type)) + geom_boxplot() +
  scale_fill_manual("Cell Type",values=neuron.cell_type.colors) + #theme_classic() + 
  stat_summary(fun.y=mean,geom='point',size=10,colour="white",shape=95,show_guide=FALSE) +
  facet_grid(. ~ age) +
  # stat_compare_means(comparisons=alpha_comparisons.list,tip.length=0.01,
  #                    method="wilcox.test",label="p.signif",vjust=0.3) +
  ylab("Shannon Index") + xlab("Cell Type") + theme_bw() +
  theme(aspect.ratio=2, plot.title=element_text(size=32,face="bold"),
        axis.text.x=element_text(angle=45,vjust=1,hjust=1),
        axis.text=element_text(size=20), axis.title=element_text(size=24), axis.ticks.x=element_blank(),
        strip.text.x = element_text(size = 24)) + 
  NoLegend() 

```

```{r}

svglite(filename = paste0(adobe_images,"fig5_a.svg"), width=10, height=10)
fig5_a
invisible(dev.off())

```

```{r, fig.width=10, fig.height=10}

fig5_a

```



```{r, include=FALSE}
## B - V1R Ratio of Cells w/ Co-Expression  

receptor.metadata <- FetchData(neuron.integrated,
                               vars=c("cell_type","age","first.receptor.count","receptor.species"))

```

```{r}

V1R.receptor.metadata <- receptor.metadata[receptor.metadata$cell_type %in% c("late_INP","V1R_iVSN","V1R_mVSN"), ]

```

```{r,include=FALSE}

##### V1R #####
# Group data by age/cell_type combinations, and how many N cells 
V1R.co_exp.cell_type <- 
   summarySE(V1R.receptor.metadata, 
             measurevar = "first.receptor.count", 
             groupvars = c("age","cell_type"))

V1R.co_exp.cell_type$None <- unlist(lapply(rownames(V1R.co_exp.cell_type), function(i){
   age <- V1R.co_exp.cell_type[i, "age"]
   cell_type <- V1R.co_exp.cell_type[i, "cell_type"]
   nrow(V1R.receptor.metadata[V1R.receptor.metadata$age == age & 
                            V1R.receptor.metadata$cell_type == cell_type &
                            V1R.receptor.metadata$receptor.species == 0, ])/V1R.co_exp.cell_type[i, "N"]
}))

V1R.co_exp.cell_type$One <- unlist(lapply(rownames(V1R.co_exp.cell_type), function(i){
   age <- V1R.co_exp.cell_type[i, "age"]
   cell_type <- V1R.co_exp.cell_type[i, "cell_type"]
   nrow(V1R.receptor.metadata[V1R.receptor.metadata$age == age & 
                            V1R.receptor.metadata$cell_type == cell_type &
                            V1R.receptor.metadata$receptor.species == 1, ])/V1R.co_exp.cell_type[i, "N"]
}))

V1R.co_exp.cell_type$Two <- unlist(lapply(rownames(V1R.co_exp.cell_type), function(i){
   age <- V1R.co_exp.cell_type[i, "age"]
   cell_type <- V1R.co_exp.cell_type[i, "cell_type"]
   nrow(V1R.receptor.metadata[V1R.receptor.metadata$age == age & 
                            V1R.receptor.metadata$cell_type == cell_type &
                            V1R.receptor.metadata$receptor.species == 2, ])/V1R.co_exp.cell_type[i, "N"]
}))

V1R.co_exp.cell_type$Three_or_more <- unlist(lapply(rownames(V1R.co_exp.cell_type), function(i){
   age <- V1R.co_exp.cell_type[i, "age"]
   cell_type <- V1R.co_exp.cell_type[i, "cell_type"]
   nrow(V1R.receptor.metadata[V1R.receptor.metadata$age == age & 
                            V1R.receptor.metadata$cell_type == cell_type &
                            V1R.receptor.metadata$receptor.species >= 3, ])/V1R.co_exp.cell_type[i, "N"]
}))

```

```{r,include=FALSE}

V1R.plot.df <- V1R.co_exp.cell_type %>% gather(Receptors, Ratio, c("Three_or_more","Two","One","None"))

```

```{r,include=FALSE}

## Order the factors and round the floats in the DF  
##### V1R #####
# Order the Receptors factor levels for plotting
V1R.plot.df$Receptors <-
  factor(as.factor(V1R.plot.df$Receptors), 
         levels = c("Three_or_more","Two","One","None"))

# Order the age factor levels for plotting
V1R.plot.df$age <-
  factor(as.factor(V1R.plot.df$age), 
         levels = c("P14","P56"))

# Order the cell_type factor levels for plotting
V1R.plot.df$cell_type <-
  factor(as.factor(V1R.plot.df$cell_type), 
         levels = c("late_INP","V1R_iVSN","V1R_mVSN"))

# Round the ratio to 4 digits
V1R.plot.df$Ratio <- round(V1R.plot.df$Ratio, digits = 4)

```

```{r,include=FALSE}

fig5_b <- ggplot(V1R.plot.df, aes(x = age, y = Ratio, fill = Receptors, label = Ratio)) + 
  geom_col() + theme(aspect.ratio=3,text=element_text(size=24),legend.position="top") +
  facet_grid(. ~ cell_type) +
  geom_text(data = subset(V1R.plot.df, Ratio > 0), size=5, position=position_stack(vjust = 0.5)) +
  geom_text(data = subset(V1R.plot.df, Receptors == "None"), size=5, aes(x=age, y=-0.01, label=paste0("N=",N)))

svglite(filename = paste0(adobe_images,"fig5_b.svg"),width=10,height=10)
fig5_b
invisible(dev.off())

```

```{r,fig.width=10,fig.height=10}

fig5_b

```


```{r}
## C - V2R Ratio of Cells w/ Co-Expression  

V2R.receptor.metadata <- receptor.metadata[receptor.metadata$cell_type %in% c("late_INP","V2R_iVSN","V2R_mVSN"), ]

```

```{r}

##### V2R #####

V2R.co_exp.cell_type <- 
   summarySE(V2R.receptor.metadata, 
             measurevar = "first.receptor.count", 
             groupvars = c("age","cell_type"))

V2R.co_exp.cell_type$None <- unlist(lapply(rownames(V2R.co_exp.cell_type), function(i){
   age <- V2R.co_exp.cell_type[i, "age"]
   cell_type <- V2R.co_exp.cell_type[i, "cell_type"]
   nrow(V2R.receptor.metadata[V2R.receptor.metadata$age == age & 
                            V2R.receptor.metadata$cell_type == cell_type &
                            V2R.receptor.metadata$receptor.species == 0, ])/V2R.co_exp.cell_type[i, "N"]
}))

V2R.co_exp.cell_type$One <- unlist(lapply(rownames(V2R.co_exp.cell_type), function(i){
   age <- V2R.co_exp.cell_type[i, "age"]
   cell_type <- V2R.co_exp.cell_type[i, "cell_type"]
   nrow(V2R.receptor.metadata[V2R.receptor.metadata$age == age & 
                            V2R.receptor.metadata$cell_type == cell_type &
                            V2R.receptor.metadata$receptor.species == 1, ])/V2R.co_exp.cell_type[i, "N"]
}))

V2R.co_exp.cell_type$Two <- unlist(lapply(rownames(V2R.co_exp.cell_type), function(i){
   age <- V2R.co_exp.cell_type[i, "age"]
   cell_type <- V2R.co_exp.cell_type[i, "cell_type"]
   nrow(V2R.receptor.metadata[V2R.receptor.metadata$age == age & 
                            V2R.receptor.metadata$cell_type == cell_type &
                            V2R.receptor.metadata$receptor.species == 2, ])/V2R.co_exp.cell_type[i, "N"]
}))

V2R.co_exp.cell_type$Three_or_more <- unlist(lapply(rownames(V2R.co_exp.cell_type), function(i){
   age <- V2R.co_exp.cell_type[i, "age"]
   cell_type <- V2R.co_exp.cell_type[i, "cell_type"]
   nrow(V2R.receptor.metadata[V2R.receptor.metadata$age == age & 
                            V2R.receptor.metadata$cell_type == cell_type &
                            V2R.receptor.metadata$receptor.species >= 3, ])/V2R.co_exp.cell_type[i, "N"]
}))

```

```{r}

V2R.plot.df <- V2R.co_exp.cell_type %>% gather(Receptors, Ratio, c("Three_or_more","Two","One","None"))

```

```{r}

##### V2R #####

# Order the Receptors factor levels for plotting
V2R.plot.df$Receptors <-
  factor(as.factor(V2R.plot.df$Receptors), 
         levels = c("Three_or_more","Two","One","None"))

# Order the age factor levels for plotting
V2R.plot.df$age <-
  factor(as.factor(V2R.plot.df$age), 
         levels = c("P14","P56"))

# Order the cell_type factor levels for plotting
V2R.plot.df$cell_type <-
  factor(as.factor(V2R.plot.df$cell_type), 
         levels = c("early_INP","late_INP","V2R_iVSN","V2R_mVSN"))

# Round the ratio to 4 digits
V2R.plot.df$Ratio <- round(V2R.plot.df$Ratio, digits = 4)

```

```{r, fig.width=16, fig.height=9,include=FALSE}

fig5_c <- ggplot(V2R.plot.df, aes(x = age, y = Ratio, fill = Receptors, label = Ratio)) +
  geom_col() + theme(aspect.ratio=3,text=element_text(size=24),legend.position="top") +
  facet_grid(. ~ cell_type) +
  geom_text(data = subset(V2R.plot.df, Ratio > 0), size=5, position = position_stack(vjust = 0.5)) +
  geom_text(data = subset(V2R.plot.df, Receptors == "None"), size=5, aes(x=age, y=-0.01, label=paste0("N=",N)))

svglite(filename = paste0(adobe_images,"fig5_c.svg"),width=10,height=10)
fig5_c
invisible(dev.off())

```

```{r}

fig5_c

```


```{r}
## D - Coexpressing V1 Receptors  

load(file=paste0(object_dir, "sig.v1r.coexpress"))
load(file=paste0(object_dir, "v1r.circos.bed"))
load(file=paste0(object_dir, "v1r.bed1"))
load(file=paste0(object_dir, "v1r.bed2"))
load(file=paste0(object_dir, "sig.v2r.coexpress"))
load(file=paste0(object_dir, "v2r.circos.bed"))
load(file=paste0(object_dir, "v2r.bed1"))
load(file=paste0(object_dir, "v2r.bed2"))
load(file=paste0(object_dir, "sig.interclass.coexpress"))
load(file=paste0(object_dir, "interclass.circos.bed"))
load(file=paste0(object_dir, "interclass.bed1"))
load(file=paste0(object_dir, "interclass.bed2"))

```


```{r}

v1r.sectors <- lapply(unique(v1r.circos.bed$chr), function(chr){
  padding <- 100000
  sector.start <- min(v1r.circos.bed[v1r.circos.bed$chr == chr, "start"]) - padding
  sector.end <- max(v1r.circos.bed[v1r.circos.bed$chr == chr, "end"]) + padding
  return(c(chr,sector.start,sector.end))
})

v1r.chr.coord <- as.data.frame(do.call(rbind, v1r.sectors))
colnames(v1r.chr.coord) <- c("chr","start","end")

major.tix <- seq(from=0,to=250000000,by=100000)

```


```{r, fig.width=40,fig.height=40}

svglite(filename = paste0(adobe_images,"fig5_d.svg"),width=25,height=25)

font.scale <- 7

circos.par(start.degree=90, gap.degree=2, track.margin=c(0.0005,0.0005))
circos.genomicInitialize(v1r.chr.coord,plotType = NULL)
# Add the geneSymbol labels
circos.genomicLabels(v1r.circos.bed, labels.column=4, side="outside", labels_height=0.29, cex=font.scale,
                     padding=0.04, connection_height=mm_h(15), line_lwd=5)

# Draw the sectors and add chr ID as text
circos.track(ylim=c(0,1), panel.fun=function(x,y){
  chr=CELL_META$sector.index
  xlim=CELL_META$xlim
  ylim=CELL_META$ylim
  circos.rect(xlim[1], 0, xlim[2], 1, col="darkred")
  circos.text(mean(xlim), mean(ylim), chr, cex=font.scale, col="white", facing="clockwise", niceFacing=TRUE)
  circos.xaxis(h=1, labels=FALSE, direction="outside", minor.ticks=1,
               major.tick.length=0.05, major.at=major.tix, lwd=5)
  }, track.height=0.22, bg.border=NA)
circos.genomicLink(v1r.bed1, v1r.bed2, col=sig.v1r.coexpress$color, lwd=5, border=NA)
invisible(dev.off())

####################################################

circos.par(start.degree=90, gap.degree=2, track.margin=c(0.0005,0.0005))
circos.genomicInitialize(v1r.chr.coord,plotType = NULL)
# Add the geneSymbol labels
circos.genomicLabels(v1r.circos.bed, labels.column=4, side="outside", labels_height=0.29, cex=font.scale,
                     padding=0.04, connection_height=mm_h(15), line_lwd=5)

# Draw the sectors and add chr ID as text
circos.track(ylim=c(0,1), panel.fun=function(x,y){
  chr=CELL_META$sector.index
  xlim=CELL_META$xlim
  ylim=CELL_META$ylim
  circos.rect(xlim[1], 0, xlim[2], 1, col="darkred")
  circos.text(mean(xlim), mean(ylim), chr, cex=font.scale, col="white", facing="clockwise", niceFacing=TRUE)
  circos.xaxis(h=1, labels=FALSE, direction="outside", minor.ticks=1,
               major.tick.length=0.05, major.at=major.tix, lwd=5)
  }, track.height=0.22, bg.border=NA)
circos.genomicLink(v1r.bed1, v1r.bed2, col=sig.v1r.coexpress$color, lwd=5, border=NA)

```



```{r}
## E - Coexpressing V2 Receptors  

v2r.sectors <- lapply(unique(v2r.circos.bed$chr), function(chr){
  padding <- 100000
  sector.start <- min(v2r.circos.bed[v2r.circos.bed$chr == chr, "start"]) - padding
  sector.end <- max(v2r.circos.bed[v2r.circos.bed$chr == chr, "end"]) + padding
  return(c(chr,sector.start,sector.end))
})

v2r.chr.coord <- as.data.frame(do.call(rbind, v2r.sectors))
colnames(v2r.chr.coord) <- c("chr","start","end")

cols.num <- c("start","end")
v2r.chr.coord[cols.num] <- sapply(v2r.chr.coord[cols.num],as.numeric)

```


```{r, fig.width=50,fig.height=50}

svglite(filename = paste0(adobe_images,"fig5_e.svg"),width=50,height=50)

font.scale <- 7

circos.par(start.degree=90, gap.degree=2, track.margin=c(0.0005,0.0005))
circos.genomicInitialize(v2r.chr.coord,plotType = NULL)
# Add the geneSymbol labels
circos.genomicLabels(v2r.circos.bed, labels.column=4, side="outside", labels_height=0.29, cex=font.scale,
                     padding=0.04, connection_height=mm_h(15), line_lwd=5)

# Draw the sectors and add chr ID as text
circos.track(ylim=c(0,1), panel.fun=function(x,y){
  chr=CELL_META$sector.index
  xlim=CELL_META$xlim
  ylim=CELL_META$ylim
  circos.rect(xlim[1], 0, xlim[2], 1, col="darkblue")
  circos.text(mean(xlim), mean(ylim), chr, cex=font.scale, col="white", facing="clockwise", niceFacing=TRUE)
  circos.xaxis(h=1, labels=FALSE, direction="outside", minor.ticks=1,
               major.tick.length=0.05, major.at=major.tix, lwd=5)
  }, track.height=0.22, bg.border=NA)
circos.genomicLink(v2r.bed1, v2r.bed2, col=sig.v2r.coexpress$color, lwd=5, border=NA)
invisible(dev.off())

####################################################

circos.par(start.degree=90, gap.degree=2, track.margin=c(0.0005,0.0005))
circos.genomicInitialize(v2r.chr.coord,plotType = NULL)
# Add the geneSymbol labels
circos.genomicLabels(v2r.circos.bed, labels.column=4, side="outside", labels_height=0.29, cex=font.scale,
                     padding=0.04, connection_height=mm_h(15), line_lwd=5)

# Draw the sectors and add chr ID as text
circos.track(ylim=c(0,1), panel.fun=function(x,y){
  chr=CELL_META$sector.index
  xlim=CELL_META$xlim
  ylim=CELL_META$ylim
  circos.rect(xlim[1], 0, xlim[2], 1, col="darkblue")
  circos.text(mean(xlim), mean(ylim), chr, cex=font.scale, col="white", facing="clockwise", niceFacing=TRUE)
  circos.xaxis(h=1, labels=FALSE, direction="outside", minor.ticks=1,
               major.tick.length=0.05, major.at=major.tix, lwd=5)
  }, track.height=0.22, bg.border=NA)
circos.genomicLink(v2r.bed1, v2r.bed2, col=sig.v2r.coexpress$color, lwd=5, border=NA)

```


```{r}
## F - Coexpressing Interclass Receptors  

interclass.sectors <- lapply(unique(interclass.circos.bed$chr), function(chr){
  padding <- 100000
  sector.start <- min(interclass.circos.bed[interclass.circos.bed$chr == chr, "start"]) - padding
  sector.end <- max(interclass.circos.bed[interclass.circos.bed$chr == chr, "end"]) + padding
  return(c(chr,sector.start,sector.end))
})

interclass.chr.coord <- as.data.frame(do.call(rbind, interclass.sectors))
colnames(interclass.chr.coord) <- c("chr","start","end")

major.tix <- seq(from=0,to=250000000,by=100000)

```


```{r, fig.width=50,fig.height=50}

svglite(filename = paste0(adobe_images,"fig5_f.svg"),width=50,height=50)

font.scale <- 7

circos.par(start.degree=90, gap.degree=2, track.margin=c(0.0005,0.0005))
circos.genomicInitialize(interclass.chr.coord,plotType = NULL)
# Add the geneSymbol labels
circos.genomicLabels(interclass.circos.bed, labels.column=4, side="outside", labels_height=0.29, cex=font.scale,
                     padding=0.04, connection_height=mm_h(15), line_lwd=5)

# Draw the sectors and add chr ID as text
circos.track(ylim=c(0,1), panel.fun=function(x,y){
  chr=CELL_META$sector.index
  xlim=CELL_META$xlim
  ylim=CELL_META$ylim
  circos.rect(xlim[1], 0, xlim[2], 1, col="darkgreen")
  circos.text(mean(xlim), mean(ylim), chr, cex=font.scale, col="white", facing="clockwise", niceFacing=TRUE)
  circos.xaxis(h=1, labels=FALSE, direction="outside", minor.ticks=1,
               major.tick.length=0.05, major.at=major.tix, lwd=5)
  }, track.height=0.22, bg.border=NA)
circos.genomicLink(interclass.bed1, interclass.bed2, col=sig.interclass.coexpress$color, lwd=5, border=NA)
invisible(dev.off())

####################################################

circos.par(start.degree=90, gap.degree=2, track.margin=c(0.0005,0.0005))
circos.genomicInitialize(interclass.chr.coord,plotType = NULL)
# Add the geneSymbol labels
circos.genomicLabels(interclass.circos.bed, labels.column=4, side="outside", labels_height=0.29, cex=font.scale,
                     padding=0.04, connection_height=mm_h(15), line_lwd=5)

# Draw the sectors and add chr ID as text
circos.track(ylim=c(0,1), panel.fun=function(x,y){
  chr=CELL_META$sector.index
  xlim=CELL_META$xlim
  ylim=CELL_META$ylim
  circos.rect(xlim[1], 0, xlim[2], 1, col="darkgreen")
  circos.text(mean(xlim), mean(ylim), chr, cex=font.scale, col="white", facing="clockwise", niceFacing=TRUE)
  circos.xaxis(h=1, labels=FALSE, direction="outside", minor.ticks=1,
               major.tick.length=0.05, major.at=major.tix, lwd=5)
  }, track.height=0.22, bg.border=NA)
circos.genomicLink(interclass.bed1, interclass.bed2, col=sig.interclass.coexpress$color, lwd=5, border=NA)

```


```{r}
## Do receptors that overlap always come from the same chromosome?  

sig.mVSN.co <- openxlsx::read.xlsx(xlsxFile=paste0(object_dir,"sig.mVSN.co.xlsx"))

```


```{r}
### What is the percentage of significant co-expressing receptor pairs that are from the same chromosome?  

paste0(round(sum(sig.mVSN.co$A_chromosome==sig.mVSN.co$B_chromosome)/nrow(sig.mVSN.co)*100,2),"%")

```

 

```{r}
### At the cell level, what is the percentage of cells with significant co-expressing receptor pairs from the same chromosome? 
# Total number of cells containing co-expressed receptors from the same chromosome
n.same <- sum(sig.mVSN.co[sig.mVSN.co$A_chromosome==sig.mVSN.co$B_chromosome, "AB.intersect"])
# Total number of cells containing co-expressed receptors from the same chromosome
n.diff <- sum(sig.mVSN.co[sig.mVSN.co$A_chromosome!=sig.mVSN.co$B_chromosome, "AB.intersect"])

paste0(round(n.same/(n.same+n.diff)*100,2),"%")

```


```{r}
### Exact Binomial Test: Are 100% of Co-expressed Pairs from the same chromosome?  

binom.test(x=c(n.same,n.diff),p=1,alternative="less",conf.level=0.95)

```



```{r}
## G-J  Resolve Receptor Overlap  

> These panels were produced by Thelma Chiremba.  

# {-}  

current.variables <- c(ls(),"current.variables")
rm(list=setdiff(current.variables,c(global.vars)))
invisible(gc())

```


```{r}

# Figure 6 - Vomeronasal Receptors and Plasma Membrane Expressed Axon-Guidance Molecules {.tabset .tabset-pills}  

#Load files generated in ```axon_guidance.Rmd```  

vr.cor.membrane.ag <- readRDS(file=paste0(object_dir,"vr.cor.membrane.ag.rds")) 
# Data frame with receptor sequence similarity scores and receptor average axon-guidance gene expression  
receptor.seq_sim.cor.ag_sim.df <- readRDS(file=paste0(object_dir,"receptor.seq_sim.cor.ag_sim.df.rds"))
# All significant VRs overlapping with surface expressed axon-guidance genes 
membrane.ag.b_in_a.mat <- readRDS(file=paste0(object_dir,"membrane.ag.b_in_a.mat.rds"))
# The above subset for strong groupings and split into V1R, V2R, and mixed-class matrices
v1r.membrane.ag.b_in_a.mat <- readRDS(file=paste0(object_dir,"v1r.membrane.ag.b_in_a.mat.rds"))
v2r.membrane.ag.b_in_a.mat <- readRDS(file=paste0(object_dir,"v2r.membrane.ag.b_in_a.mat.rds"))

```


```{r,fig.height=10,fig.width=10}

## A - Vomeronasal Receptor Correlation of Surface Axon-Guidance Gene Expression  

membrane.ag.cluster.n <- 20

fig6_a <-
  pheatmap(vr.cor.membrane.ag,fontsize=20,show_rownames=FALSE,show_colnames=FALSE,
           cutree_rows=membrane.ag.cluster.n,cutree_cols=membrane.ag.cluster.n)

# Save image to svg file
svglite(filename = paste0(adobe_images,"fig6_a.svg"),height=10,width=10)
grid.draw(rectGrob(gp=gpar(fill="white", lwd=0)))
grid.draw(fig6_a[[4]])
invisible(dev.off())

```


```{r}

## B - Scatter Plot: (VR Sequence Similarity) vs (VR-Specific Average-Axon-Guidance Gene Expression Correlation)  

fig6_b <- ggplot(data=receptor.seq_sim.cor.ag_sim.df, aes(x=vr_seq_sim, y=pm_ag_exp_cor)) + 
  geom_point(size=0.5) + theme_classic() +
  geom_pointdensity() + scale_color_viridis() + 
  labs(y="Guidance Molecule Similarity",x="VR Similarity") +
  stat_summary_bin(fun.y='mean', bins=100, color='orange', 
                   geom='line', orientation='x', size=1.5, show.legend=TRUE) + 
  stat_summary_bin(fun.y='median', bins=100, color = 'dodgerblue', 
                   geom='line', linetype='dashed', orientation='x', size=1.5, show.legend=TRUE) +
  stat_summary_bin(fun.y='mean', bins=100, 
                   color='magenta', geom='line', orientation='y', size=1.5, show.legend=TRUE) + 
  stat_summary_bin(fun.y='median', bins=100, color = 'cyan', 
                   geom='line', linetype='dashed', orientation='y', size=1.5, show.legend=TRUE) +
  theme(legend.position="left",aspect.ratio=1,axis.text=element_text(size=20),legend.text=element_text(size=24),
        axis.title=element_text(size=32),legend.title=element_text(size=24),panel.border=element_blank(),
        plot.background=element_blank())
  
```

```{r, fig.width=10, fig.height=10}

svglite(filename = paste0(adobe_images,"fig6_b.svg"), width=10, height=10)
fig6_b
invisible(dev.off())
fig6_b

```


```{r}
dim(v1r.membrane.ag.b_in_a.mat)
```

```{r,fig.height=19,fig.width=10}

## C - V1R Average Surface Axon-Guidance Gene Expression  

# The dendrogram for rows
dst <- dist(v1r.membrane.ag.b_in_a.mat)
hc_row <- hclust(dst)
row_dend <- as.dendrogram(hc_row)
row_dend <- seriate_dendrogram(row_dend, dst, method="OLO")

# The dendrogram for columns
dst <- dist(t(v1r.membrane.ag.b_in_a.mat))
hc_row <- hclust(dst)
col_dend <- as.dendrogram(hc_row)
col_dend <- seriate_dendrogram(col_dend, dst, method="OLO")

fig6_c <- pheatmap(v1r.membrane.ag.b_in_a.mat,fontsize=10,treeheight_row=25,treeheight_col=25,
                   cluster_rows=as.hclust(row_dend), cluster_cols=as.hclust(col_dend))
# Save image to svg file
svglite(filename = paste0(adobe_images,"fig6_c.svg"),width=5,height=9.5)
grid.draw(rectGrob(gp=gpar(fill="white", lwd=0)))
grid.draw(fig6_c[[4]])
invisible(dev.off())

```



```{r}
## D - V2R Average Surface Axon-Guidance Gene Expression  
dim(v2r.membrane.ag.b_in_a.mat)
```

```{r,fig.height=15,fig.width=7.5}

# The dendrogram for rows
dst <- dist(v2r.membrane.ag.b_in_a.mat)
hc_row <- hclust(dst)
row_dend <- as.dendrogram(hc_row)
row_dend <- seriate_dendrogram(row_dend, dst, method="OLO")

# The dendrogram for columns
dst <- dist(t(v2r.membrane.ag.b_in_a.mat))
hc_row <- hclust(dst)
col_dend <- as.dendrogram(hc_row)
col_dend <- seriate_dendrogram(col_dend, dst, method="OLO")

fig6_d <- pheatmap(v2r.membrane.ag.b_in_a.mat,cluster_rows=as.hclust(row_dend),
                   cluster_cols=as.hclust(col_dend),fontsize=10,treeheight_row=25,
                   treeheight_col=25)
# Save image to svg file
svglite(filename = paste0(adobe_images,"fig6_d.svg"),height=15,width=7.5)
grid.draw(rectGrob(gp=gpar(fill="white", lwd=0)))
grid.draw(fig6_d[[4]])
invisible(dev.off())

```



```{r}
# Supplemental Figure 6.1  
dim(membrane.ag.b_in_a.mat)

```

```{r}

membrane.ag.b_in_a.mat.filtered <- membrane.ag.b_in_a.mat[grep("^Vmn1r|^Vmn2r", rownames(membrane.ag.b_in_a.mat), value=TRUE), ]
membrane.ag.b_in_a.mat.filtered <- membrane.ag.b_in_a.mat.filtered[ ,colSums(membrane.ag.b_in_a.mat.filtered) > 0]

```

```{r}

dim(membrane.ag.b_in_a.mat.filtered)

```

```{r fig.width=8, fig.height=20}

# The dendrogram for rows
dst <- dist(membrane.ag.b_in_a.mat.filtered)
hc_row <- hclust(dst)
row_dend <- as.dendrogram(hc_row)
row_dend <- seriate_dendrogram(row_dend, dst, method="OLO")

# The dendrogram for columns
dst <- dist(t(membrane.ag.b_in_a.mat.filtered))
hc_row <- hclust(dst)
col_dend <- as.dendrogram(hc_row)
col_dend <- seriate_dendrogram(col_dend, dst, method="OLO")

fig6_supplemental_1 <- pheatmap(membrane.ag.b_in_a.mat.filtered,fontsize=10,treeheight_row=25,treeheight_col=25,
                                cluster_rows=as.hclust(row_dend), cluster_cols=as.hclust(col_dend))
# Save image to svg file
pdf(file=paste0(adobe_images,"fig6_supplemental_1.pdf"),height=20,width=8)
grid.draw(fig6_supplemental_1[[4]])
invisible(dev.off())

```

 

```{r}
# Supplemental Figure 6.2 - VR-specific Plasma Membrane Axon Guidance Gene Expression Correlation 
# Get the transpose matrix for average ag gene expression in a VR cell
vr.cor.membrane.ag.df <- t(membrane.ag.b_in_a.mat)
# Get the column names for the transpose matrix
vr.cor.membrane.ag.names <- rownames(membrane.ag.b_in_a.mat)
colnames(vr.cor.membrane.ag.df) <- vr.cor.membrane.ag.names

vr.cor.membrane.ag <- cor(vr.cor.membrane.ag.df, method="pearson")
rm(vr.cor.membrane.ag.names,vr.cor.membrane.ag.df)

```


```{r,fig.height=36,fig.width=36}

membrane.ag.cluster.n <- 25

fig6_supplemental_2 <- pheatmap(vr.cor.membrane.ag,fontsize=12,cutree_rows=membrane.ag.cluster.n,cutree_cols=membrane.ag.cluster.n,
                                treeheight_row=25,treeheight_col=25)
# Save image to svg file
pdf(file=paste0(adobe_images,"fig6_supplemental_2.pdf"),height=36,width=36)
grid.draw(rectGrob(gp=gpar(fill="white", lwd=0)))
grid.draw(fig6_supplemental_2[[4]])
invisible(dev.off())

```



```{r}

# Supplemental Figure 6.3 - Plasma Membrane Axon Guidance Gene VR Expression Correlation  
membrane.ag.cor.vr <- cor(membrane.ag.b_in_a.mat)

```

```{r,fig.height=15,fig.width=15}

membrane.ag.cluster.n <- 20

fig6_supplemental_3 <- pheatmap(membrane.ag.cor.vr,fontsize=12,cutree_rows=membrane.ag.cluster.n,cutree_cols=membrane.ag.cluster.n,
                                treeheight_row=25,treeheight_col=25)
# Save image to svg file
pdf(file=paste0(adobe_images,"fig6_supplemental_3.pdf"),height=15,width=15)
grid.draw(rectGrob(gp=gpar(fill="white", lwd=0)))
grid.draw(fig6_supplemental_3[[4]])
invisible(dev.off())

```



```{r}
# {-}  
current.variables <- c(ls(),"current.variables")
rm(list=setdiff(current.variables,c(global.vars)))
invisible(gc())

```


```{r}

# Figure 7 - Axon-Guidance / Transcription Factor Gene Co-expression with Receptors  

#Load the data we will need to plot panels for Figure 7.  

# Load the significant co-expression data frames
sig.vr.coexpression <- openxlsx::read.xlsx(xlsxFile=paste0(object_dir,"sig.mVSN.co.xlsx"))
rownames(sig.vr.coexpression) <- paste(sig.vr.coexpression$setA,sig.vr.coexpression$setB,sep="_")
sig.vr.ag.coexpression <- openxlsx::read.xlsx(xlsxFile=paste0(object_dir,"sig.vr.ag.coexpression.xlsx"))
rownames(sig.vr.ag.coexpression) <- paste(sig.vr.ag.coexpression$setA,sig.vr.ag.coexpression$setB,sep="_")
sig.vr.tf.coexpression <- openxlsx::read.xlsx(xlsxFile=paste0(object_dir,"sig.vr.tf.coexpression.xlsx"))
rownames(sig.vr.tf.coexpression) <- paste(sig.vr.tf.coexpression$setA,sig.vr.tf.coexpression$setB,sep="_")
vr.ag_tf.results <- openxlsx::read.xlsx(xlsxFile=paste0(object_dir,"vr.ag_tf.results.xlsx"))
rownames(vr.ag_tf.results) <- paste(vr.ag_tf.results$setA,vr.ag_tf.results$setB,sep="_")

# Load the gene names for each category
vr.names <- readRDS(file=paste0(object_dir,"final.receptor.names.rds"))
ag.names <- readRDS(file=paste0(object_dir,"final.membrane_ag.names.rds"))
tf.names <- readRDS(file=paste0(object_dir,"final.dna_binding_tf.names.rds"))

```

```{r}

VRs <- unique(sort(vr.ag_tf.results$VR))
AGs <- unique(sort(vr.ag_tf.results$AG))
TFs <- unique(sort(vr.ag_tf.results$TF))

```

```{r}

# Create a list of matrices with the Jaccard Index for AG-TF co-expression, for each VR
vno.mat.list <- lapply(VRs, function(vr){    # Loop through unique VRs
  # Create an empty matrix for this VR with AGs on columns and TFs on rows
  empty_mat <- matrix(data=0,ncol=length(AGs),nrow=length(TFs))
  rownames(empty_mat) <- TFs
  colnames(empty_mat) <- AGs
  # Subset our results for rows that contain this VR
  sub.df <- vr.ag_tf.results[vr.ag_tf.results$VR==vr, ]
  # For our results data frame, iterate over rows
  for(i in 1:nrow(sub.df)){
    # The AG for this row
    ag.name <- sub.df[i, ]$AG
    # The TF for this row
    tf.name <- sub.df[i, ]$TF
    # The Jaccard Index for this AG and this TF for this VR
    empty_mat[tf.name,ag.name] <- sub.df[i, ]$jaccard.index
  }
  return(empty_mat)    # Return the matrix for this VR
})
# Name matrices for their VR
names(vno.mat.list) <- VRs

```


```{r}
#We will convert all of our VR-specific matrices to vectors, then check Pearson's correlation between VRs

# Convert all matrices to vectors and check their pearson correlation
vno.vec.mat <- sapply(vno.mat.list, as.vector)
vno.cor.mat <- cor(vno.vec.mat)

```



```{r}
#Now we will plot our VR-specific AG-TF co-expression correlation heatmap.
fig7_a.colors <- wesanderson::wes_palette("Darjeeling1")
# Create a row annotation data frame for figure 7A
fig7_a.row_annotations <- sapply(rownames(vno.cor.mat), function(rec){
  if(startsWith(x=rec,prefix="Fpr")){
    return("Fpr")
  } else if(startsWith(x=rec,prefix="Olfr")){
    return("Olfr")
  } else if(startsWith(x=rec,prefix="Vmn1r")){
    return("V1R")
  } else if(startsWith(x=rec,prefix="Vmn2r")){
    return("V2R")
  }
})

fig7_a.row_annotations <- as.data.frame(fig7_a.row_annotations)
colnames(fig7_a.row_annotations) <- "type"

fig7_a.ann_colors <- list(type=c(Fpr=fig7_a.colors[5],
                                 Olfr=fig7_a.colors[2],
                                 V1R=fig7_a.colors[3],
                                 V2R=fig7_a.colors[4]))

```



```{r,fig.width=12,fig.height=12}
## A  
# Plot the cross VR-specific correlation of AGxTF co-expression
fig7_a <-
  pheatmap(vno.cor.mat,fontsize=20,annotation_row=fig7_a.row_annotations,
           show_rownames=FALSE,show_colnames=FALSE,annotation_colors=fig7_a.ann_colors)

```

```{r}

# Save image to svg file
svglite(filename = paste0(adobe_images,"fig7_a.svg"),height=16,width=16)
grid.draw(rectGrob(gp=gpar(fill="white", lwd=0)))
grid.draw(fig7_a[[4]])
invisible(dev.off())

```

 

```{r}
## B 
vr.order.3d <- readRDS(file=paste0(object_dir,"vr.order.3d.rds"))
ag.order.3d <- readRDS(file=paste0(object_dir,"ag.order.3d.rds"))
tf.order.3d <- readRDS(file=paste0(object_dir,"tf.order.3d.rds"))

```

```{r}

# Mimic the color scheme of pheatmap
fig7_palette <- colorRampPalette(rev(RColorBrewer::brewer.pal(n=7, name="RdYlBu")))(100)
# Subset our results data frame
fig7_b.scatter.df <- vr.ag_tf.results[ ,c("VR","AG","TF","jaccard.index")]

```

```{r,include=FALSE}

# Convert categorical variables to factors
fig7_b.scatter.df$AG <- factor(fig7_b.scatter.df$AG, levels = rev(ag.order.3d))
fig7_b.scatter.df$TF <- factor(fig7_b.scatter.df$TF, levels = rev(tf.order.3d))
fig7_b.scatter.df$VR <- factor(fig7_b.scatter.df$VR, levels = rev(vr.order.3d))

# Convert factors to underlying numeric representation
fig7_b.scatter.df$AG_numeric <- as.numeric(fig7_b.scatter.df$AG)
fig7_b.scatter.df$TF_numeric <- as.numeric(fig7_b.scatter.df$TF)
fig7_b.scatter.df$VR_numeric <- as.numeric(fig7_b.scatter.df$VR)

svglite(filename = paste0(adobe_images,"fig7_b.svg"),height=10,width=10)
# Plotting the 3D scatterplot
scatter3D(
  x = fig7_b.scatter.df$AG_numeric,
  y = fig7_b.scatter.df$TF_numeric,
  z = fig7_b.scatter.df$VR_numeric,
  colvar = fig7_b.scatter.df$jaccard.index,  # Color by jaccard.index
  col=fig7_palette,
  pch = 19,  # Point shape
  theta = 45, phi = 30,  # Viewing angle
  xlab = "AG", ylab = "TF", zlab = "VR",  # Axis labels
  clab = "Jaccard Index",
  cex = 0.25    # Shrink the size of the data points
)
invisible(dev.off())

```

```{r,fig.height=10,fig.width=10}

scatter3D(
  x = fig7_b.scatter.df$AG_numeric,
  y = fig7_b.scatter.df$TF_numeric,
  z = fig7_b.scatter.df$VR_numeric,
  colvar = fig7_b.scatter.df$jaccard.index,  # Color by jaccard.index
  col=fig7_palette,
  pch = 19,  # Point shape
  theta = 45, phi = 30,  # Viewing angle
  xlab = "AG", ylab = "TF", zlab = "VR",  # Axis labels
  clab = "Jaccard Index",
  cex = 0.25    # Shrink the size of the data points
)

```



```{r}
## C  
vr.goi <- c("Vmn1r69","Vmn1r185","Vmn1r34","Vmn1r56","Vmn1r60","Vmn1r61","Vmn2r1",
         "Vmn2r3","Vmn2r7","Vmn2r73","Vmn2r88","Vmn2r8","Vmn2r91","Vmn2r118")

ag.goi <- c("Flot1","Lgr4","Pcdh9","Kirrel2","Cdk5r2","Ncam1","Efna5","Nrp2","B3gnt2","Pcdh10",
            "Epha5","Robo2","Pcdh17","Tenm2","Gap43","Kirrel3","Plxnd1","App","Bsg","Efna3")

tf.goi <- c("Trps1","Pbx2","Meis2","Nfix","Tfap2e","Pou2f1","Batf3","Tcf4","Sp8","Cebpb",
            "Nhlh1","Foxg1","Foxp1","Atf5","Jund","Ebf2","Ebf3","Klf7","Emx2","Ikzf4")

filtered.vr.ag_tf.results <- vr.ag_tf.results[vr.ag_tf.results$VR %in% vr.goi &
                                                vr.ag_tf.results$AG %in% ag.goi &
                                                vr.ag_tf.results$TF %in% tf.goi, ]

filtered.vr.ag_tf.results

```

```{r}

vr.order.3d.filtered <- vr.order.3d[vr.order.3d %in% vr.goi]
ag.order.3d.filtered <- ag.order.3d[ag.order.3d %in% ag.goi]
tf.order.3d.filtered <- tf.order.3d[tf.order.3d %in% tf.goi]

```

```{r}

filtered.vno.mat.list <- lapply(vr.order.3d.filtered, function(vr){
  empty_mat <- matrix(data=0,ncol=length(ag.order.3d.filtered),nrow=length(tf.order.3d.filtered))
  rownames(empty_mat) <- tf.order.3d.filtered
  colnames(empty_mat) <- ag.order.3d.filtered
  sub.df <- filtered.vr.ag_tf.results[filtered.vr.ag_tf.results$VR==vr, ]
  for(i in 1:nrow(sub.df)){
    ag.name <- sub.df[i, ]$AG
    tf.name <- sub.df[i, ]$TF
    empty_mat[tf.name,ag.name] <- sub.df[i, ]$jaccard.index
  }
  return(empty_mat)
})
names(filtered.vno.mat.list) <- vr.order.3d.filtered

```

```{r}

# Re-order the matrices by our new ag and tf orders
filtered.vno.mat.list <- lapply(filtered.vno.mat.list, function(mat){
  mat <- mat[tf.order.3d.filtered,ag.order.3d.filtered]
  return(mat)
})

```

```{r}

# Correlated V1Rs (Thelma)
Vmn1r185 <- pheatmap(filtered.vno.mat.list$Vmn1r185, fontsize = 24, cluster_cols = FALSE, 
                     cluster_rows = FALSE, main="Vmn1r185", show_rownames = FALSE, silent=TRUE)
Vmn1r34 <- pheatmap(filtered.vno.mat.list$Vmn1r34, fontsize = 24, cluster_cols = FALSE, 
                    cluster_rows = FALSE, main="Vmn1r34", show_rownames = FALSE, silent=TRUE)
Vmn1r69 <- pheatmap(filtered.vno.mat.list$Vmn1r69, fontsize = 24, cluster_cols = FALSE, 
                    cluster_rows = FALSE, main="Vmn1r69", silent=TRUE)


svglite(filename = paste0(adobe_images,"fig7_c.svg"),width=24,height=8)
grid.arrange(Vmn1r185[[4]], Vmn1r34[[4]], Vmn1r69[[4]], ncol = 3, widths=c(1,1,1.14))
invisible(dev.off())

```

```{r,fig.width=24,fig.height=8}

grid.arrange(Vmn1r185[[4]], Vmn1r34[[4]], Vmn1r69[[4]], ncol = 3, widths=c(1,1,1.14))

```



```{r}
## D  
Vmn1r56 <- pheatmap(filtered.vno.mat.list$Vmn1r56, fontsize = 24, cluster_cols = FALSE, 
                     cluster_rows = FALSE, main="Vmn1r56", show_rownames = FALSE, silent=TRUE)
Vmn1r60 <- pheatmap(filtered.vno.mat.list$Vmn1r60, fontsize = 24, cluster_cols = FALSE, 
                    cluster_rows = FALSE, main="Vmn1r60", show_rownames = FALSE, silent=TRUE)
Vmn1r61 <- pheatmap(filtered.vno.mat.list$Vmn1r61, fontsize = 24, cluster_cols = FALSE, 
                    cluster_rows = FALSE, main="Vmn1r61", silent=TRUE)


svglite(filename = paste0(image_dir,"fig7_d.svg"),width=24,height=8)
grid.arrange(Vmn1r56[[4]], Vmn1r60[[4]], Vmn1r61[[4]], ncol = 3, widths=c(1,1,1.14))
invisible(dev.off())

```

```{r,fig.width=24,fig.height=8}

grid.arrange(Vmn1r56[[4]], Vmn1r60[[4]], Vmn1r61[[4]], ncol = 3, widths=c(1,1,1.14))

```



```{r}
## E  
Vmn2r1 <- pheatmap(filtered.vno.mat.list$Vmn2r1, fontsize = 24, cluster_cols = FALSE, 
                     cluster_rows = FALSE, main="Vmn2r1", show_rownames = FALSE, silent=TRUE)
Vmn2r3 <- pheatmap(filtered.vno.mat.list$Vmn2r3, fontsize = 24, cluster_cols = FALSE, 
                    cluster_rows = FALSE, main="Vmn2r3", show_rownames = FALSE, silent=TRUE)
Vmn2r7 <- pheatmap(filtered.vno.mat.list$Vmn2r7, fontsize = 24, cluster_cols = FALSE, 
                    cluster_rows = FALSE, main="Vmn2r7", silent=TRUE)


svglite(filename = paste0(image_dir,"fig7_e.svg"),width=24,height=8)
grid.arrange(Vmn2r1[[4]], Vmn2r3[[4]], Vmn2r7[[4]], ncol = 3, widths=c(1,1,1.14))
invisible(dev.off())

```

```{r,fig.width=24,fig.height=8}

grid.arrange(Vmn2r1[[4]], Vmn2r3[[4]], Vmn2r7[[4]], ncol = 3, widths=c(1,1,1.14))

```



```{r}
## F  
Vmn2r8 <- pheatmap(filtered.vno.mat.list$Vmn2r8, fontsize = 24, cluster_cols = FALSE, 
                     cluster_rows = FALSE, main="Vmn2r8", show_rownames = FALSE, silent=TRUE)
Vmn2r73 <- pheatmap(filtered.vno.mat.list$Vmn2r73, fontsize = 24, cluster_cols = FALSE, 
                    cluster_rows = FALSE, main="Vmn2r73", show_rownames = FALSE, silent=TRUE)
Vmn2r88 <- pheatmap(filtered.vno.mat.list$Vmn2r88, fontsize = 24, cluster_cols = FALSE, 
                    cluster_rows = FALSE, main="Vmn2r88", silent=TRUE)


svglite(filename = paste0(image_dir,"fig7_f.svg"),width=24,height=8)
grid.arrange(Vmn2r8[[4]], Vmn2r73[[4]], Vmn2r88[[4]], ncol = 3, widths=c(1,1,1.14))
invisible(dev.off())

```

```{r,fig.width=24,fig.height=8}

grid.arrange(Vmn2r8[[4]], Vmn2r73[[4]], Vmn2r88[[4]], ncol = 3, widths=c(1,1,1.14))

```



```{r}
## G  
Vmn2r91 <- pheatmap(filtered.vno.mat.list$Vmn2r91, fontsize = 24, cluster_cols = FALSE, 
                     cluster_rows = FALSE, main="Vmn2r91", show_rownames = FALSE, silent=TRUE)
Vmn2r118 <- pheatmap(filtered.vno.mat.list$Vmn2r118, fontsize = 24, cluster_cols = FALSE, 
                    cluster_rows = FALSE, main="Vmn2r118", silent=TRUE)

svglite(filename = paste0(image_dir,"fig7_g.svg"),width=16,height=8)
grid.arrange(Vmn2r91[[4]], Vmn2r118[[4]], ncol = 2, widths=c(1,1.14))
invisible(dev.off())

```

```{r,fig.width=16,fig.height=8}

grid.arrange(Vmn2r91[[4]], Vmn2r118[[4]], ncol = 2, widths=c(1,1.16))

```





```{r}

# Correlated V1Rs - alternative
pheatmap(filtered.vno.mat.list$Vmn1r56, fontsize = 30, cluster_cols = FALSE, cluster_rows = FALSE, main="Vmn1r56")
pheatmap(filtered.vno.mat.list$Vmn1r60, fontsize = 30, cluster_cols = FALSE, cluster_rows = FALSE, main="Vmn1r60")
pheatmap(filtered.vno.mat.list$Vmn1r61, fontsize = 30, cluster_cols = FALSE, cluster_rows = FALSE, main="Vmn1r61")

# Correlated V2Rs - expected
pheatmap(filtered.vno.mat.list$Vmn2r1, fontsize = 30, cluster_cols = FALSE, cluster_rows = FALSE, main="Vmn2r1")
pheatmap(filtered.vno.mat.list$Vmn2r3, fontsize = 30, cluster_cols = FALSE, cluster_rows = FALSE, main="Vmn2r3")
pheatmap(filtered.vno.mat.list$Vmn2r7, fontsize = 30, cluster_cols = FALSE, cluster_rows = FALSE, main="Vmn2r7")

# Correlated V2Rs - expected
pheatmap(filtered.vno.mat.list$Vmn2r8, fontsize = 30, cluster_cols = FALSE, cluster_rows = FALSE, main="Vmn2r8")
pheatmap(filtered.vno.mat.list$Vmn2r73, fontsize = 30, cluster_cols = FALSE, cluster_rows = FALSE, main="Vmn2r73")
pheatmap(filtered.vno.mat.list$Vmn2r88, fontsize = 30, cluster_cols = FALSE, cluster_rows = FALSE, main="Vmn2r88")

# Uncorrelated V2Rs
pheatmap(filtered.vno.mat.list$Vmn2r91, fontsize = 30, cluster_cols = FALSE, cluster_rows = FALSE, main="Vmn2r91")
pheatmap(filtered.vno.mat.list$Vmn2r118, fontsize = 30, cluster_cols = FALSE, cluster_rows = FALSE, main="Vmn2r118")

```














```{r, eval=FALSE}

# Figure 6 - Resolve Molecular Cartography {.tabset .tabset-pills}  

load(file=paste0(object_dir,"resolve.seu.POSTpredict"))
load(file=paste0(object_dir,"resolve.datasets.POSTpredict"))
load(file=paste0(object_dir,"resolve.predictions.assay"))

```

```{r}

resolve.cell.types <- 
  as.factor(unlist(lapply(colnames(resolve.predictions.assay), function(cell){
  rownames(resolve.predictions.assay)[which.max(x=resolve.predictions.assay[ ,cell])]
})))

```

```{r}

Idents(resolve.seu.integrated) <- resolve.cell.types
levels(resolve.seu.integrated@active.ident)[levels(resolve.seu.integrated@active.ident)=="unknown-mVSN"] <- "sVSN"

levels(resolve.seu.integrated@active.ident) <- gsub("-","_", levels(resolve.seu.integrated@active.ident))

resolve.seu.integrated@active.ident <-
  factor(as.factor(resolve.seu.integrated@active.ident), 
         levels = c("V1R_mVSN","V2R_mVSN","mOSN","sVSN",
                    "V1R_iVSN","V2R_iVSN","INP","GBC","HBC","MV",
                    "SUS","OEC","Endothelial","LP","Microglia",
                    "Microglia_rp","Immune_Fpr","T_cells"))

```


```{r}

## B  

resolve.cell_type.colors <- hue_pal()(length(levels(resolve.seu.integrated@active.ident)))

names(resolve.cell_type.colors) <- levels(resolve.seu.integrated@active.ident)

resolve.cell_type.colors["V1R_mVSN"] <- "#00FF00"   #green
resolve.cell_type.colors["V2R_mVSN"] <- "#FF7F50"   #coral
resolve.cell_type.colors["sVSN"] <- "#8F00FF"       #violet 
resolve.cell_type.colors["LP"] <- "#0000FF"         #blue   
resolve.cell_type.colors["SUS"] <- "#FF0000"        #red
resolve.cell_type.colors["HBC"] <- "#FFFF00"        #yellow
resolve.cell_type.colors["V1R_iVSN"] <- "#DFFF00"   #chartreuse 
resolve.cell_type.colors["V2R_iVSN"] <- "#FFA500"   #orange

```

```{r}

colors <- c("V1R_mVSN"="#45B500","V2R_mVSN"="#E7851E","mOSN"="#D39200","sVSN"="#9C8DFF",
            "V1R_iVSN"="#00C087","V2R_iVSN"="#D09400","INP"="#FF61C7","GBC"="#EFFD5F", 
            "HBC"="#FFFF00","MV"="#00C0B2","SUS"="#F8766D","OEC"="#00BCD6",
            "Endothelial"="#00B3F2","LP"="#29A3FF","Microglia"="#00BC51","Microglia_rp"="#D277FF",
            "Immune_Fpr"="#B2A100","T_cells"="#F166E8")

```

```{r}

# Figure 6B, part I
fig6_b.part1 <- DimPlot(resolve.seu.integrated,cols=colors) + 
  theme(text=element_text(size=24),axis.text=element_text(size=18),aspect.ratio=1)

svglite(filename = paste0(adobe_images,"fig6_b.part1.svg"), width=9, height=9)
fig6_b.part1 
invisible(dev.off())
fig6_b.part1

```


```{r}

fig6.samples <- resolve.datasets[["slide7.B2_2"]]
cell_names <- colnames(fig6.samples)
cell_types <- resolve.seu.integrated@active.ident[cell_names]
fig6.samples <- AddMetaData(fig6.samples, cell_types, col.name="cell_type")
fig6_b <- SpatialDimPlot(object=fig6.samples, group.by="cell_type", 
                         pt.size.factor=2, cols = colors) +
  theme(aspect.ratio=1, legend.title=element_text(size=24),
        legend.text=element_text(size=18),
        legend.box.background=element_rect(fill="blue",colour="transparent"),
        legend.key = element_rect(fill="transparent",colour="transparent")) 

```

```{r, fig.width=10, fig.height=10}

svglite(filename = paste0(adobe_images,"fig6_b.svg"), width=10, height=10)
fig6_b 
invisible(dev.off())
fig6_b

```


```{r}

## C - HBC/GBC/INP

fig6_c.p1 <- SpatialFeaturePlot(fig6.samples, features=rownames(resolve.predictions.assay)[10],
                                pt.size.factor=4) + theme(aspect.ratio=1) +
  scale_fill_gradient(low="grey85",high="magenta")

fig6_c.p2 <- SpatialFeaturePlot(fig6.samples, features=rownames(resolve.predictions.assay)[17],
                                pt.size.factor=4) + theme(aspect.ratio=1) +
  scale_fill_gradient(low="grey85",high="magenta")

fig6_c.p3 <- SpatialFeaturePlot(fig6.samples, features=rownames(resolve.predictions.assay)[8],
                                pt.size.factor=4) + theme(aspect.ratio=1) +
  scale_fill_gradient(low="grey85",high="magenta")

fig6_c <- ggarrange(fig6_c.p1, fig6_c.p2, fig6_c.p3, ncol=1, nrow=3)

```

```{r, fig.width=5, fig.height=15}

svglite(filename = paste0(adobe_images,"fig6_c.svg"), width=5, height=15)
fig6_c 
invisible(dev.off())
fig6_c

```


```{r}

## D - V1r & V2r Immature VSNs

fig6_d.p1 <- SpatialFeaturePlot(fig6.samples, features=rownames(resolve.predictions.assay)[7],
                                pt.size.factor=4) + theme(aspect.ratio=1) +
  scale_fill_gradient(low="grey85",high="magenta")

fig6_d.p2 <- SpatialFeaturePlot(fig6.samples, features=rownames(resolve.predictions.assay)[6],
                                pt.size.factor=4) + theme(aspect.ratio=1) +
  scale_fill_gradient(low="grey85",high="magenta")

fig6_d <- ggarrange(fig6_d.p1, fig6_d.p2, ncol=2, nrow=1)

```

```{r, fig.width=10, fig.height=5}

svglite(filename = paste0(adobe_images,"fig6_d.svg"), width=10, height=5)
fig6_d 
invisible(dev.off())
fig6_d

```


```{r}

## E - V1R & V2R Mature VSNs  

fig6_e.p1 <- SpatialFeaturePlot(fig6.samples, features=rownames(resolve.predictions.assay)[1],
                                pt.size.factor=4) + theme(aspect.ratio=1) +
  scale_fill_gradient(low="grey85",high="magenta")

fig6_e.p2 <- SpatialFeaturePlot(fig6.samples, features=rownames(resolve.predictions.assay)[2],
                                pt.size.factor=4) + theme(aspect.ratio=1) +
  scale_fill_gradient(low="grey85",high="magenta")

fig6_e <- ggarrange(fig6_e.p1, fig6_e.p2, ncol=2, nrow=1)

```

```{r, fig.width=10, fig.height=5}

svglite(filename = paste0(adobe_images,"fig6_e.svg"), width=10, height=5)
fig6_e 
invisible(dev.off())
fig6_e

```



```{r}

## F - sVSNs & Lamina Propria  

fig6_f.p1 <- SpatialFeaturePlot(fig6.samples, features=rownames(resolve.predictions.assay)[5],
                                pt.size.factor=4) + theme(aspect.ratio=1) +
  scale_fill_gradient(low="grey85",high="magenta")

fig6_f.p2 <- SpatialFeaturePlot(fig6.samples, features=rownames(resolve.predictions.assay)[11],
                                pt.size.factor=4) + theme(aspect.ratio=1) +
  scale_fill_gradient(low="grey85",high="magenta")

fig6_f <- ggarrange(fig6_f.p1, fig6_f.p2, ncol=2, nrow=1)

```

```{r, fig.width=10, fig.height=5}

svglite(filename = paste0(adobe_images,"fig6_f.svg"), width=10, height=5)
fig6_f 
invisible(dev.off())
fig6_f

```





```{r}

# Figure 9 - Vomeronasal Receptors and Transcription Factors {.tabset .tabset-pills}  

vr.cor.transcription_factor <- readRDS(file=paste0(object_dir,"vr.cor.transcription_factor.rds"))
# Data frame with receptor sequence similarity scores and receptor average axon-guidance gene expression  
receptor.seq_sim.cor.tf_sim.df <- readRDS(file=paste0(object_dir,"receptor.seq_sim.cor.tf_sim.df.rds"))
# All significant VRs overlapping with surface expressed axon-guidance genes 
transcription_factor.b_in_a.mat <- readRDS(file=paste0(object_dir,"transcription_factor.b_in_a.mat.rds"))
# The above subset for strong groupings and split into V1R, V2R, and mixed-class matrices
v1r.transcription_factor.b_in_a.mat <- readRDS(file=paste0(object_dir,"v1r.transcription_factor.b_in_a.mat.rds"))
v2r.transcription_factor.b_in_a.mat <- readRDS(file=paste0(object_dir,"v2r.transcription_factor.b_in_a.mat.rds"))

```


```{r,fig.height=10,fig.width=10}

## A - Vomeronasal Receptor Correlation of Transcription Factor Expression  

transcription_factor.cluster.n <- 20

fig9_a <-
  pheatmap(vr.cor.transcription_factor,fontsize=20,show_rownames=FALSE,show_colnames=FALSE,
           cutree_rows=transcription_factor.cluster.n,cutree_cols=transcription_factor.cluster.n)

# Save image to svg file
svglite(filename = paste0(adobe_images,"fig9_a.svg"),height=10,width=10)
grid.draw(rectGrob(gp=gpar(fill="white", lwd=0)))
grid.draw(fig9_a[[4]])
invisible(dev.off())

```


```{r}

## B - Does the similarity of VR transcription factor gene expression correlate with the similarity of VR gene sequences?  

fig9_b <- ggplot(data=receptor.seq_sim.cor.tf_sim.df, aes(x=vr_seq_sim, y=tf_exp_cor)) + 
  geom_point(size=0.5) + labs(y="Transcription Factor Similarity",x="VR Similarity") +
  geom_pointdensity() + scale_color_viridis() + theme_classic() +
  stat_summary_bin(fun.y='mean', bins=100, color='orange', 
                   geom='line', orientation='x', size=1.5, show.legend=TRUE) + 
  stat_summary_bin(fun.y='median', bins=100, color = 'blue', 
                   geom='line', linetype='dashed', orientation='x', size=1.5, show.legend=TRUE) +
  stat_summary_bin(fun.y='mean', bins=100, 
                   color='magenta', geom='line', orientation='y', size=1.5, show.legend=TRUE) + 
  stat_summary_bin(fun.y='median', bins=100, color = 'cyan', 
                   geom='line', linetype='dashed', orientation='y', size=1.5, show.legend=TRUE) +
  theme(legend.position="left",aspect.ratio=1,axis.text=element_text(size=20),legend.text=element_text(size=24),
        axis.title=element_text(size=32),legend.title=element_text(size=24),panel.border=element_blank(),
        plot.background=element_blank())
  
```

```{r, fig.width=10, fig.height=10}

svglite(filename = paste0(adobe_images,"fig9_b.svg"), width=10, height=10)
fig9_b
invisible(dev.off())
fig9_b

```


```{r}

## C  

dim(v1r.transcription_factor.b_in_a.mat)

```

```{r,fig.height=5,fig.width=10}

# The dendrogram for rows
dst <- dist(v1r.transcription_factor.b_in_a.mat)
hc_row <- hclust(dst)
row_dend <- as.dendrogram(hc_row)
row_dend <- seriate_dendrogram(row_dend, dst, method="OLO")

# The dendrogram for columns
dst <- dist(t(v1r.transcription_factor.b_in_a.mat))
hc_row <- hclust(dst)
col_dend <- as.dendrogram(hc_row)
col_dend <- seriate_dendrogram(col_dend, dst, method="OLO")

fig9_c <- pheatmap(v1r.transcription_factor.b_in_a.mat,fontsize=8,treeheight_row=25,treeheight_col=25,
                                cluster_rows=as.hclust(row_dend), cluster_cols=as.hclust(col_dend))
# Save image to svg file
svglite(filename = paste0(adobe_images,"fig9_c.svg"),height=5,width=10)
grid.draw(rectGrob(gp=gpar(fill="white", lwd=0)))
grid.draw(fig9_c[[4]])
invisible(dev.off())


```


```{r}

## D  

dim(v2r.transcription_factor.b_in_a.mat)

```

```{r,fig.height=9,fig.width=14}

# The dendrogram for rows
dst <- dist(v2r.transcription_factor.b_in_a.mat)
hc_row <- hclust(dst)
row_dend <- as.dendrogram(hc_row)
row_dend <- seriate_dendrogram(row_dend, dst, method="OLO")

# The dendrogram for columns
dst <- dist(t(v2r.transcription_factor.b_in_a.mat))
hc_row <- hclust(dst)
col_dend <- as.dendrogram(hc_row)
col_dend <- seriate_dendrogram(col_dend, dst, method="OLO")

fig9_d <- pheatmap(v2r.transcription_factor.b_in_a.mat,fontsize=8,treeheight_row=25,treeheight_col=25,
                   cluster_rows=as.hclust(row_dend), cluster_cols=as.hclust(col_dend))
# Save image to svg file
svglite(filename = paste0(adobe_images,"fig9_d.svg"),height=9,width=14)
grid.draw(rectGrob(gp=gpar(fill="white", lwd=0)))
grid.draw(fig9_d[[4]])
invisible(dev.off())

```



















```{r, include=FALSE}

## G - Box Plot: Receptor Statistics  

# Capture the receptor metadata in a dataframe
receptor.metadata <- 
  FetchData(mature.neurons,vars=c("cell_type",
                                  "percent.V1R","percent.V2R","percent.Olfr","percent.Fpr",
                                  "V1R.receptor.cnt","V2R.receptor.cnt","Olfr.receptor.cnt","Fpr.receptor.cnt",
                                  "V1R.receptor.ratio","V2R.receptor.ratio","Olfr.receptor.ratio","Fpr.receptor.ratio"))

```


```{r, fig.width=14, fig.height=10.5, include=FALSE}

my_comparisons <- list(c("V1R_mVSN","V2R_mVSN"),c("V2R_mVSN","mOSN"),c("mOSN","sVSN"),
                      c("V1R_mVSN","mOSN"),c("V2R_mVSN","sVSN"),c("V1R_mVSN","sVSN"))

my_theme <- theme(plot.title=element_text(hjust = 0.5,size=14), # Center the title and set font size
                  axis.title.x=element_blank(),axis.ticks.x=element_blank(), 
                  axis.text.x=element_blank(),aspect.ratio=1,
                  legend.key.size=unit(4,'mm'),
                  legend.title.align=0)

p1 <- ggplot(receptor.metadata,aes(x=cell_type,y=percent.V1R,fill=cell_type)) + geom_boxplot() +
  scale_fill_manual("Cell Type",values=mature.cell_type.colors) + theme_classic() +
  scale_y_log10() +
  stat_compare_means(comparisons=my_comparisons,tip.length=0.01,
                     method="wilcox.test",label="p.signif",vjust=0.3) +
  ggtitle("Percent V1R Expression") + ylab("count") + my_theme
p2 <- ggplot(receptor.metadata,aes(x=cell_type,y=percent.V2R,fill=cell_type)) + geom_boxplot() +
  scale_fill_manual("Cell Type",values=mature.cell_type.colors) + theme_classic() +
  scale_y_log10() +
  stat_compare_means(comparisons=my_comparisons,tip.length=0.01,
                     method="wilcox.test",label="p.signif",vjust=0.3) +
  ggtitle("Percent V2R Expression") + ylab("count") + my_theme
p3 <- ggplot(receptor.metadata,aes(x=cell_type,y=percent.Olfr,fill=cell_type)) + geom_boxplot() +
  scale_fill_manual("Cell Type",values=mature.cell_type.colors) + theme_classic() +
  scale_y_log10() +
  stat_compare_means(comparisons=my_comparisons,tip.length=0.01,
                     method="wilcox.test",label="p.signif",vjust=0.3) +
  ggtitle("Percent Olfr Expression") + ylab("percent") + my_theme 
p4 <- ggplot(receptor.metadata,aes(x=cell_type,y=percent.Fpr,fill=cell_type)) + geom_boxplot() +
  scale_fill_manual("Cell Type",values=mature.cell_type.colors) + theme_classic() +
  scale_y_log10() +
  stat_compare_means(comparisons=my_comparisons,tip.length=0.01,
                     method="wilcox.test",label="p.signif",vjust=0.3) +
  ggtitle("Percent Fpr Expression") + ylab("Log Normalized Count") + my_theme 
p5 <- ggplot(receptor.metadata,aes(x=cell_type,y=V1R.receptor.cnt,fill=cell_type)) + geom_boxplot() +
  scale_fill_manual("Cell Type",values=mature.cell_type.colors) + theme_classic() +
  scale_y_log10() +
  stat_compare_means(comparisons=my_comparisons,tip.length=0.01,
                     method="wilcox.test",label="p.signif",vjust=0.3) +
  ggtitle("V1R Count") + ylab("Log Normalized Count") + my_theme 
p6 <- ggplot(receptor.metadata,aes(x=cell_type,y=V2R.receptor.cnt,fill=cell_type)) + geom_boxplot() +
  scale_fill_manual("Cell Type",values=mature.cell_type.colors) + theme_classic() +
  scale_y_log10() +
  stat_compare_means(comparisons=my_comparisons,tip.length=0.01,
                     method="wilcox.test",label="p.signif",vjust=0.3) +
  ggtitle("V2R Count") + ylab("Log Normalized Count") + my_theme 
p7 <- ggplot(receptor.metadata,aes(x=cell_type,y=Olfr.receptor.cnt,fill=cell_type)) + geom_boxplot() +
  scale_fill_manual("Cell Type",values=mature.cell_type.colors) + theme_classic() +
  scale_y_log10() +
  stat_compare_means(comparisons=my_comparisons,tip.length=0.01,
                     method="wilcox.test",label="p.signif",vjust=0.3) +
  ggtitle("Olfr Count") + ylab("Log Normalized Count") + my_theme 
p8 <- ggplot(receptor.metadata,aes(x=cell_type,y=Fpr.receptor.cnt,fill=cell_type)) + geom_boxplot() +
  scale_fill_manual("Cell Type",values=mature.cell_type.colors) + theme_classic() +
  scale_y_log10() +
  stat_compare_means(comparisons=my_comparisons,tip.length=0.01,
                     method="wilcox.test",label="p.signif",vjust=0.3) +
  ggtitle("Fpr Count") + ylab("Log Normalized Count") + my_theme 
p9 <- ggplot(receptor.metadata,aes(x=cell_type,y=V1R.receptor.ratio,fill=cell_type)) + geom_boxplot() +
  scale_fill_manual("Cell Type",values=mature.cell_type.colors) + theme_classic() +
  scale_y_log10() +
  stat_compare_means(comparisons=my_comparisons,tip.length=0.01,
                     method="wilcox.test",label="p.signif",vjust=0.3) +
  ggtitle("V1R Ratio") + ylab("Log Normalized Count") + my_theme 
p10 <- ggplot(receptor.metadata,aes(x=cell_type,y=V2R.receptor.ratio,fill=cell_type)) + geom_boxplot() +
  scale_fill_manual("Cell Type",values=mature.cell_type.colors) + theme_classic() +
  scale_y_log10() +
  stat_compare_means(comparisons=my_comparisons,tip.length=0.01,
                     method="wilcox.test",label="p.signif",vjust=0.3) +
  ggtitle("V2R Ratio") + ylab("Log Normalized Count") + my_theme 
p11 <- ggplot(receptor.metadata,aes(x=cell_type,y=Olfr.receptor.ratio,fill=cell_type)) + geom_boxplot() +
  scale_fill_manual("Cell Type",values=mature.cell_type.colors) + theme_classic() +
  scale_y_log10() +
  stat_compare_means(comparisons=my_comparisons,tip.length=0.01,
                     method="wilcox.test",label="p.signif",vjust=0.3) +
  ggtitle("Olfr Ratio") + ylab("Log Normalized Count") + my_theme 
p12 <- ggplot(receptor.metadata,aes(x=cell_type,y=Fpr.receptor.ratio,fill=cell_type)) + geom_boxplot() +
  scale_fill_manual("Cell Type",values=mature.cell_type.colors) + theme_classic() +
  scale_y_log10() +
  stat_compare_means(comparisons=my_comparisons,tip.length=0.01,
                     method="wilcox.test",label="p.signif",vjust=0.3) +
  ggtitle("Fpr Ratio") + ylab("Log Normalized Count") + my_theme 
fig4_iv <- ggarrange(p1,p2,p3,p4,p5,p6,p7,p8,p9,p10,p11,p12,ncol=4,nrow=3,
                     common.legend=TRUE,align="hv",legend="right") +
  ggtitle("Receptor Characteristics by Cell-Type") +
  theme(plot.title = element_text(hjust = 0.5,face="bold",size=24))
svglite(filename = paste0(adobe_images,"fig4_iv.svg"), width=14, height=10.5)
fig4_iv
invisible(dev.off())

```


```{r}

fig4_iv

```



```{r, include=FALSE}

# {-}  

temp.variables <- c("receptor.names","receptor.df")
current.variables <- c(ls(),"current.variables")
rm(list=setdiff(current.variables,c(global.vars,temp.variables)))
invisible(gc())

```




















```{r, include=FALSE}

# Figure 3 - Age differences in V1R and V2R Lineages {.tabset .tabset-pills}  

load(file = paste0(object_dir,"v1r.slingshot.integrated.postPseudotime"))
load(file = paste0(fitGAM_dir, "v1r.neuron.sce.all_genes.cedar.tradeSeq"))
load(file = paste0(object_dir,"v1r.neuron.sds.slingshot"))
load(file = paste0(object_dir,"v1r.slingshot.count.matrix"))

load(file = paste0(object_dir,"v2r.slingshot.integrated.postPseudotime"))
load(file = paste0(fitGAM_dir, "v2r.neuron.sce.all_genes.cedar.tradeSeq"))
load(file = paste0(object_dir,"v2r.neuron.sds.slingshot"))
load(file = paste0(object_dir,"v2r.slingshot.count.matrix"))

load(file=paste0(fitGAM_dir, "v1r.neuron.condRes.tradeSeq"))
load(file=paste0(fitGAM_dir, "v2r.neuron.condRes.tradeSeq"))
v1r.BP.condRes <- openxlsx::read.xlsx(paste0(object_dir,"v1r.BP.conditionResults.xlsx"))
v1r.MF.condRes <- openxlsx::read.xlsx(paste0(object_dir,"v1r.MF.conditionResults.xlsx"))
v1r.CC.condRes <- openxlsx::read.xlsx(paste0(object_dir,"v1r.CC.conditionResults.xlsx"))
v2r.BP.condRes <- openxlsx::read.xlsx(paste0(object_dir,"v2r.BP.conditionResults.xlsx"))
v2r.MF.condRes <- openxlsx::read.xlsx(paste0(object_dir,"v2r.MF.conditionResults.xlsx"))
v2r.CC.condRes <- openxlsx::read.xlsx(paste0(object_dir,"v2r.CC.conditionResults.xlsx"))

```


```{r, fig.width=10, fig.height=10, include=FALSE}

## A - 2D PCA of V1R and V2R lineage  

fig3_i <- DimPlot(v1r.slingshot.integrated,pt.size=2) +
  ggtitle("V1R Lineage") +
  theme(aspect.ratio = 1)
svglite(filename = paste0(adobe_images,"fig3_i.svg"), width=10, height=10)
fig3_i
invisible(dev.off())
fig3_i

```


```{r, fig.width=10, fig.height=10, include=FALSE}

fig3_ii <- DimPlot(v2r.slingshot.integrated, pt.size = 2) +
  ggtitle("V2R Lineage") +
  theme(aspect.ratio = 1)
svglite(filename = paste0(adobe_images,"fig3_ii.svg"), width=10, height=10)
fig3_ii
invisible(dev.off())
fig3_ii

```


```{r, fig.width=8,fig.height=16}

fig3_a <- ggarrange(plotlist=list(V1R=fig3_i,V2R=fig3_ii), nrow=2, ncol=1)

svglite(filename = paste0(adobe_images,"fig3_a.svg"), width=8, height=16)
fig3_a
invisible(dev.off())

fig3_a

```


```{r, include=FALSE}

## B - Pseudotime Principal Curves  
# Custom color palette  
pal <- viridis(100, end = 0.95)

```


```{r}

# Save image to file
svglite(filename=paste0(adobe_images,"fig3_b.svg"),width=10,height=10)

colors <- pal[cut(v1r.slingshot.integrated$v1r.pseudotime, breaks = 100)]
plot(reducedDim(v1r.neuron.sds), col = colors, pch = 16, cex = 1, main = "V1R Pseudotime", asp=1)
lines(v1r.neuron.sds, lwd = 2, col = 'black')

invisible(dev.off())

# Print image to screen
plot(reducedDim(v1r.neuron.sds), col = colors, pch = 16, cex = 1, main = "V1R Pseudotime", asp=1)
lines(v1r.neuron.sds, lwd = 2, col = 'black')

```


```{r}

# Save image to file
svglite(filename=paste0(adobe_images,"fig3_iv.svg"),width=10,height=10)

colors <- pal[cut(v2r.slingshot.integrated$v2r.pseudotime, breaks = 100)]
plot(reducedDim(v2r.neuron.sds), col = colors, pch = 16, cex = 1, main = "V2R Pseudotime", asp=1)
lines(v2r.neuron.sds, lwd = 2, col = 'black')

invisible(dev.off())

# Print image to screen
plot(reducedDim(v2r.neuron.sds), col = colors, pch = 16, cex = 1, main = "V2R Pseudotime", asp=1)
lines(v2r.neuron.sds, lwd = 2, col = 'black')

```


```{r, include=FALSE}

# Extract Metadata to Plot w/ Pseudotime
v1r.slingshot.integrated.metadata <- v1r.slingshot.integrated@meta.data
v2r.slingshot.integrated.metadata <- v2r.slingshot.integrated@meta.data

```


```{r, include=FALSE}

## C - Density plots along pseudotime  

fig3_v <- ggdensity(data=subset(v1r.slingshot.integrated.metadata,
                                !is.na(v1r.slingshot.integrated.metadata)),"v1r.pseudotime",
                    color="age",fill="age",title="V1R Lineage") +
  theme(aspect.ratio = 1) + xlab("pseudotime") + 
  annotate("text",x=90,y=0.06,label="Kolmogorov-Smirnov Test: p-value < 2.2e-16")

svglite(filename = paste0(adobe_images,"fig3_v.svg"), width=10, height=10)
fig3_v
invisible(dev.off())
fig3_v

```


```{r, include=FALSE}

## V1R Two-sample Kolmogorov-Smirnov test  
v1r.pseudotime <- v1r.slingshot.integrated.metadata[ ,"v1r.pseudotime",drop=FALSE]

# Get age specific cells for statistical test
p14.v1r.pseudotime <- v1r.pseudotime[grep("^P14", rownames(v1r.pseudotime)), ]
p56.v1r.pseudotime <- v1r.pseudotime[grep("^P56", rownames(v1r.pseudotime)), ]
# Remove NAs for cells that don't fall in the lineage
p14.v1r.pseudotime <- na.omit(p14.v1r.pseudotime)
p56.v1r.pseudotime <- na.omit(p56.v1r.pseudotime)
###############################
### Kolmogorov-Smirnov Test ###  
###############################
ks.test(p14.v1r.pseudotime,p56.v1r.pseudotime)

```


```{r, include=FALSE}

fig3_vi <- ggdensity(data=subset(v2r.slingshot.integrated.metadata,
                                !is.na(v2r.slingshot.integrated.metadata)),"v2r.pseudotime",
                    color="age",fill="age",title="V2R Lineage") +
  theme(aspect.ratio = 1) + xlab("pseudotime") + 
  annotate("text",x=75,y=0.06,label="Kolmogorov-Smirnov Test: p-value = 2.339e-10")

svglite(filename = paste0(adobe_images,"fig3_vi.svg"), width=10, height=10)
fig3_vi
invisible(dev.off())
fig3_vi

```

```{r, include=FALSE}

## V2R Two-sample Kolmogorov-Smirnov test  
v2r.pseudotime <- v2r.slingshot.integrated.metadata[ ,"v2r.pseudotime",drop=FALSE]

# Get age specific cells for statistical test
p14.v2r.pseudotime <- v2r.pseudotime[grep("^P14", rownames(v2r.pseudotime)),]
p56.v2r.pseudotime <- v2r.pseudotime[grep("^P56", rownames(v2r.pseudotime)),]
# Remove NAs for cells that don't fall in the lineage
p14.v2r.pseudotime <- na.omit(p14.v2r.pseudotime)
p56.v2r.pseudotime <- na.omit(p56.v2r.pseudotime)
########################
### Kolmogorov-Smirnov Test
########################
ks.test(p14.v2r.pseudotime,
        p56.v2r.pseudotime)

```


```{r, fig.width=8,fig.height=16,include=FALSE}

fig3_c <- ggarrange(plotlist=list(V1R=fig3_v,V2R=fig3_vi),nrow=2,ncol=1,common.legend=TRUE)

svglite(filename = paste0(adobe_images,"fig3_c.svg"), width=8, height=16)
fig3_c
invisible(dev.off())

```


```{r}

fig3_c

```


```{r, include=FALSE}

## D - Heatmap: V1R DEGs by Age  

v1r.neuron.condRes <- na.omit(v1r.neuron.condRes)

# Significantly Differetially Expressed Genes (DEGs)
sig.v1r.condRes <- rownames(v1r.neuron.condRes[v1r.neuron.condRes$padj <= 0.05, ])

```


```{r, fig.width=10, fig.height=10}

v1r.clusters <- 7

v1r.condRes.yhatSmooth <- predictSmooth(v1r.neuron.sce,gene=sig.v1r.condRes,nPoints=50,tidy=FALSE)
v1r.condRes.yhatSmoothScaled <- t(scale(t(v1r.condRes.yhatSmooth)))
v1r.condRes.heatSmooth_P14 <- pheatmap(v1r.condRes.yhatSmoothScaled[ ,1:50], 
                                       cluster_cols=FALSE, cutree_rows = v1r.clusters,
                                       show_rownames=FALSE,show_colnames=FALSE, 
                                       main="P14 V1R", legend=FALSE, silent=TRUE)

v1r.cl <- cutree(v1r.condRes.heatSmooth_P14$tree_row, k = v1r.clusters)
v1r.condRes.heatmap.genes <-
  rownames(v1r.condRes.yhatSmoothScaled[v1r.condRes.heatSmooth_P14$tree_row$order, ])
v1r.gene2cluster <- setNames(nm=c('gene','cluster'),stack(v1r.cl[v1r.condRes.heatmap.genes])[2:1])
data.table::fwrite(v1r.gene2cluster, sep="\t",
                   file=paste0(object_dir,"v1r.condRes.heatmap.txt"))

v1r.cl.gaps <- cumsum(table(v1r.cl[v1r.condRes.heatmap.genes])[unique(v1r.cl[v1r.condRes.heatmap.genes])])

v1r.condRes.matchingHeatmap_P56 <- 
  pheatmap(v1r.condRes.yhatSmoothScaled[v1r.condRes.heatSmooth_P14$tree_row$order,51:100],
           cluster_cols=FALSE, cluster_rows=FALSE, gaps_row = v1r.cl.gaps, 
           show_rownames=FALSE, show_colnames=FALSE, #cellwidth=10,cellheight=0.1,
           main="P56 V1R", legend=TRUE, silent=TRUE)

svglite(filename = paste0(adobe_images,"fig3_vii.svg"),width=10,height=10)
fig3_vii <- grid.arrange(v1r.condRes.heatSmooth_P14[[4]], v1r.condRes.matchingHeatmap_P56[[4]], ncol = 2)
grid.draw(rectGrob(gp=gpar(fill="white", lwd=0)))
grid.draw(fig3_vii)
invisible(dev.off())

grid.draw(rectGrob(gp=gpar(fill="white", lwd=0)))
grid.draw(fig3_vii)

```



```{r, include=FALSE}

## E - Heatmap: V2R DEGs by Age  
v2r.neuron.condRes <- na.omit(v2r.neuron.condRes)

# Significantly Differetially Expressed Genes (DEGs)
sig.v2r.condRes <- rownames(v2r.neuron.condRes[v2r.neuron.condRes$padj <= 0.05, ])

```


```{r, fig.width=10, fig.height=10}

v2r.clusters <- 5

v2r.condRes.yhatSmooth <- predictSmooth(v2r.neuron.sce,gene=sig.v2r.condRes,nPoints=50,tidy=FALSE)
v2r.condRes.yhatSmoothScaled <- t(scale(t(v2r.condRes.yhatSmooth)))
v2r.condRes.heatSmooth_P14 <- pheatmap(v2r.condRes.yhatSmoothScaled[ ,1:50],
                                       cluster_cols=FALSE, cutree_rows = v2r.clusters,
                                       show_rownames=FALSE,show_colnames=FALSE, 
                                       main="P14 V2R", legend=FALSE, silent=TRUE)

v2r.cl <- cutree(v2r.condRes.heatSmooth_P14$tree_row, k=v2r.clusters)
v2r.condRes.heatmap.genes <-
  rownames(v2r.condRes.yhatSmoothScaled[v2r.condRes.heatSmooth_P14$tree_row$order, ])
v2r.gene2cluster <- setNames(nm=c('gene','cluster'),stack(v2r.cl[v2r.condRes.heatmap.genes])[2:1])
data.table::fwrite(v2r.gene2cluster, sep="\t",
                   file=paste0(object_dir,"v2r.condRes.heatmap.txt"))

v2r.cl.gaps <- cumsum(table(v2r.cl[v2r.condRes.heatmap.genes])[unique(v2r.cl[v2r.condRes.heatmap.genes])])

v2r.condRes.matchingHeatmap_P56 <- 
pheatmap(v2r.condRes.yhatSmoothScaled[v2r.condRes.heatSmooth_P14$tree_row$order,51:100],
           cluster_cols=FALSE, cluster_rows=FALSE, gaps_row = v2r.cl.gaps,
           show_rownames=FALSE, show_colnames=FALSE, 
           main="P56 V2R", legend=TRUE, silent=TRUE)

svglite(filename = paste0(adobe_images,"fig3_viii.svg"),width=10,height=10)
fig3_viii <- grid.arrange(v2r.condRes.heatSmooth_P14[[4]], v2r.condRes.matchingHeatmap_P56[[4]], ncol = 2)
grid.draw(rectGrob(gp=gpar(fill="white", lwd=0)))
grid.draw(fig3_viii)
invisible(dev.off())

grid.draw(rectGrob(gp=gpar(fill="white", lwd=0)))
grid.draw(fig3_viii)

```


```{r, include=FALSE}

### V1R ###
v1r.neuron.condRes$padj <- as.numeric(scientific(v1r.neuron.condRes$padj))

```


```{r, include=FALSE, eval=FALSE}

## Plot Significant DEGs Between P14 and Adult, over Pseudotime  
# create the plot list with geneSymbol as title, padj as subtitle
v1r.neuron.condRes.pl <- lapply(v1r.condRes.heatmap.genes, function(goi){
  plotSmoothers(v1r.neuron.sce, assays(v1r.neuron.sce)$counts,
                gene = goi,
                alpha = 1, border = TRUE) + 
    ggtitle(goi, subtitle = paste("padj:",v1r.neuron.condRes[goi,"padj"])) +
    theme(aspect.ratio=1)
})

# What are the dimensions of the multi-plot page?
v1r.neuron.condRes.ml <- ggarrange(plotlist=v1r.neuron.condRes.pl, nrow=2, ncol=2, common.legend = TRUE)

# Export the plot to a pdf
ggexport(v1r.neuron.condRes.ml, 
         filename = paste0(image_dir,"v1r.neuron.condRes.DEGs.pdf"))

```


```{r, include=FALSE}

### V2R ###  
v2r.neuron.condRes$padj <- as.numeric(scientific(v2r.neuron.condRes$padj))

```


```{r, include=FALSE, eval=FALSE}

# create the plot list with geneSymbol as title, padj as subtitle
v2r.neuron.condRes.pl <- lapply(v2r.condRes.heatmap.genes, function(goi){
  plotSmoothers(v2r.neuron.sce, assays(v2r.neuron.sce)$counts,
                gene = goi,
                alpha = 1, border = TRUE) + 
    ggtitle(goi, subtitle = paste("padj:",v2r.neuron.condRes[goi,"padj"])) +
    theme(aspect.ratio=1)
})

# What are the dimensions of the multi-plot page?
v2r.neuron.condRes.ml <- ggarrange(plotlist=v2r.neuron.condRes.pl, nrow=2, ncol=2, common.legend = TRUE)

# Export the plot to a pdf
ggexport(v2r.neuron.condRes.ml, 
         filename = paste0(image_dir,"v2r.neuron.condRes.DEGs.pdf"))

```


```{r, include=FALSE}

## F - Gene Set Enrichment Plots: __V1R__ lineage P14 vs P56  
# Combind GO Results for P14 vs P56 in V1R lineage
v1r.condRes.GO <- rbind(v1r.BP.condRes,v1r.MF.condRes,v1r.CC.condRes)
# Just significant results
v1r.condRes.GO <- v1r.condRes.GO[v1r.condRes.GO$padj <= 0.05, ]
# Just GO Terms with a positive NES
v1r.condRes.GO <- v1r.condRes.GO[v1r.condRes.GO$NES > 0, ]
# # Just GO Terms with less than 1000 associated genes
# v1r.condRes.GO <- v1r.condRes.GO[v1r.condRes.GO$size < 1000, ]
# Create a column for GO category
v1r.condRes.GO$Category <- substring(v1r.condRes.GO$pathway,first=3,last=4)
v1r.condRes.GO$pathway <- substring(v1r.condRes.GO$pathway,first=6)
# Sort by normalized enrichment score
v1r.condRes.GO <- v1r.condRes.GO[order(v1r.condRes.GO$NES), ]
v1r.condRes.GO$pathway <- factor(v1r.condRes.GO$pathway, levels=v1r.condRes.GO$pathway)
v1r.condRes.GO <- na.omit(v1r.condRes.GO)

# Combind GO Results for P14 vs P56 in V2R lineage
v2r.condRes.GO <- rbind(v2r.BP.condRes,v2r.MF.condRes,v2r.CC.condRes)
# Just significant results
v2r.condRes.GO <- v2r.condRes.GO[v2r.condRes.GO$padj <= 0.05, ]
# Just GO Terms with a positive NES
v2r.condRes.GO <- v2r.condRes.GO[v2r.condRes.GO$NES > 0, ]
# # Just GO Terms with less than 1000 associated genes
# v2r.condRes.GO <- v2r.condRes.GO[v2r.condRes.GO$size < 1000, ]
# Create a column for GO category
v2r.condRes.GO$Category <- substring(v2r.condRes.GO$pathway,first=3,last=4) 
v2r.condRes.GO$pathway <- substring(v2r.condRes.GO$pathway,first=6)
# Sort by normalized enrichment score
v2r.condRes.GO <- v2r.condRes.GO[order(v2r.condRes.GO$NES), ]
v2r.condRes.GO$pathway <- factor(v2r.condRes.GO$pathway, levels=v2r.condRes.GO$pathway)
v2r.condRes.GO <- na.omit(v2r.condRes.GO)

v1r.condRes.GO
v2r.condRes.GO

```


```{r, include=FALSE}

# Final Filtering of GO Terms for Redundancy and Irrelevance
#v1r.condRes.GO <- v1r.condRes.GO[-c(3,6,12)]
v1r.condRes.GO <- v1r.condRes.GO[v1r.condRes.GO$size > 10, ]
#v2r.condRes.GO <- v2r.condRes.GO[-c(3,6,12)]
v2r.condRes.GO <- v2r.condRes.GO[v2r.condRes.GO$size > 10, ]


```


```{r, include=FALSE}

# Which heatmap clusters contain the genes associated with a given GO term?  
lapply(1:nrow(v1r.condRes.GO), function(i){
  gene.set <- strsplit(v1r.condRes.GO$leadingEdge[i],", ")[[1]]
  heatmap.genes <- intersect(gene.set,v1r.gene2cluster$gene)
  table(v1r.gene2cluster[v1r.gene2cluster$gene %in% heatmap.genes, "cluster"])
})

```


```{r, include=FALSE}

# Which heatmap clusters contain the genes associated with a given GO term?  
lapply(1:nrow(v2r.condRes.GO), function(i){
  gene.set <- strsplit(v2r.condRes.GO$leadingEdge[i],", ")[[1]]
  heatmap.genes <- intersect(gene.set,v2r.gene2cluster$gene)
  unique(v2r.gene2cluster$cluster)
  table(v2r.gene2cluster[v2r.gene2cluster$gene %in% heatmap.genes, "cluster"])
})

```


```{r fig.height=6, fig.width=14}

fig3_ix <- ggplot(v1r.condRes.GO, aes(x=pathway,y=NES,fill=Category)) +
  geom_bar(stat="identity") +
  theme(aspect.ratio = 1,plot.title=element_text(face="bold")) +
  ggtitle("Enriched GO Terms P14 Vs P56 in V1R Lineage") +
  xlab("GO Term") +
  coord_flip()

svglite(filename = paste0(adobe_images,"fig3_ix.svg"), width=14, height=6)
fig3_ix
invisible(dev.off())

fig3_ix

```


```{r fig.height=6, fig.width=14}

## G - Gene Set Enrichment Plots: __V2R__ lineage P14 vs P56  
fig3_x <- ggplot(v2r.condRes.GO, aes(x=pathway,y=NES,fill=Category)) +
  geom_bar(stat="identity") +
  theme(aspect.ratio = 1,plot.title=element_text(face="bold")) +
  ggtitle("Enriched GO Terms P14 Vs P56 in V2R Lineage") +
  xlab("GO Term") +
  coord_flip()

svglite(filename = paste0(adobe_images,"fig3_x.svg"), width=14, height=6)
fig3_x
invisible(dev.off())

fig3_x

```


```{r include=FALSE}

# {-}  
current.variables <- c(ls(),"current.variables")
rm(list=setdiff(current.variables,c(global.vars)))
invisible(gc())

```





```{r,include=FALSE}

## F - Box Plots: Receptor Coexpression Statistics  

# Receptor data supporting a role for co-expression
receptor.metadata <- 
  FetchData(mature.neurons,vars=c("cell_type","receptor.species","V1R.species","V2R.species",
                                  "Olfr.species","total.receptor.cnt","first.receptor.count",
                                  "second.receptor.count","third.receptor.count","Shannon.Index",
                                  "first.receptor.percent","second.receptor.percent","third.receptor.percent"))

```


```{r,fig.width=14,fig.height=10.5,include=FALSE}

# Characteristics of Receptors Across Cell-Types  
my_comparisons = list(c("V1R_mVSN","V2R_mVSN"),c("V2R_mVSN","mOSN"),c("mOSN","sVSN"),
                      c("V1R_mVSN","mOSN"),c("V2R_mVSN","sVSN"),c("V1R_mVSN","sVSN"))

my_theme <- theme(plot.title=element_text(hjust = 0.5,size=14), # Center the title and set font size
                  axis.title.x=element_blank(),axis.ticks.x=element_blank(), 
                  axis.text.x=element_blank(),aspect.ratio=1,
                  legend.key.size=unit(4,'mm'),
                  legend.title.align=0)

p1 <- ggplot(receptor.metadata,aes(x=cell_type,y=receptor.species,fill=cell_type)) + geom_violin() +
  scale_fill_manual("Cell Type",values=mature.cell_type.colors) + theme_classic() +
  stat_summary(fun.y=mean,geom='point',size=20,colour="black",shape=95,show_guide=FALSE) +
  stat_compare_means(comparisons=my_comparisons,tip.length=0.01,
                     method="wilcox.test",label="p.signif",vjust=0.3) +
  ggtitle("receptor.species") + ylab("count") + my_theme
p2 <- ggplot(receptor.metadata,aes(x=cell_type,y=V1R.species,fill=cell_type)) + geom_violin() +
  scale_fill_manual("Cell Type",values=mature.cell_type.colors) + theme_classic() +
  stat_summary(fun.y=mean,geom='point',size=20,colour="black",shape=95,show_guide=FALSE) +
  stat_compare_means(comparisons=my_comparisons,tip.length=0.01,
                     method="wilcox.test",label="p.signif",vjust=0.3) +
  ggtitle("V1R.species") + ylab("count") + my_theme
p3 <- ggplot(receptor.metadata,aes(x=cell_type,y=V2R.species,fill=cell_type)) + geom_violin() +
  scale_fill_manual("Cell Type",values=mature.cell_type.colors) + theme_classic() +
  stat_summary(fun.y=mean,geom='point',size=20,colour="black",shape=95,show_guide=FALSE) +
  stat_compare_means(comparisons=my_comparisons,tip.length=0.01,
                     method="wilcox.test",label="p.signif",vjust=0.3) +
  ggtitle("V2R.species") + ylab("percent") + my_theme 
p4 <- ggplot(receptor.metadata,aes(x=cell_type,y=Olfr.species,fill=cell_type)) + geom_violin() +
  scale_fill_manual("Cell Type",values=mature.cell_type.colors) + theme_classic() +
  stat_summary(fun.y=mean,geom='point',size=20,colour="black",shape=95,show_guide=FALSE) +
  stat_compare_means(comparisons=my_comparisons,tip.length=0.01,
                     method="wilcox.test",label="p.signif",vjust=0.3) +
  ggtitle("Olfr.species") + ylab("Log Normalized Count") + my_theme 
p5 <- ggplot(receptor.metadata,aes(x=cell_type,y=total.receptor.cnt,fill=cell_type)) + geom_violin() +
  scale_fill_manual("Cell Type",values=mature.cell_type.colors) + theme_classic() +
  scale_y_log10() +
  stat_summary(fun.y=mean,geom='point',size=20,colour="black",shape=95,show_guide=FALSE) +
  stat_compare_means(comparisons=my_comparisons,tip.length=0.01,
                     method="wilcox.test",label="p.signif",vjust=0.3) +
  ggtitle("total.receptor.cnt") + ylab("Log Normalized Count") + my_theme 
p6 <- ggplot(receptor.metadata,aes(x=cell_type,y=first.receptor.count,fill=cell_type)) + geom_violin() +
  scale_fill_manual("Cell Type",values=mature.cell_type.colors) + theme_classic() +
  scale_y_log10() +
  stat_summary(fun.y=mean,geom='point',size=20,colour="black",shape=95,show_guide=FALSE) +
  stat_compare_means(comparisons=my_comparisons,tip.length=0.01,
                     method="wilcox.test",label="p.signif",vjust=0.3) +
  ggtitle("first.receptor.count") + ylab("Log Normalized Count") + my_theme 
p7 <- ggplot(receptor.metadata,aes(x=cell_type,y=second.receptor.count,fill=cell_type)) + geom_violin() +
  scale_fill_manual("Cell Type",values=mature.cell_type.colors) + theme_classic() +
  scale_y_log10() +
  stat_summary(fun.y=mean,geom='point',size=20,colour="black",shape=95,show_guide=FALSE) +
  stat_compare_means(comparisons=my_comparisons,tip.length=0.01,
                     method="wilcox.test",label="p.signif",vjust=0.3) +
  ggtitle("second.receptor.count") + ylab("Log Normalized Count") + my_theme 
p8 <- ggplot(receptor.metadata,aes(x=cell_type,y=third.receptor.count,fill=cell_type)) + geom_violin() +
  scale_fill_manual("Cell Type",values=mature.cell_type.colors) + theme_classic() +
  scale_y_log10() +
  stat_summary(fun.y=mean,geom='point',size=20,colour="black",shape=95,show_guide=FALSE) +
  stat_compare_means(comparisons=my_comparisons,tip.length=0.01,
                     method="wilcox.test",label="p.signif",vjust=0.3) +
  ggtitle("third.receptor.count") + ylab("Log Normalized Count") + my_theme 
p9 <- ggplot(receptor.metadata,aes(x=cell_type,y=Shannon.Index,fill=cell_type)) + geom_violin() +
  scale_fill_manual("Cell Type",values=mature.cell_type.colors) + theme_classic() +
  stat_summary(fun.y=mean,geom='point',size=20,colour="black",shape=95,show_guide=FALSE) +
  stat_compare_means(comparisons=my_comparisons,tip.length=0.01,
                     method="wilcox.test",label="p.signif",vjust=0.3) +
  ggtitle("Shannon.Index") + ylab("Log Normalized Count") + my_theme 
p10 <- ggplot(receptor.metadata,aes(x=cell_type,y=first.receptor.percent,fill=cell_type)) + geom_violin() +
  scale_fill_manual("Cell Type",values=mature.cell_type.colors) + theme_classic() +
  stat_summary(fun.y=mean,geom='point',size=20,colour="black",shape=95,show_guide=FALSE) +
  stat_compare_means(comparisons=my_comparisons,tip.length=0.01,
                     method="wilcox.test",label="p.signif",vjust=0.3) +
  ggtitle("first.receptor.percent") + ylab("Log Normalized Count") + my_theme 
p11 <- ggplot(receptor.metadata,aes(x=cell_type,y=second.receptor.percent,fill=cell_type)) + geom_violin() +
  scale_fill_manual("Cell Type",values=mature.cell_type.colors) + theme_classic() +
  stat_summary(fun.y=mean,geom='point',size=20,colour="black",shape=95,show_guide=FALSE) +
  stat_compare_means(comparisons=my_comparisons,tip.length=0.01,
                     method="wilcox.test",label="p.signif",vjust=0.3) +
  ggtitle("second.receptor.percent") + ylab("Log Normalized Count") + my_theme 
p12 <- ggplot(receptor.metadata,aes(x=cell_type,y=third.receptor.percent,fill=cell_type)) + geom_violin() +
  scale_fill_manual("Cell Type",values=mature.cell_type.colors) + theme_classic() +
  stat_summary(fun.y=mean,geom='point',size=20,colour="black",shape=95,show_guide=FALSE) +
  stat_compare_means(comparisons=my_comparisons,tip.length=0.01,
                     method="wilcox.test",label="p.signif",vjust=0.3) +
  ggtitle("third.receptor.percent") + ylab("Log Normalized Count") + my_theme 
fig5_iv <- ggarrange(p1,p2,p3,p4,p5,p6,p7,p8,p9,p10,p11,p12,ncol=4,nrow=3,
                     common.legend=TRUE,align="hv",legend="right") +
  ggtitle("Receptor Coexpression Characteristics by Cell-Type") +
  theme(plot.title = element_text(hjust = 0.5,face="bold",size=24))
svglite(filename = paste0(adobe_images,"fig5_iv.svg"), width=14, height=10.5)
fig5_iv
invisible(dev.off())

```

```{r}

fig5_iv

```


```{r include=FALSE}

current.variables <- c(ls(),"current.variables")
rm(list=setdiff(current.variables,c(global.vars)))
invisible(gc())

```




```{r}

load(file=paste0(object_dir,"inp.integrated.preCellID"))            # INP Seurat Object

V1R_INP1_V2R_INP1.DEGs <- 
  openxlsx::read.xlsx(xlsxFile=paste0(object_dir,"v1r.v.v2r.inp1.markers.xlsx"),rowNames=TRUE)
V1R_INP2_V2R_INP2.DEGs <- 
  openxlsx::read.xlsx(xlsxFile=paste0(object_dir,"v1r.v.v2r.inp2.markers.xlsx"),rowNames=TRUE)
V1R_iVSN_V2R_iVSN.DEGs <- 
  read.xlsx(xlsxFile = paste0(object_dir,"V1R_iVSN_V2R_iVSN.DEGs.xlsx"),rowNames=TRUE)
V1R_mVSN_V2R_mVSN.DEGs <- 
  read.xlsx(xlsxFile = paste0(object_dir,"V1R_mVSN_V2R_mVSN.DEGs.xlsx"),rowNames=TRUE)

```

```{r}

# Let's get the significant V2R upregulated genes from each DEG test.
v2r.inp1.up <- rownames(V1R_INP1_V2R_INP1.DEGs[V1R_INP1_V2R_INP1.DEGs$p_val_adj <= 0.05 &
                                        V1R_INP1_V2R_INP1.DEGs$avg_log2FC < 0, ])
v2r.inp2.up <- rownames(V1R_INP2_V2R_INP2.DEGs[V1R_INP2_V2R_INP2.DEGs$p_val_adj <= 0.05 &
                                        V1R_INP2_V2R_INP2.DEGs$avg_log2FC < 0, ])
v2r.ivsn.up <- rownames(V1R_iVSN_V2R_iVSN.DEGs[V1R_iVSN_V2R_iVSN.DEGs$p_val_adj <= 0.05 &
                                                 V1R_iVSN_V2R_iVSN.DEGs$avg_log2FC < 0, ])
v2r.mvsn.up <- rownames(V1R_mVSN_V2R_mVSN.DEGs[V1R_mVSN_V2R_mVSN.DEGs$p_val_adj <= 0.05 &
                                                 V1R_mVSN_V2R_mVSN.DEGs$avg_log2FC < 0, ])

v2r.up.genes <- names(head(sort(table(sort(c(v2r.inp1.up,v2r.inp2.up,v2r.ivsn.up,v2r.mvsn.up))),decreasing=TRUE),14))

```

```{r,fig.width=12,fig.height=12}

#gene <- "Crabp1"
genes <- c("Gnao1","Cnpy1","Krt8","Dpysl3")

FeaturePlot(object=inp.integrated,features=genes,pt.size=1.5,
            order=TRUE,cols=c("cyan","magenta"),slot="data",ncol=2) & 
  theme(aspect.ratio=1)
FeaturePlot(object=neuron.integrated,features=genes,pt.size=1.5,
            order=TRUE,cols=c("cyan","magenta"),slot="data",ncol=2) & 
  theme(aspect.ratio=1)

```

# {-}  

###  *Thank You! Come Again!*